<!doctype html>
<html class="no-js" lang="en" data-content_root="">
  <head><meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="index" title="Index" href="genindex.html"><link rel="search" title="Search" href="search.html"><link rel="next" title="Contributing" href="contributing.html"><link rel="prev" title="Contents" href="index.html">

    <!-- Generated with Sphinx 7.1.2 and Furo 2025.12.19 -->
        <title>scran, in Python - scranpy 0.3.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo.css?v=7bdb33bb" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo-extensions.css?v=8dab3a3b" />
    
    


<style>
  body {
    --color-code-background: #eeffcc;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle site navigation sidebar">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc" aria-label="Toggle table of contents sidebar">
<label class="overlay sidebar-overlay" for="__navigation"></label>
<label class="overlay toc-overlay" for="__toc"></label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <span class="icon"><svg><use href="#svg-menu"></use></svg></span>
      </label>
    </div>
    <div class="header-center">
      <a href="index.html"><div class="brand">scranpy 0.3.0 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle" aria-label="Toggle Light / Dark / Auto color theme">
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <span class="icon"><svg><use href="#svg-toc"></use></svg></span>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="index.html">
  
  <span class="sidebar-brand-text">scranpy 0.3.0 documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1 current current-page"><a class="current reference internal" href="#">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributions</a></li>
<li class="toctree-l1"><a class="reference internal" href="license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="authors.html">Authors</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="api/modules.html">Module Reference</a><input aria-label="Toggle navigation of Module Reference" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><span class="icon"><svg><use href="#svg-arrow-right"></use></svg></span></label><ul>
<li class="toctree-l2"><a class="reference internal" href="api/scranpy.html">scranpy package</a></li>
</ul>
</li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="view-this-page">
  <a class="muted-link" href="_sources/readme.md.txt" title="View this page">
    <svg><use href="#svg-eye"></use></svg>
    <span class="visually-hidden">View this page</span>
  </a>
</div>
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle" aria-label="Toggle Light / Dark / Auto color theme">
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <span class="icon"><svg><use href="#svg-toc"></use></svg></span>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <!-- These are examples of badges you might want to add to your README:
     please update the URLs accordingly

[![Built Status](https://api.cirrus-ci.com/github/<USER>/scranpy.svg?branch=main)](https://cirrus-ci.com/github/<USER>/scranpy)
[![ReadTheDocs](https://readthedocs.org/projects/scranpy/badge/?version=latest)](https://scranpy.readthedocs.io/en/stable/)
[![Coveralls](https://img.shields.io/coveralls/github/<USER>/scranpy/main.svg)](https://coveralls.io/r/<USER>/scranpy)
[![Conda-Forge](https://img.shields.io/conda/vn/conda-forge/scranpy.svg)](https://anaconda.org/conda-forge/scranpy)
[![Twitter](https://img.shields.io/twitter/url/http/shields.io.svg?style=social&label=Twitter)](https://twitter.com/scranpy)
-->
<p><a class="reference external" href="https://pyscaffold.org/"><img alt="Project generated with PyScaffold" src="https://img.shields.io/badge/-PyScaffold-005CA0?logo=pyscaffold" /></a>
<a class="reference external" href="https://pypi.org/project/scranpy/"><img alt="PyPI-Server" src="https://img.shields.io/pypi/v/scranpy.svg" /></a>
<a class="reference external" href="https://pepy.tech/project/scranpy"><img alt="Downloads" src="https://static.pepy.tech/badge/scranpy/month" /></a>
<img alt="Unit tests" src="https://github.com/libscran/scranpy/actions/workflows/run-tests.yml/badge.svg" /></p>
<section class="tex2jax_ignore mathjax_ignore" id="scran-in-python">
<h1>scran, in Python<a class="headerlink" href="#scran-in-python" title="Permalink to this heading">¶</a></h1>
<section id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this heading">¶</a></h2>
<p>The <strong>scranpy</strong> package provides Python bindings to the single-cell analysis methods in the <a class="reference external" href="https://github.com/libscran"><strong>libscran</strong></a> C++ libraries.
It performs the standard steps in a typical single-cell analysis including quality control, normalization, feature selection, dimensionality reduction, clustering and marker detection.
This package is effectively a mirror of its counterparts in Javascript (<a class="reference external" href="https://npmjs.com/package/scran.js"><strong>scran.js</strong></a>) and R (<a class="reference external" href="https://github.com/libscran/scrapper"><strong>scrapper</strong></a>),
which are based on the same underlying C++ libraries and concepts.</p>
</section>
<section id="quick-start">
<h2>Quick start<a class="headerlink" href="#quick-start" title="Permalink to this heading">¶</a></h2>
<p>Let’s fetch a dataset from the <a class="reference external" href="https://github.com/BiocPy/scrnaseq"><strong>scrnaseq</strong></a> package:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">scrnaseq</span> 
<span class="n">sce</span> <span class="o">=</span> <span class="n">scrnaseq</span><span class="o">.</span><span class="n">fetch_dataset</span><span class="p">(</span><span class="s2">&quot;zeisel-brain-2015&quot;</span><span class="p">,</span> <span class="s2">&quot;2023-12-14&quot;</span><span class="p">,</span> <span class="n">realize_assays</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">sce</span><span class="p">)</span>
<span class="c1">## class: SingleCellExperiment</span>
<span class="c1">## dimensions: (20006, 3005)</span>
<span class="c1">## assays(1): [&#39;counts&#39;]</span>
<span class="c1">## row_data columns(1): [&#39;featureType&#39;]</span>
<span class="c1">## row_names(20006): [&#39;Tspan12&#39;, &#39;Tshz1&#39;, &#39;Fnbp1l&#39;, ..., &#39;mt-Rnr2&#39;, &#39;mt-Rnr1&#39;, &#39;mt-Nd4l&#39;]</span>
<span class="c1">## column_data columns(9): [&#39;tissue&#39;, &#39;group #&#39;, &#39;total mRNA mol&#39;, &#39;well&#39;, &#39;sex&#39;, &#39;age&#39;, &#39;diameter&#39;, &#39;level1class&#39;, &#39;level2class&#39;]</span>
<span class="c1">## column_names(3005): [&#39;1772071015_C02&#39;, &#39;1772071017_G12&#39;, &#39;1772071017_A05&#39;, ..., &#39;1772063068_D01&#39;, &#39;1772066098_A12&#39;, &#39;1772058148_F03&#39;]</span>
<span class="c1">## main_experiment_name: gene</span>
<span class="c1">## reduced_dims(0): []</span>
<span class="c1">## alternative_experiments(2): [&#39;repeat&#39;, &#39;ERCC&#39;]</span>
<span class="c1">## row_pairs(0): []</span>
<span class="c1">## column_pairs(0): []</span>
<span class="c1">## metadata(0):</span>
</pre></div>
</div>
<p>Then we call <strong>scranpy</strong>’s <code class="docutils literal notranslate"><span class="pre">analyze()</span></code> functions, with some additional information about the mitochondrial subset for quality control purposes.
This will perform all of the usual steps for a routine single-cell analysis,
as described in Bioconductor’s <a class="reference external" href="https://bioconductor.org/books/OSCA">Orchestrating single cell analysis</a> book.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">analyze_se</span><span class="p">(</span>
    <span class="n">sce</span><span class="p">,</span>
    <span class="n">rna_qc_subsets</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;mito&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">name</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;mt-&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">sce</span><span class="o">.</span><span class="n">get_row_names</span><span class="p">()]</span>
    <span class="p">}</span>
<span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s2">&quot;x&quot;</span><span class="p">])</span>
<span class="c1">## class: SingleCellExperiment</span>
<span class="c1">## dimensions: (20006, 2874)</span>
<span class="c1">## assays(2): [&#39;counts&#39;, &#39;logcounts&#39;]</span>
<span class="c1">## row_data columns(6): [&#39;featureType&#39;, &#39;mean&#39;, &#39;variance&#39;, &#39;fitted&#39;, &#39;residual&#39;, &#39;hvg&#39;]</span>
<span class="c1">## row_names(20006): [&#39;Tspan12&#39;, &#39;Tshz1&#39;, &#39;Fnbp1l&#39;, ..., &#39;mt-Rnr2&#39;, &#39;mt-Rnr1&#39;, &#39;mt-Nd4l&#39;]</span>
<span class="c1">## column_data columns(15): [&#39;tissue&#39;, &#39;group #&#39;, &#39;total mRNA mol&#39;, ..., &#39;keep&#39;, &#39;size_factor&#39;, &#39;graph_cluster&#39;]</span>
<span class="c1">## column_names(2874): [&#39;1772071015_C02&#39;, &#39;1772071017_G12&#39;, &#39;1772071017_A05&#39;, ..., &#39;1772066097_D04&#39;, &#39;1772063068_D01&#39;, &#39;1772066098_A12&#39;]</span>
<span class="c1">## main_experiment_name: gene</span>
<span class="c1">## reduced_dims(3): [&#39;PCA&#39;, &#39;TSNE&#39;, &#39;UMAP&#39;]</span>
<span class="c1">## alternative_experiments(2): [&#39;repeat&#39;, &#39;ERCC&#39;]</span>
<span class="c1">## row_pairs(0): []</span>
<span class="c1">## column_pairs(0): []</span>
<span class="c1">## metadata(2): qc PCA</span>
</pre></div>
</div>
<p>We can extract useful bits and pieces from the <a class="reference external" href="https://github.com/BiocPy/singlecellexperiment"><code class="docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a> stored as <code class="docutils literal notranslate"><span class="pre">x</span></code>:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s2">&quot;x&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">()[:,[</span><span class="s2">&quot;sum&quot;</span><span class="p">,</span> <span class="s2">&quot;detected&quot;</span><span class="p">,</span> <span class="s2">&quot;subset_proportion_mito&quot;</span><span class="p">,</span> <span class="s2">&quot;size_factor&quot;</span><span class="p">,</span> <span class="s2">&quot;graph_cluster&quot;</span><span class="p">])</span>
<span class="c1">##                               sum          detected subset_proportion_mito         size_factor    graph_cluster</span>
<span class="c1">##                &lt;ndarray[float64]&gt; &lt;ndarray[uint32]&gt;     &lt;ndarray[float64]&gt;  &lt;ndarray[float64]&gt; &lt;ndarray[int64]&gt;</span>
<span class="c1">## 1772071015_C02            22354.0              4871    0.03462467567325758  1.4587034836358181                0</span>
<span class="c1">## 1772071017_G12            22869.0              4712    0.04901832174559447  1.4923096522889652                0</span>
<span class="c1">## 1772071017_A05            32594.0              6055   0.029207829661900962  2.1269115749139242                0</span>
<span class="c1">##                               ...               ...                    ...                 ...              ...</span>
<span class="c1">## 1772066097_D04             2574.0              1441   0.005827505827505828 0.16796558856932076                6</span>
<span class="c1">## 1772063068_D01             4993.0              2001    0.19587422391347886 0.32581669919449047                6</span>
<span class="c1">## 1772066098_A12             3099.0              1510    0.06550500161342368 0.20222430418660645               13</span>

<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s2">&quot;x&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">get_reduced_dimension</span><span class="p">(</span><span class="s2">&quot;TSNE&quot;</span><span class="p">))</span>
<span class="c1">## [[  5.56742365 -28.68868021]</span>
<span class="c1">##  [  5.60398273 -28.02309408]</span>
<span class="c1">##  [  4.77422687 -28.70557818]</span>
<span class="c1">##  ...</span>
<span class="c1">##  [ 18.76434892   8.48223628]</span>
<span class="c1">##  [ 17.69108131   3.82950607]</span>
<span class="c1">##  [ 14.09402365   6.71953971]]</span>

<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s2">&quot;x&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">get_reduced_dimension</span><span class="p">(</span><span class="s2">&quot;UMAP&quot;</span><span class="p">))</span>
<span class="c1">## [[10.36782455 -1.76653302]</span>
<span class="c1">##  [10.24362564 -1.65133715]</span>
<span class="c1">##  [10.46131039 -1.62113261]</span>
<span class="c1">##  ...</span>
<span class="c1">##  [-3.18933988 -6.42624807]</span>
<span class="c1">##  [-2.90072227 -5.33980703]</span>
<span class="c1">##  [-5.86072397 -7.62296963]]</span>
</pre></div>
</div>
<p>We can also inspect the top markers for each cluster in each modality:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s2">&quot;markers&quot;</span><span class="p">][</span><span class="s2">&quot;rna&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">get_names</span><span class="p">())</span>
<span class="c1">## [&#39;0&#39;, &#39;1&#39;, &#39;2&#39;, &#39;3&#39;, &#39;4&#39;, &#39;5&#39;, &#39;6&#39;, &#39;7&#39;, &#39;8&#39;, &#39;9&#39;, &#39;10&#39;, &#39;11&#39;, &#39;12&#39;, &#39;13&#39;]</span>

<span class="nb">print</span><span class="p">(</span><span class="n">scranpy</span><span class="o">.</span><span class="n">preview_markers</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s2">&quot;markers&quot;</span><span class="p">][</span><span class="s2">&quot;rna&quot;</span><span class="p">][</span><span class="s2">&quot;0&quot;</span><span class="p">]))</span>
<span class="c1">## BiocFrame with 10 rows and 3 columns</span>
<span class="c1">##                       mean           detected                lfc</span>
<span class="c1">##         &lt;ndarray[float64]&gt; &lt;ndarray[float64]&gt; &lt;ndarray[float64]&gt;</span>
<span class="c1">##    Gad1  4.722606264371632 0.9967105263157895 4.5144234560400935</span>
<span class="c1">##    Gad2 4.3881067939564335 0.9967105263157895  4.212918994316072</span>
<span class="c1">##   Ndrg4  4.336870587905828 0.9967105263157895 2.5370838460299865</span>
<span class="c1">##   Stmn3  4.676829189457486  0.993421052631579 2.6318375392719537</span>
<span class="c1">##  Vstm2a  2.866641571231044 0.9572368421052632 2.6178131058482004</span>
<span class="c1">##  Nap1l5    4.2569402746771                1.0 3.0585216754783877</span>
<span class="c1">##  Slc6a1 3.6726655518924356 0.9901315789473684 3.0484528790789396</span>
<span class="c1">##   Rab3c 3.8515412644109457 0.9802631578947368 2.9407642089786044</span>
<span class="c1">##  Tspyl4 3.3162021744969783                1.0   2.10987030054009</span>
<span class="c1">## Slc32a1    1.9797938787573 0.8947368421052633 1.9555828312546393</span>
</pre></div>
</div>
<p>Check out the <a class="reference external" href="https://libscran.github.io/scranpy">reference documentation</a> for more details.</p>
</section>
<section id="multiple-batches">
<h2>Multiple batches<a class="headerlink" href="#multiple-batches" title="Permalink to this heading">¶</a></h2>
<p>To demonstrate, let’s grab two pancreas datasets from the <strong>scrnaseq</strong> package.
Each dataset represents a separate batch of cells generated in different studies.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">scrnaseq</span> 
<span class="n">gsce</span> <span class="o">=</span> <span class="n">scrnaseq</span><span class="o">.</span><span class="n">fetch_dataset</span><span class="p">(</span><span class="s2">&quot;grun-pancreas-2016&quot;</span><span class="p">,</span> <span class="s2">&quot;2023-12-14&quot;</span><span class="p">,</span> <span class="n">realize_assays</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">msce</span> <span class="o">=</span> <span class="n">scrnaseq</span><span class="o">.</span><span class="n">fetch_dataset</span><span class="p">(</span><span class="s2">&quot;muraro-pancreas-2016&quot;</span><span class="p">,</span> <span class="s2">&quot;2023-12-19&quot;</span><span class="p">,</span> <span class="n">realize_assays</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<p>They don’t have the same features, so we’ll just take the intersection of their row names before combining them into a single <code class="docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code> object:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">biocutils</span>
<span class="n">common</span> <span class="o">=</span> <span class="n">biocutils</span><span class="o">.</span><span class="n">intersect</span><span class="p">(</span><span class="n">gsce</span><span class="o">.</span><span class="n">get_row_names</span><span class="p">(),</span> <span class="n">msce</span><span class="o">.</span><span class="n">get_row_names</span><span class="p">())</span>
<span class="n">combined</span> <span class="o">=</span> <span class="n">biocutils</span><span class="o">.</span><span class="n">relaxed_combine_columns</span><span class="p">(</span>
    <span class="n">gsce</span><span class="p">[</span><span class="n">biocutils</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="n">common</span><span class="p">,</span> <span class="n">gsce</span><span class="o">.</span><span class="n">get_row_names</span><span class="p">()),</span> <span class="p">:],</span>
    <span class="n">msce</span><span class="p">[</span><span class="n">biocutils</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="n">common</span><span class="p">,</span> <span class="n">msce</span><span class="o">.</span><span class="n">get_row_names</span><span class="p">()),</span> <span class="p">:]</span>
<span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">combined</span><span class="p">)</span>
<span class="c1">## class: SingleCellExperiment</span>
<span class="c1">## dimensions: (18499, 4800)</span>
<span class="c1">## assays(1): [&#39;counts&#39;]</span>
<span class="c1">## row_data columns(2): [&#39;symbol&#39;, &#39;chr&#39;]</span>
<span class="c1">## row_names(18499): [&#39;A1BG-AS1__chr19&#39;, &#39;A1BG__chr19&#39;, &#39;A1CF__chr10&#39;, ..., &#39;ZYX__chr7&#39;, &#39;ZZEF1__chr17&#39;, &#39;ZZZ3__chr1&#39;]</span>
<span class="c1">## column_data columns(4): [&#39;donor&#39;, &#39;sample&#39;, &#39;label&#39;, &#39;plate&#39;]</span>
<span class="c1">## column_names(4800): [&#39;D2ex_1&#39;, &#39;D2ex_2&#39;, &#39;D2ex_3&#39;, ..., &#39;D30-8_94&#39;, &#39;D30-8_95&#39;, &#39;D30-8_96&#39;]</span>
<span class="c1">## main_experiment_name: endogenous</span>
<span class="c1">## reduced_dims(0): []</span>
<span class="c1">## alternative_experiments(0): []</span>
<span class="c1">## row_pairs(0): []</span>
<span class="c1">## column_pairs(0): []</span>
<span class="c1">## metadata(0):</span>
</pre></div>
</div>
<p>We can now perform a batch-aware analysis, where the blocking factor is also used in relevant functions to avoid problems with batch effects.
This yields mostly the same set of results as before, but with an extra MNN-corrected embedding for clustering, visualization, etc.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="n">block</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;grun&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="n">gsce</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="s2">&quot;muraro&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="n">msce</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">analyze</span><span class="p">(</span><span class="n">combined</span><span class="p">,</span> <span class="n">block</span><span class="o">=</span><span class="n">block</span><span class="p">)</span> <span class="c1"># no mitochondrial genes in this case...</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s2">&quot;x&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">get_reduced_dimension_names</span><span class="p">())</span>
<span class="c1">## [&#39;PCA&#39;, &#39;MNN&#39;, &#39;TSNE&#39;, &#39;UMAP&#39;]</span>
</pre></div>
</div>
<p>The blocking factor is also used during marker detection to ensure that any batch effects do not interfere with marker scoring.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s2">&quot;markers&quot;</span><span class="p">][</span><span class="s2">&quot;rna&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">get_names</span><span class="p">())</span>
<span class="c1">## [&#39;0&#39;, &#39;1&#39;, &#39;2&#39;, &#39;3&#39;, &#39;4&#39;, &#39;5&#39;, &#39;6&#39;, &#39;7&#39;, &#39;8&#39;, &#39;9&#39;, &#39;10&#39;, &#39;11&#39;, &#39;12&#39;, &#39;13&#39;, &#39;14&#39;]</span>

<span class="nb">print</span><span class="p">(</span><span class="n">scranpy</span><span class="o">.</span><span class="n">preview_markers</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s2">&quot;markers&quot;</span><span class="p">][</span><span class="s2">&quot;rna&quot;</span><span class="p">][</span><span class="s2">&quot;0&quot;</span><span class="p">]))</span>
<span class="c1">## BiocFrame with 10 rows and 3 columns</span>
<span class="c1">##                              mean           detected                lfc</span>
<span class="c1">##                &lt;ndarray[float64]&gt; &lt;ndarray[float64]&gt; &lt;ndarray[float64]&gt;</span>
<span class="c1">##    PRSS1__chr7 7.4960140622383635                1.0  6.112707795838959</span>
<span class="c1">## PLA2G1B__chr12 5.3256781931032435 0.9767441860465116 4.9768663423161525</span>
<span class="c1">##   SPINK1__chr5  7.226025132762255                1.0  5.484853630696096</span>
<span class="c1">##  PRSS3P2__chr7  6.414465893460902                1.0  5.474642067942027</span>
<span class="c1">##   CTRB1__chr16  6.242340887932875                1.0  5.439850813378681</span>
<span class="c1">##   CTRB2__chr16  6.211977585350613 0.9767441860465116  5.390247433870743</span>
<span class="c1">##   CELA3A__chr1  5.749981006377225 0.9922480620155039  5.379827932369857</span>
<span class="c1">##     CPA1__chr7  5.956052489801642 0.9922480620155039  5.398848064205473</span>
<span class="c1">##    REG1A__chr2  8.505688048945858                1.0   5.69363720328606</span>
<span class="c1">##     CPA2__chr7  4.394240756079104 0.9224806201550388  4.167356591402403</span>
</pre></div>
</div>
</section>
<section id="multiple-modalities">
<h2>Multiple modalities<a class="headerlink" href="#multiple-modalities" title="Permalink to this heading">¶</a></h2>
<p>Let’s grab a 10X Genomics immune profiling dataset (see <a class="reference external" href="https://github.com/kanaverse/random-test-files/releases/download/10x-immune-v1.0.0/immune_3.0.0-tenx.h5">here</a>),
which contains count data for the entire transcriptome and targeted proteins:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">singlecellexperiment</span>
<span class="nb">input</span> <span class="o">=</span> <span class="n">singlecellexperiment</span><span class="o">.</span><span class="n">read_tenx_h5</span><span class="p">(</span><span class="s2">&quot;immune_3.0.0-tenx.h5&quot;</span><span class="p">,</span> <span class="n">realize_assays</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">input</span><span class="o">.</span><span class="n">set_row_names</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">get_row_data</span><span class="p">()</span><span class="o">.</span><span class="n">get_column</span><span class="p">(</span><span class="s2">&quot;id&quot;</span><span class="p">),</span> <span class="n">in_place</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<p>We split our <code class="docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code> into separate objects for genes and ADTs.
We use the gene dataset as the main experiment and store the ADT object as an alternative experiment.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">feattypes</span> <span class="o">=</span> <span class="nb">input</span><span class="o">.</span><span class="n">get_row_data</span><span class="p">()</span><span class="o">.</span><span class="n">get_column</span><span class="p">(</span><span class="s2">&quot;feature_type&quot;</span><span class="p">)</span>
<span class="n">gene_data</span> <span class="o">=</span> <span class="nb">input</span><span class="p">[[</span><span class="n">x</span> <span class="o">==</span> <span class="s2">&quot;Gene Expression&quot;</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">feattypes</span><span class="p">],:]</span>
<span class="n">adt_data</span> <span class="o">=</span> <span class="nb">input</span><span class="p">[[</span><span class="n">x</span> <span class="o">==</span> <span class="s2">&quot;Antibody Capture&quot;</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">feattypes</span><span class="p">],:]</span>

<span class="n">sce</span> <span class="o">=</span> <span class="n">gene_data</span>
<span class="n">sce</span><span class="o">.</span><span class="n">set_alternative_experiment</span><span class="p">(</span><span class="s2">&quot;ADT&quot;</span><span class="p">,</span> <span class="n">adt_data</span><span class="p">,</span> <span class="n">in_place</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">sce</span><span class="p">)</span>
<span class="c1">## class: SingleCellExperiment</span>
<span class="c1">## dimensions: (33555, 8258)</span>
<span class="c1">## assays(1): [&#39;counts&#39;]</span>
<span class="c1">## row_data columns(7): [&#39;feature_type&#39;, &#39;genome&#39;, &#39;id&#39;, &#39;name&#39;, &#39;pattern&#39;, &#39;read&#39;, &#39;sequence&#39;]</span>
<span class="c1">## row_names(33555): [&#39;ENSG00000243485&#39;, &#39;ENSG00000237613&#39;, &#39;ENSG00000186092&#39;, ..., &#39;IgG2b&#39;, &#39;CD127&#39;, &#39;CD15&#39;]</span>
<span class="c1">## column_data columns(1): [&#39;barcodes&#39;]</span>
<span class="c1">## column_names(0):</span>
<span class="c1">## main_experiment_name:</span>
<span class="c1">## reduced_dims(0): []</span>
<span class="c1">## alternative_experiments(0): []</span>
<span class="c1">## row_pairs(0): []</span>
<span class="c1">## column_pairs(0): []</span>
<span class="c1">## metadata(0):</span>

<span class="nb">print</span><span class="p">(</span><span class="n">sce</span><span class="o">.</span><span class="n">get_alternative_experiment</span><span class="p">(</span><span class="s2">&quot;ADT&quot;</span><span class="p">))</span>
<span class="c1">## class: SingleCellExperiment</span>
<span class="c1">## dimensions: (17, 8258)</span>
<span class="c1">## assays(1): [&#39;counts&#39;]</span>
<span class="c1">## row_data columns(7): [&#39;feature_type&#39;, &#39;genome&#39;, &#39;id&#39;, &#39;name&#39;, &#39;pattern&#39;, &#39;read&#39;, &#39;sequence&#39;]</span>
<span class="c1">## row_names(17): [&#39;CD3&#39;, &#39;CD19&#39;, &#39;CD45RA&#39;, ..., &#39;IgG2b&#39;, &#39;CD127&#39;, &#39;CD15&#39;]</span>
<span class="c1">## column_data columns(1): [&#39;barcodes&#39;]</span>
<span class="c1">## column_names(0):</span>
<span class="c1">## main_experiment_name:</span>
<span class="c1">## reduced_dims(0): []</span>
<span class="c1">## alternative_experiments(0): []</span>
<span class="c1">## row_pairs(0): []</span>
<span class="c1">## column_pairs(0): []</span>
<span class="c1">## metadata(0):</span>
</pre></div>
</div>
<p>And now we can run the analysis, with some additional specification of the IgG subsets for ADT-related quality control.
ADT-specific results are stored in the alternative experiment of the output <code class="docutils literal notranslate"><span class="pre">x</span></code>:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">analyze_se</span><span class="p">(</span>
    <span class="n">sce</span><span class="p">,</span>
    <span class="n">adt_altexp</span><span class="o">=</span><span class="s2">&quot;ADT&quot;</span><span class="p">,</span>
    <span class="n">rna_qc_subsets</span> <span class="o">=</span> <span class="p">{</span> 
        <span class="s2">&quot;mito&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">n</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;MT-&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">gene_data</span><span class="o">.</span><span class="n">get_row_data</span><span class="p">()</span><span class="o">.</span><span class="n">get_column</span><span class="p">(</span><span class="s2">&quot;name&quot;</span><span class="p">)]</span>
    <span class="p">},</span>
    <span class="n">adt_qc_subsets</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;igg&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">n</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;IgG&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">adt_data</span><span class="o">.</span><span class="n">get_row_data</span><span class="p">()</span><span class="o">.</span><span class="n">get_column</span><span class="p">(</span><span class="s2">&quot;name&quot;</span><span class="p">)]</span>
    <span class="p">}</span>
<span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s2">&quot;x&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">get_alternative_experiment</span><span class="p">(</span><span class="s2">&quot;ADT&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">()[:,[</span><span class="s2">&quot;sum&quot;</span><span class="p">,</span> <span class="s2">&quot;detected&quot;</span><span class="p">,</span> <span class="s2">&quot;subset_sum_igg&quot;</span><span class="p">,</span> <span class="s2">&quot;size_factor&quot;</span><span class="p">]])</span>
<span class="c1">## BiocFrame with 6779 rows and 4 columns</span>
<span class="c1">##                       sum          detected     subset_sum_igg        size_factor</span>
<span class="c1">##        &lt;ndarray[float64]&gt; &lt;ndarray[uint32]&gt; &lt;ndarray[float64]&gt; &lt;ndarray[float64]&gt;</span>
<span class="c1">##    [0]             2410.0                17               21.0 0.7935940817266915</span>
<span class="c1">##    [1]             2637.0                17               30.0 0.7941033166783458</span>
<span class="c1">##    [2]             5551.0                17               29.0  0.895364130394313</span>
<span class="c1">##                       ...               ...                ...                ...</span>
<span class="c1">## [6776]             5079.0                17               11.0 0.7920783948957917</span>
<span class="c1">## [6777]             1757.0                17               12.0 0.6649272283074951</span>
<span class="c1">## [6778]             2312.0                17               33.0 0.7684763715276961</span>

<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s2">&quot;x&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">get_alternative_experiment</span><span class="p">(</span><span class="s2">&quot;ADT&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">get_reduced_dimension_names</span><span class="p">())</span>
<span class="c1">## [&#39;PCA&#39;]</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">analyze_se()</span></code> combines the RNA and ADT data into a single embedding for downstream steps like visualization and clustering.
This ensures that those steps will use information from both modalities.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s2">&quot;x&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">get_reduced_dimension_names</span><span class="p">())</span>
<span class="c1">## [&#39;PCA&#39;, &#39;combined&#39;, &#39;TSNE&#39;, &#39;UMAP&#39;]</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">biocutils</span>
<span class="nb">print</span><span class="p">(</span><span class="n">biocutils</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s2">&quot;x&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">()[</span><span class="s2">&quot;graph_cluster&quot;</span><span class="p">]))</span>
<span class="c1">## [&#39;0&#39;=401, &#39;1&#39;=801, &#39;2&#39;=282, &#39;3&#39;=1030, &#39;4&#39;=1142, &#39;5&#39;=279, &#39;6&#39;=922, &#39;7&#39;=198, &#39;8&#39;=375, &#39;9&#39;=212, &#39;10&#39;=884, &#39;11&#39;=47, &#39;12&#39;=13, &#39;13&#39;=48, &#39;14&#39;=58, &#39;15&#39;=87]</span>
</pre></div>
</div>
<p>Similarly, <code class="docutils literal notranslate"><span class="pre">analyze_se()</span></code> will compute marker statistics for the ADT data:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">scranpy</span><span class="o">.</span><span class="n">preview_markers</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s2">&quot;markers&quot;</span><span class="p">][</span><span class="s2">&quot;adt&quot;</span><span class="p">][</span><span class="s2">&quot;0&quot;</span><span class="p">]))</span>
<span class="c1">## BiocFrame with 10 rows and 3 columns</span>
<span class="c1">##                      mean           detected                  lfc</span>
<span class="c1">##        &lt;ndarray[float64]&gt; &lt;ndarray[float64]&gt;   &lt;ndarray[float64]&gt;</span>
<span class="c1">##    CD3 10.710798038087853                1.0    3.073794979159913</span>
<span class="c1">##  CD127  7.202083836362728                1.0   2.1670932585149587</span>
<span class="c1">## CD45RO  7.062122120526255                1.0   1.6391671901349703</span>
<span class="c1">##   CD8a  7.530206276892356                1.0   1.8636563581508114</span>
<span class="c1">##   CD56  5.110427162087626                1.0   0.6727102618980374</span>
<span class="c1">##   PD-1  5.739783018333582                1.0  0.19200190878211132</span>
<span class="c1">##   IgG1  4.330276551600134                1.0 -0.02840790745859983</span>
<span class="c1">##   CD25  5.088783799767125                1.0  -0.1269303811218677</span>
<span class="c1">##  TIGIT 3.6629345003643983 0.9975062344139651 -0.28573778753415674</span>
<span class="c1">##  IgG2a  3.143818721927284 0.9900249376558603  -0.3191747955354897</span>
</pre></div>
</div>
</section>
<section id="customizing-the-analysis">
<h2>Customizing the analysis<a class="headerlink" href="#customizing-the-analysis" title="Permalink to this heading">¶</a></h2>
<p>Most parameters can be changed by modifying the relevant arguments in <code class="docutils literal notranslate"><span class="pre">analyze_se()</span></code>.
For example:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">scrnaseq</span> 
<span class="n">sce</span> <span class="o">=</span> <span class="n">scrnaseq</span><span class="o">.</span><span class="n">fetch_dataset</span><span class="p">(</span><span class="s2">&quot;zeisel-brain-2015&quot;</span><span class="p">,</span> <span class="s2">&quot;2023-12-14&quot;</span><span class="p">,</span> <span class="n">realize_assays</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">is_mito</span> <span class="o">=</span> <span class="p">[</span><span class="n">name</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;mt-&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">sce</span><span class="o">.</span><span class="n">get_row_names</span><span class="p">()]</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">analyze</span><span class="p">(</span>
    <span class="n">sce</span><span class="p">,</span>
    <span class="n">rna_qc_subsets</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;mito&quot;</span><span class="p">:</span> <span class="n">is_mito</span>
    <span class="p">},</span>
    <span class="n">more_build_graph_args</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;num_neighbors&quot;</span><span class="p">:</span> <span class="mi">10</span>
    <span class="p">},</span>
    <span class="n">more_cluster_graph_args</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;multilevel_resolution&quot;</span><span class="p">:</span> <span class="mi">2</span>
    <span class="p">},</span>
    <span class="n">more_rna_pca_args</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;number&quot;</span><span class="p">:</span> <span class="mi">15</span>
    <span class="p">},</span>
    <span class="n">more_tsne_args</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;perplexity&quot;</span><span class="p">:</span> <span class="mi">25</span>
    <span class="p">},</span>
    <span class="n">more_umap_args</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;min_dist&quot;</span><span class="p">:</span> <span class="mf">0.05</span>
    <span class="p">}</span>
<span class="p">)</span>
</pre></div>
</div>
<p>For finer control, users can call each step individually via lower-level functions.
A typical RNA analysis might be implemented as:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">res</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">quick_rna_qc_se</span><span class="p">(</span><span class="n">sce</span><span class="p">,</span> <span class="n">subsets</span><span class="o">=</span><span class="p">{</span> <span class="s2">&quot;mito&quot;</span><span class="p">:</span> <span class="n">is_mito</span> <span class="p">})</span>
<span class="n">res</span> <span class="o">=</span> <span class="n">res</span><span class="p">[:,</span><span class="n">res</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">()[</span><span class="s2">&quot;keep&quot;</span><span class="p">]]</span>
<span class="n">res</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">normalize_rna_counts_se</span><span class="p">(</span><span class="n">res</span><span class="p">,</span> <span class="n">size_factors</span><span class="o">=</span><span class="n">res</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">()[</span><span class="s2">&quot;sum&quot;</span><span class="p">])</span>
<span class="n">res</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">choose_rna_hvgs_se</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>
<span class="n">res</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">run_pca_se</span><span class="p">(</span><span class="n">res</span><span class="p">,</span> <span class="n">features</span><span class="o">=</span><span class="n">res</span><span class="o">.</span><span class="n">get_row_data</span><span class="p">()[</span><span class="s2">&quot;hvg&quot;</span><span class="p">])</span>
<span class="n">res</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">run_all_neighbor_steps_se</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>
<span class="n">markers</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">score_markers_se</span><span class="p">(</span><span class="n">res</span><span class="p">,</span> <span class="n">groups</span><span class="o">=</span><span class="n">res</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">()[</span><span class="s2">&quot;clusters&quot;</span><span class="p">])</span>
</pre></div>
</div>
<p>Check out <a class="reference internal" href="#src/scranpy/_se_analyze.py"><span class="xref myst"><code class="docutils literal notranslate"><span class="pre">analyze_se()</span></code> source code</span></a> for more details.</p>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="contributing.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Contributing</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="index.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Home</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2023, Jayaram Kancherla
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">scran, in Python</a><ul>
<li><a class="reference internal" href="#overview">Overview</a></li>
<li><a class="reference internal" href="#quick-start">Quick start</a></li>
<li><a class="reference internal" href="#multiple-batches">Multiple batches</a></li>
<li><a class="reference internal" href="#multiple-modalities">Multiple modalities</a></li>
<li><a class="reference internal" href="#customizing-the-analysis">Customizing the analysis</a></li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script data-url_root="./" id="documentation_options" src="_static/documentation_options.js?v=c02300dd"></script>
    <script src="_static/doctools.js?v=888ff710"></script>
    <script src="_static/sphinx_highlight.js?v=4825356b"></script>
    <script src="_static/scripts/furo.js?v=46bd48cc"></script>
    </body>
</html>