<!doctype html>
<html class="no-js" lang="en" data-content_root="">
  <head><meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <meta name="color-scheme" content="light dark"><link rel="index" title="Index" href="../../genindex.html"><link rel="search" title="Search" href="../../search.html">

    <!-- Generated with Sphinx 7.1.2 and Furo 2025.12.19 -->
        <title>scranpy._se_analyze - scranpy 0.3.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/furo.css?v=7bdb33bb" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/furo-extensions.css?v=8dab3a3b" />
    
    


<style>
  body {
    --color-code-background: #eeffcc;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle site navigation sidebar">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc" aria-label="Toggle table of contents sidebar">
<label class="overlay sidebar-overlay" for="__navigation"></label>
<label class="overlay toc-overlay" for="__toc"></label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <span class="icon"><svg><use href="#svg-menu"></use></svg></span>
      </label>
    </div>
    <div class="header-center">
      <a href="../../index.html"><div class="brand">scranpy 0.3.0 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle" aria-label="Toggle Light / Dark / Auto color theme">
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon no-toc" for="__toc">
        <span class="icon"><svg><use href="#svg-toc"></use></svg></span>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../../index.html">
  
  <span class="sidebar-brand-text">scranpy 0.3.0 documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="../../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul>
<li class="toctree-l1"><a class="reference internal" href="../../readme.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../contributing.html">Contributions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../authors.html">Authors</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../changelog.html">Changelog</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../api/modules.html">Module Reference</a><input aria-label="Toggle navigation of Module Reference" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><span class="icon"><svg><use href="#svg-arrow-right"></use></svg></span></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../api/scranpy.html">scranpy package</a></li>
</ul>
</li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle" aria-label="Toggle Light / Dark / Auto color theme">
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon no-toc" for="__toc">
            <span class="icon"><svg><use href="#svg-toc"></use></svg></span>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <h1>Source code for scranpy._se_analyze</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span><span class="w"> </span><span class="nn">typing</span><span class="w"> </span><span class="kn">import</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">Union</span><span class="p">,</span> <span class="n">Literal</span><span class="p">,</span> <span class="n">Sequence</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">delayedarray</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">summarizedexperiment</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">singlecellexperiment</span> 
<span class="kn">import</span><span class="w"> </span><span class="nn">biocutils</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">knncolle</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">._se_quick_rna_qc</span><span class="w"> </span><span class="kn">import</span> <span class="n">quick_rna_qc_se</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">._se_quick_adt_qc</span><span class="w"> </span><span class="kn">import</span> <span class="n">quick_adt_qc_se</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">._se_quick_crispr_qc</span><span class="w"> </span><span class="kn">import</span> <span class="n">quick_crispr_qc_se</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">._se_normalize_rna_counts</span><span class="w"> </span><span class="kn">import</span> <span class="n">normalize_rna_counts_se</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">._se_normalize_adt_counts</span><span class="w"> </span><span class="kn">import</span> <span class="n">normalize_adt_counts_se</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">._se_normalize_crispr_counts</span><span class="w"> </span><span class="kn">import</span> <span class="n">normalize_crispr_counts_se</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">._se_choose_rna_hvgs</span><span class="w"> </span><span class="kn">import</span> <span class="n">choose_rna_hvgs_se</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">._se_run_pca</span><span class="w"> </span><span class="kn">import</span> <span class="n">run_pca_se</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">._se_scale_by_neighbors</span><span class="w"> </span><span class="kn">import</span> <span class="n">scale_by_neighbors_se</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">._se_correct_mnn</span><span class="w"> </span><span class="kn">import</span> <span class="n">correct_mnn_se</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">._se_run_all_neighbor_steps</span><span class="w"> </span><span class="kn">import</span> <span class="n">run_all_neighbor_steps_se</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">._se_cluster_kmeans</span><span class="w"> </span><span class="kn">import</span> <span class="n">cluster_kmeans_se</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">._se_score_markers</span><span class="w"> </span><span class="kn">import</span> <span class="n">score_markers_se</span>


<div class="viewcode-block" id="analyze_se"><a class="viewcode-back" href="../../api/scranpy.html#scranpy.analyze_se">[docs]</a><span class="k">def</span><span class="w"> </span><span class="nf">analyze_se</span><span class="p">(</span> 
    <span class="n">x</span><span class="p">:</span> <span class="n">summarizedexperiment</span><span class="o">.</span><span class="n">SummarizedExperiment</span><span class="p">,</span>
    <span class="n">rna_altexp</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Union</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">str</span><span class="p">,</span> <span class="nb">bool</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">adt_altexp</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Union</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">str</span><span class="p">,</span> <span class="nb">bool</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">crispr_altexp</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Union</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">str</span><span class="p">,</span> <span class="nb">bool</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">rna_assay_type</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;counts&quot;</span><span class="p">,</span>
    <span class="n">adt_assay_type</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;counts&quot;</span><span class="p">,</span>
    <span class="n">crispr_assay_type</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;counts&quot;</span><span class="p">,</span>
    <span class="n">block</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Sequence</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">block_name</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;block&quot;</span><span class="p">,</span>
    <span class="n">rna_qc_subsets</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">dict</span><span class="p">,</span> <span class="n">Sequence</span><span class="p">,</span> <span class="n">biocutils</span><span class="o">.</span><span class="n">NamedList</span><span class="p">]</span> <span class="o">=</span> <span class="p">[],</span>
    <span class="n">rna_qc_output_prefix</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">more_rna_qc_args</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="p">{},</span>
    <span class="n">adt_qc_subsets</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">dict</span><span class="p">,</span> <span class="n">Sequence</span><span class="p">,</span> <span class="n">biocutils</span><span class="o">.</span><span class="n">NamedList</span><span class="p">]</span> <span class="o">=</span> <span class="p">[],</span>
    <span class="n">adt_qc_output_prefix</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">more_adt_qc_args</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="p">{},</span>
    <span class="n">crispr_qc_output_prefix</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">more_crispr_qc_args</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="p">{},</span>
    <span class="n">filter_cells</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">rna_norm_output_name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;logcounts&quot;</span><span class="p">,</span>
    <span class="n">more_rna_norm_args</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="p">{},</span>
    <span class="n">adt_norm_output_name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;logcounts&quot;</span><span class="p">,</span>
    <span class="n">more_adt_norm_args</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="p">{},</span>
    <span class="n">crispr_norm_output_name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;logcounts&quot;</span><span class="p">,</span>
    <span class="n">more_crispr_norm_args</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="p">{},</span>
    <span class="n">rna_hvg_output_prefix</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">more_rna_hvg_args</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="p">{},</span>
    <span class="n">rna_pca_output_name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;PCA&quot;</span><span class="p">,</span>
    <span class="n">more_rna_pca_args</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="p">{},</span>
    <span class="n">adt_pca_output_name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;PCA&quot;</span><span class="p">,</span>
    <span class="n">more_adt_pca_args</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="p">{},</span>
    <span class="n">use_rna_pcs</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">use_adt_pcs</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">scale_output_name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;combined&quot;</span><span class="p">,</span>
    <span class="n">more_scale_args</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="p">{},</span>
    <span class="n">mnn_output_name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;MNN&quot;</span><span class="p">,</span>
    <span class="n">more_mnn_args</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="p">{},</span>
    <span class="n">more_umap_args</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="p">{},</span>
    <span class="n">more_tsne_args</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="p">{},</span>
    <span class="n">cluster_graph_output_name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;graph_cluster&quot;</span><span class="p">,</span>
    <span class="n">more_build_graph_args</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="p">{},</span>
    <span class="n">more_cluster_graph_args</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="p">{},</span>
    <span class="n">more_neighbor_args</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="p">{},</span>
    <span class="n">kmeans_clusters</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">kmeans_clusters_output_name</span> <span class="o">=</span> <span class="s2">&quot;kmeans_cluster&quot;</span><span class="p">,</span>
    <span class="n">more_kmeans_args</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="p">{},</span>
    <span class="n">clusters_for_markers</span><span class="p">:</span> <span class="n">Sequence</span><span class="p">[</span><span class="n">Literal</span><span class="p">[</span><span class="s2">&quot;graph&quot;</span><span class="p">,</span> <span class="s2">&quot;kmeans&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;graph&quot;</span><span class="p">,</span> <span class="s2">&quot;kmeans&quot;</span><span class="p">],</span>
    <span class="n">more_rna_marker_args</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="p">{},</span>
    <span class="n">more_adt_marker_args</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="p">{},</span>
    <span class="n">more_crispr_marker_args</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="p">{},</span>
    <span class="n">nn_parameters</span> <span class="o">=</span> <span class="n">knncolle</span><span class="o">.</span><span class="n">AnnoyParameters</span><span class="p">(),</span>
    <span class="n">num_threads</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">3</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">biocutils</span><span class="o">.</span><span class="n">NamedList</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Execute a simple single-cell analysis pipeline, starting from a count matrix and ending with clusters, visualizations and markers.</span>
<span class="sd">    This is equivalent to:</span>

<span class="sd">    - Running :py:func:`~scranpy.quick_rna_qc_se`,</span>
<span class="sd">      :py:func:`~scranpy.quick_adt_qc_se`</span>
<span class="sd">      and/or :py:func:`~scranpy.quick_crispr_qc_se`,</span>
<span class="sd">      for quality control.</span>
<span class="sd">    - Subsetting ``x`` to only retain the high-quality cells in all modalities. </span>
<span class="sd">    - Running :py:func:`~scranpy.normalize_rna_counts_se`,</span>
<span class="sd">      :py:func:`~scranpy.normalize_adt_counts_se`</span>
<span class="sd">      and/or :py:func:`~scranpy.normalize_crispr_counts_se`,</span>
<span class="sd">      for normalization.</span>
<span class="sd">    - Running :py:func:`~scranpy.choose_rna_hvgs_se` to identify highly variable genes.</span>
<span class="sd">    - Running :py:func:`~scranpy.run_pca_se` on the RNA and/or ADT data.</span>
<span class="sd">    - Running :py:func:`~scranpy.scale_by_neighbors_se` if multiple modalities are present.</span>
<span class="sd">    - Running :py:func:`~scranpy.correct_mnn_se` if multiple batches are present. </span>
<span class="sd">    - Running :py:func:`~scranpy.run_all_neighbor_steps_se` to obtain t-SNE and UMAP coordinates, and to perform graph-based clustering.</span>
<span class="sd">    - Running :py:func:`~scranpy.cluster_kmeans_se` to perform k-means clustering, if requested. </span>
<span class="sd">    - Running :py:func:`~scranpy.score_markers_se` to compute markers for each modality based on one of the clusterings. </span>

<span class="sd">    Args:</span>
<span class="sd">        x:</span>
<span class="sd">            A :py:class:`~summarizedexperiment.SummarizedExperiment` object or one of its subclasses.</span>
<span class="sd">            Rows correspond to genomic features (genes, ADTs, or CRISPR guides) and columns correspond to cells.</span>

<span class="sd">        rna_altexp:</span>
<span class="sd">            Name or index of the alternative experiment of ``x`` containing the RNA data.</span>
<span class="sd">            Alternatively ``False``, in which case the main experiment is assumed to contain the RNA data.</span>
<span class="sd">            Alternatively ``None``, in which case it is assumed that no RNA data is available.</span>

<span class="sd">        adt_altexp:</span>
<span class="sd">            Name or index of the alternative experiment of ``x`` containing the ADT data.</span>
<span class="sd">            Alternatively ``False``, in which case the main experiment is assumed to contain the ADT data.</span>
<span class="sd">            Alternatively ``None``, in which case it is assumed that no ADT data is available.</span>

<span class="sd">        crispr_altexp:</span>
<span class="sd">            Name or index of the alternative experiment of ``x`` containing the CRISPR data.</span>
<span class="sd">            Alternatively ``False``, in which case the main experiment is assumed to contain the CRISPR data.</span>
<span class="sd">            Alternatively ``None``, in which case it is assumed that no CRISPR data is available.</span>

<span class="sd">        rna_assay_type:</span>
<span class="sd">            Name or index of the assay containing the RNA count data.</span>
<span class="sd">            Only used if ``rna_altexp`` is not ``None``.</span>

<span class="sd">        adt_assay_type:</span>
<span class="sd">            Name or index of the assay containing the ADT count data.</span>
<span class="sd">            Only used if ``rna_altexp`` is not ``None``.</span>

<span class="sd">        crispr_assay_type:</span>
<span class="sd">            Name or index of the assay containing the CRISPR count data.</span>
<span class="sd">            Only used if ``rna_altexp`` is not ``None``.</span>

<span class="sd">        block:</span>
<span class="sd">            Blocking factor, as a sequence containing the block of origin (e.g., batch, sample) for each cell in ``x``.</span>
<span class="sd">            Alternatively ``None``, if all cells are from the same block.</span>

<span class="sd">        block_name:</span>
<span class="sd">            Name of the column in which to store the blocking factor in the column data of the output ``SummarizedExperiment``.</span>
<span class="sd">            Only used if ``block`` is not ``None``.</span>
<span class="sd">            If ``None``, the blocking factor is not stored in the output.</span>

<span class="sd">        rna_qc_subsets:</span>
<span class="sd">            Passed to :py:func:`~scranpy.quick_rna_qc_se` as the ``subsets`` argument.</span>
<span class="sd">            Only used if ``rna_altexp`` is not ``None``.</span>

<span class="sd">        rna_qc_output_prefix:</span>
<span class="sd">            Passed to :py:func:`~scranpy.quick_rna_qc_se` as the ``output_prefix`` argument.</span>
<span class="sd">            Only used if ``rna_altexp`` is not ``None``.</span>

<span class="sd">        more_rna_qc_args:</span>
<span class="sd">            Additional arguments to pass to :py:func:`~scranpy.quick_rna_qc_se`.</span>
<span class="sd">            Only used if ``rna_altexp`` is not ``None``.</span>

<span class="sd">        adt_qc_subsets:</span>
<span class="sd">            Passed to :py:func:`~scranpy.quick_adt_qc_se` as the ``subsets`` argument.</span>
<span class="sd">            Only used if ``adt_altexp`` is not ``None``.</span>

<span class="sd">        adt_qc_output_prefix:</span>
<span class="sd">            Passed to :py:func:`~scranpy.quick_adt_qc_se` as the ``output_prefix`` argument.</span>
<span class="sd">            Only used if ``adt_altexp`` is not ``None``.</span>

<span class="sd">        more_adt_qc_args:</span>
<span class="sd">            Additional arguments to pass to :py:func:`~scranpy.quick_adt_qc_se`.</span>
<span class="sd">            Only used if ``adt_altexp`` is not ``None``.</span>

<span class="sd">        crispr_qc_output_prefix:</span>
<span class="sd">            Passed to :py:func:`~scranpy.quick_crispr_qc_se` as the ``output_prefix`` argument.</span>
<span class="sd">            Only used if ``crispr_altexp`` is not ``None``.</span>

<span class="sd">        more_crispr_qc_args:</span>
<span class="sd">            Additional arguments to pass to :py:func:`~scranpy.quick_crispr_qc_se`.</span>
<span class="sd">            Only used if ``crispr_altexp`` is not ``None``.</span>

<span class="sd">        filter_cells:</span>
<span class="sd">            Whether to filter ``x`` to only retain high-quality cells in all modalities.</span>
<span class="sd">            If ``False`` QC metrics and thresholds are still computed but are not used to filter the count matrices.</span>

<span class="sd">        rna_norm_output_name:</span>
<span class="sd">            Passed to :py:func:`~scranpy.normalize_rna_counts_se` as the ``output_name`` argument.</span>
<span class="sd">            Only used if ``rna_altexp`` is not ``None``.</span>

<span class="sd">        more_rna_norm_args:</span>
<span class="sd">            Additional arguments to pass to :py:func:`~scranpy.normalize_rna_counts_se`.</span>
<span class="sd">            Only used if ``rna_altexp`` is not ``None``.</span>

<span class="sd">        adt_norm_output_name:</span>
<span class="sd">            Passed to :py:func:`~scranpy.normalize_adt_counts_se` as the ``output_name`` argument.</span>
<span class="sd">            Only used if ``adt_altexp`` is not ``None``.</span>

<span class="sd">        more_adt_norm_args:</span>
<span class="sd">            Additional arguments to pass to :py:func:`~scranpy.normalize_adt_counts_se`.</span>
<span class="sd">            Only used if ``adt_altexp`` is not ``None``.</span>

<span class="sd">        crispr_norm_output_name:</span>
<span class="sd">            Passed to :py:func:`~scranpy.normalize_crispr_counts_se` as the ``output_name`` argument.</span>
<span class="sd">            Only used if ``crispr_altexp`` is not ``None``.</span>

<span class="sd">        more_crispr_norm_args:</span>
<span class="sd">            Additional arguments to pass to :py:func:`~scranpy.normalize_crispr_counts_se`.</span>
<span class="sd">            Only used if ``crispr_altexp`` is not ``None``.</span>

<span class="sd">        rna_hvg_output_prefix:</span>
<span class="sd">            Passed to :py:func:`~scranpy.choose_rna_hvgs_se` as the ``output_prefix`` argument.</span>
<span class="sd">            Only used if ``rna_altexp`` is not ``None``.</span>

<span class="sd">        more_rna_hvg_args:</span>
<span class="sd">            Additional arguments to pass to :py:func:`~scranpy.choose_rna_hvgs_se`.</span>
<span class="sd">            Only used if ``rna_altexp`` is not ``None``.</span>

<span class="sd">        rna_pca_output_name:</span>
<span class="sd">            Passed to :py:func:`~scranpy.run_pca_se` as the ``output_name`` argument.</span>
<span class="sd">            Only used if ``rna_altexp`` is not ``None``.</span>

<span class="sd">        more_rna_pca_args:</span>
<span class="sd">            Additional arguments to pass to :py:func:`~scranpy.run_pca_se`.</span>
<span class="sd">            Only used if ``rna_altexp`` is not ``None``.</span>

<span class="sd">        adt_pca_output_name:</span>
<span class="sd">            Passed to :py:func:`~scranpy.run_pca_se` as the ``output_name`` argument.</span>
<span class="sd">            Only used if ``adt_altexp`` is not ``None``.</span>

<span class="sd">        more_adt_pca_args:</span>
<span class="sd">            Additional arguments to pass to :py:func:`~scranpy.run_pca_se`.</span>
<span class="sd">            Only used if ``adt_altexp`` is not ``None``.</span>

<span class="sd">        use_rna_pcs:</span>
<span class="sd">            Whether to use the RNA-derived PCs for downstream steps (i.e., clustering, visualization).</span>
<span class="sd">            Only used if ``rna_altexp`` is not ``None``.</span>

<span class="sd">        use_adt_pcs:</span>
<span class="sd">            Whether to use the ADT-derived PCs for downstream steps (i.e., clustering, visualization).</span>
<span class="sd">            Only used if ``adt_altexp`` is not ``None``.</span>

<span class="sd">        scale_output_name:</span>
<span class="sd">            Passed to :py:func:`~scranpy.scale_by_neighbors_se` as the ``output_name`` argument.</span>
<span class="sd">            Only used if multiple modalities are available and their corresponding ``use_*_pcs`` arguments are ``True``.</span>

<span class="sd">        more_scale_args:</span>
<span class="sd">            Additional arguments to pass to :py:func:`~scranpy.scale_by_neighbors_se`.</span>
<span class="sd">            Only used if multiple modalities are available and their corresponding ``use_*_pcs`` arguments are ``True``.</span>

<span class="sd">        mnn_output_name:</span>
<span class="sd">            Passed to :py:func:`~scranpy.correct_mnn_se` as the ``output_name`` argument.</span>
<span class="sd">            Only used if ``block`` is not ``None``.</span>

<span class="sd">        more_mnn_args:</span>
<span class="sd">            Additional arguments to pass to :py:func:`~scranpy.correct_mnn_se`.</span>
<span class="sd">            Only used if ``block`` is not ``None``.</span>

<span class="sd">        more_tsne_args:</span>
<span class="sd">            Passed to :py:func:`~scranpy.run_all_neighbor_steps_se`.</span>

<span class="sd">        more_umap_args:</span>
<span class="sd">            Passed to :py:func:`~scranpy.run_all_neighbor_steps_se`.</span>

<span class="sd">        more_build_graph_args:</span>
<span class="sd">            Passed to :py:func:`~scranpy.run_all_neighbor_steps_se`.</span>

<span class="sd">        cluster_graph_output_name:</span>
<span class="sd">            Passed to :py:func:`~scranpy.run_all_neighbor_steps_se` as ``cluster_output_name``.</span>

<span class="sd">        more_cluster_graph_args:</span>
<span class="sd">            Passed to :py:func:`~scranpy.run_all_neighbor_steps_se`.</span>

<span class="sd">        more_neighbor_args:</span>
<span class="sd">            Passed to :py:func:`~scranpy.run_all_neighbor_steps_se`.</span>

<span class="sd">        kmeans_clusters:</span>
<span class="sd">            Passed to :py:func:`~scranpy.cluster_kmeans_se` as the ``k`` argument.</span>
<span class="sd">            If ``None``, k-means clustering is not performed.</span>

<span class="sd">        kmeans_clusters_output_name:</span>
<span class="sd">            Passed to :py:func:`~scranpy.cluster_kmeans_se` as the ``output_name`` argument.</span>
<span class="sd">            Ignored if ``kmeans_clusters`` is ``None``.</span>

<span class="sd">        more_kmeans_args:</span>
<span class="sd">            Additional arguments to pass to :py:func:`~scranpy.cluster_kmeans_se`. </span>
<span class="sd">            Ignored if ``kmeans_clusters`` is ``None``.</span>

<span class="sd">        clusters_for_markers:</span>
<span class="sd">            List of clustering algorithms, specifying the clusters to be used for marker detection.</span>
<span class="sd">            The first available clustering will be chosen.</span>
<span class="sd">            If no clustering is available from the list, markers will not be computed.</span>

<span class="sd">        more_rna_marker_args:</span>
<span class="sd">            Additional arguments to pass to :py:func:`~scranpy.score_markers_se` for the RNA data.</span>
<span class="sd">            Ignored if no suitable clusterings are available or if ``rna_altexp`` is ``None``</span>

<span class="sd">        more_adt_marker_args:</span>
<span class="sd">            Additional arguments to pass to :py:func:`~scranpy.score_markers_se` for the ADT data.</span>
<span class="sd">            Ignored if no suitable clusterings are available or if ``adt_altexp`` is ``None``</span>

<span class="sd">        more_crispr_marker_args:</span>
<span class="sd">            Additional arguments to pass to :py:func:`~scranpy.score_markers_se` for the CRISPR data.</span>
<span class="sd">            Ignored if no suitable clusterings are available or if ``crispr_altexp`` is ``None``</span>

<span class="sd">        nn_parameters:</span>
<span class="sd">            Parameters for the nearest-neighbor search.</span>

<span class="sd">        num_threads:</span>
<span class="sd">            Number of threads to use in each step.</span>

<span class="sd">    Returns:</span>
<span class="sd">        A :py:class:`~biocutils.NamedList.NamedList` containing:</span>

<span class="sd">        - ``x``: a :py:class:`~singlecellexperiment.SingleCellExperiment.SingleCellExperiment` that is a copy of the input ``x``.</span>
<span class="sd">          It is also decorated with the results of each analysis step. </span>
<span class="sd">        - ``markers``: a list of list of :py:class:`~biocframe.BiocFrame.BiocFrame` objects containing the marker statistics for each modality.</span>
<span class="sd">          Each inner list corresponds to a modality (RNA, ADT, etc.) while each ``BiocFrame`` corresponds to a group.</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; import scranpy</span>
<span class="sd">        &gt;&gt;&gt; sce = scranpy.get_test_rna_data_se()</span>
<span class="sd">        &gt;&gt;&gt; is_mito = [ s.startswith(&quot;mt-&quot;) for s in sce.get_row_names() ]</span>
<span class="sd">        &gt;&gt;&gt; res = scranpy.analyze_se(</span>
<span class="sd">        &gt;&gt;&gt;     sce, </span>
<span class="sd">        &gt;&gt;&gt;     rna_qc_subsets={ &quot;mito&quot;: is_mito }</span>
<span class="sd">        &gt;&gt;&gt; )</span>
<span class="sd">        &gt;&gt;&gt; res[&quot;x&quot;].get_assay_names()</span>
<span class="sd">        &gt;&gt;&gt; res[&quot;x&quot;].get_reduced_dimension_names()</span>
<span class="sd">        &gt;&gt;&gt; print(res[&quot;x&quot;].get_column_data())</span>
<span class="sd">        &gt;&gt;&gt; import biocutils </span>
<span class="sd">        &gt;&gt;&gt; print(biocutils.table(res[&quot;x&quot;].get_column_data()[&quot;graph_cluster&quot;]))</span>
<span class="sd">        &gt;&gt;&gt; print(scranpy.preview_markers(res[&quot;markers&quot;][&quot;rna&quot;][0]))</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1">############ Quality control o(*°▽°*)o #############</span>

    <span class="n">collected_filters</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">if</span> <span class="n">rna_altexp</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">_get_modality</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">rna_altexp</span><span class="p">)</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">quick_rna_qc_se</span><span class="p">(</span>
            <span class="n">tmp</span><span class="p">,</span>
            <span class="n">output_prefix</span><span class="o">=</span><span class="n">rna_qc_output_prefix</span><span class="p">,</span>
            <span class="n">subsets</span><span class="o">=</span><span class="n">rna_qc_subsets</span><span class="p">,</span>
            <span class="n">block</span><span class="o">=</span><span class="n">block</span><span class="p">,</span>
            <span class="n">num_threads</span><span class="o">=</span><span class="n">num_threads</span><span class="p">,</span>
            <span class="n">assay_type</span><span class="o">=</span><span class="n">rna_assay_type</span><span class="p">,</span>
            <span class="o">**</span><span class="n">more_rna_qc_args</span>
        <span class="p">)</span>
        <span class="n">collected_filters</span><span class="p">[</span><span class="s2">&quot;rna&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">()[</span><span class="n">_maybe_concat</span><span class="p">(</span><span class="n">rna_qc_output_prefix</span><span class="p">,</span> <span class="s2">&quot;keep&quot;</span><span class="p">)]</span>
        <span class="n">x</span> <span class="o">=</span> <span class="n">_set_modality</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">rna_altexp</span><span class="p">,</span> <span class="n">tmp</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">adt_altexp</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">_get_modality</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">adt_altexp</span><span class="p">)</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">quick_adt_qc_se</span><span class="p">(</span>
            <span class="n">tmp</span><span class="p">,</span>
            <span class="n">output_prefix</span><span class="o">=</span><span class="n">adt_qc_output_prefix</span><span class="p">,</span>
            <span class="n">subsets</span><span class="o">=</span><span class="n">adt_qc_subsets</span><span class="p">,</span>
            <span class="n">block</span><span class="o">=</span><span class="n">block</span><span class="p">,</span>
            <span class="n">num_threads</span><span class="o">=</span><span class="n">num_threads</span><span class="p">,</span>
            <span class="n">assay_type</span><span class="o">=</span><span class="n">adt_assay_type</span><span class="p">,</span>
            <span class="o">**</span><span class="n">more_adt_qc_args</span>
        <span class="p">)</span>
        <span class="n">collected_filters</span><span class="p">[</span><span class="s2">&quot;adt&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">()[</span><span class="n">_maybe_concat</span><span class="p">(</span><span class="n">adt_qc_output_prefix</span><span class="p">,</span> <span class="s2">&quot;keep&quot;</span><span class="p">)]</span>
        <span class="n">x</span> <span class="o">=</span> <span class="n">_set_modality</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">adt_altexp</span><span class="p">,</span> <span class="n">tmp</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">crispr_altexp</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">_get_modality</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">crispr_altexp</span><span class="p">)</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">quick_crispr_qc_se</span><span class="p">(</span>
            <span class="n">tmp</span><span class="p">,</span>
            <span class="n">output_prefix</span><span class="o">=</span><span class="n">crispr_qc_output_prefix</span><span class="p">,</span>
            <span class="n">num_threads</span><span class="o">=</span><span class="n">num_threads</span><span class="p">,</span>
            <span class="n">block</span><span class="o">=</span><span class="n">block</span><span class="p">,</span>
            <span class="n">assay_type</span><span class="o">=</span><span class="n">crispr_assay_type</span><span class="p">,</span>
            <span class="o">**</span><span class="n">more_crispr_qc_args</span>
        <span class="p">)</span>
        <span class="n">collected_filters</span><span class="p">[</span><span class="s2">&quot;crispr&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">()[</span><span class="n">_maybe_concat</span><span class="p">(</span><span class="n">crispr_qc_output_prefix</span><span class="p">,</span> <span class="s2">&quot;keep&quot;</span><span class="p">)]</span>
        <span class="n">x</span> <span class="o">=</span> <span class="n">_set_modality</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">crispr_altexp</span><span class="p">,</span> <span class="n">tmp</span><span class="p">)</span>

    <span class="c1"># Combining all filters.</span>
    <span class="n">combined_qc_filter</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="n">numpy</span><span class="o">.</span><span class="n">bool</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">keep</span> <span class="ow">in</span> <span class="n">collected_filters</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
        <span class="n">combined_qc_filter</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">logical_and</span><span class="p">(</span><span class="n">combined_qc_filter</span><span class="p">,</span> <span class="n">keep</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">collected_filters</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">()</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">set_column</span><span class="p">(</span><span class="s2">&quot;combined_keep&quot;</span><span class="p">,</span> <span class="n">combined_qc_filter</span><span class="p">)</span>
        <span class="n">x</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">set_column_data</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">filter_cells</span><span class="p">:</span>
        <span class="n">x</span> <span class="o">=</span> <span class="n">_delayify_assays</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
        <span class="n">x</span> <span class="o">=</span> <span class="n">x</span><span class="p">[:,</span><span class="n">combined_qc_filter</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">block</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">filter_indices</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">combined_qc_filter</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">k</span><span class="p">:</span>
                    <span class="n">filter_indices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
            <span class="n">block</span> <span class="o">=</span> <span class="n">biocutils</span><span class="o">.</span><span class="n">subset</span><span class="p">(</span><span class="n">block</span><span class="p">,</span> <span class="n">filter_indices</span><span class="p">)</span>

    <span class="c1">############ Normalization ( ꈍᴗꈍ) #############</span>

    <span class="k">if</span> <span class="n">rna_altexp</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">_get_modality</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">rna_altexp</span><span class="p">)</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">normalize_rna_counts_se</span><span class="p">(</span>
            <span class="n">tmp</span><span class="p">,</span>
            <span class="n">assay_type</span><span class="o">=</span><span class="n">rna_assay_type</span><span class="p">,</span>
            <span class="n">output_name</span><span class="o">=</span><span class="n">rna_norm_output_name</span><span class="p">,</span>
            <span class="n">size_factors</span><span class="o">=</span><span class="n">tmp</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">()[</span><span class="n">_maybe_concat</span><span class="p">(</span><span class="n">rna_qc_output_prefix</span><span class="p">,</span> <span class="s2">&quot;sum&quot;</span><span class="p">)],</span>
            <span class="n">block</span><span class="o">=</span><span class="n">block</span><span class="p">,</span>
            <span class="o">**</span><span class="n">more_rna_norm_args</span>
        <span class="p">)</span>
        <span class="n">x</span> <span class="o">=</span> <span class="n">_set_modality</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">rna_altexp</span><span class="p">,</span> <span class="n">tmp</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">adt_altexp</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">_get_modality</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">adt_altexp</span><span class="p">)</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">normalize_adt_counts_se</span><span class="p">(</span>
            <span class="n">tmp</span><span class="p">,</span>
            <span class="n">assay_type</span><span class="o">=</span><span class="n">adt_assay_type</span><span class="p">,</span>
            <span class="n">output_name</span><span class="o">=</span><span class="n">adt_norm_output_name</span><span class="p">,</span>
            <span class="n">block</span><span class="o">=</span><span class="n">block</span><span class="p">,</span>
            <span class="o">**</span><span class="n">more_adt_norm_args</span>
        <span class="p">)</span>
        <span class="n">x</span> <span class="o">=</span> <span class="n">_set_modality</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">adt_altexp</span><span class="p">,</span> <span class="n">tmp</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">crispr_altexp</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">_get_modality</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">crispr_altexp</span><span class="p">)</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">normalize_crispr_counts_se</span><span class="p">(</span>
            <span class="n">tmp</span><span class="p">,</span>
            <span class="n">assay_type</span><span class="o">=</span><span class="n">crispr_assay_type</span><span class="p">,</span>
            <span class="n">output_name</span><span class="o">=</span><span class="n">crispr_norm_output_name</span><span class="p">,</span>
            <span class="n">size_factors</span><span class="o">=</span><span class="n">tmp</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">()[</span><span class="n">_maybe_concat</span><span class="p">(</span><span class="n">crispr_qc_output_prefix</span><span class="p">,</span> <span class="s2">&quot;sum&quot;</span><span class="p">)],</span>
            <span class="n">block</span><span class="o">=</span><span class="n">block</span><span class="p">,</span>
            <span class="o">**</span><span class="n">more_crispr_norm_args</span>
        <span class="p">)</span>
        <span class="n">x</span> <span class="o">=</span> <span class="n">_set_modality</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">crispr_altexp</span><span class="p">,</span> <span class="n">tmp</span><span class="p">)</span>

    <span class="c1">############ Variance modelling (～￣▽￣)～ #############</span>

    <span class="k">if</span> <span class="n">rna_altexp</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">_get_modality</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">rna_altexp</span><span class="p">)</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">choose_rna_hvgs_se</span><span class="p">(</span>
            <span class="n">tmp</span><span class="p">,</span>
            <span class="n">output_prefix</span><span class="o">=</span><span class="n">rna_hvg_output_prefix</span><span class="p">,</span>
            <span class="n">block</span><span class="o">=</span><span class="n">block</span><span class="p">,</span>
            <span class="o">**</span><span class="n">more_rna_hvg_args</span>
        <span class="p">)</span>
        <span class="n">x</span> <span class="o">=</span> <span class="n">_set_modality</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">rna_altexp</span><span class="p">,</span> <span class="n">tmp</span><span class="p">)</span>

    <span class="c1">############ Principal components analysis \(&gt;⩊&lt;)/ #############</span>

    <span class="k">if</span> <span class="n">rna_altexp</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">_get_modality</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">rna_altexp</span><span class="p">)</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">run_pca_se</span><span class="p">(</span>
            <span class="n">tmp</span><span class="p">,</span>
            <span class="n">assay_type</span><span class="o">=</span><span class="n">rna_norm_output_name</span><span class="p">,</span>
            <span class="n">output_name</span><span class="o">=</span><span class="n">rna_pca_output_name</span><span class="p">,</span>
            <span class="n">features</span><span class="o">=</span><span class="n">tmp</span><span class="o">.</span><span class="n">get_row_data</span><span class="p">()[</span><span class="n">_maybe_concat</span><span class="p">(</span><span class="n">rna_hvg_output_prefix</span><span class="p">,</span> <span class="s2">&quot;hvg&quot;</span><span class="p">)],</span>
            <span class="n">block</span><span class="o">=</span><span class="n">block</span><span class="p">,</span>
            <span class="o">**</span><span class="n">more_rna_pca_args</span>
        <span class="p">)</span>
        <span class="n">x</span> <span class="o">=</span> <span class="n">_set_modality</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">rna_altexp</span><span class="p">,</span> <span class="n">tmp</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">adt_altexp</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">_get_modality</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">adt_altexp</span><span class="p">)</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">run_pca_se</span><span class="p">(</span>
            <span class="n">tmp</span><span class="p">,</span>
            <span class="n">assay_type</span><span class="o">=</span><span class="n">adt_norm_output_name</span><span class="p">,</span>
            <span class="n">output_name</span><span class="o">=</span><span class="n">adt_pca_output_name</span><span class="p">,</span>
            <span class="n">features</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">block</span><span class="o">=</span><span class="n">block</span><span class="p">,</span>
            <span class="o">**</span><span class="n">more_adt_pca_args</span>
        <span class="p">)</span>
        <span class="n">x</span> <span class="o">=</span> <span class="n">_set_modality</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">adt_altexp</span><span class="p">,</span> <span class="n">tmp</span><span class="p">)</span>

    <span class="c1">############ Combining modalities („• ᴗ •„) #############</span>

    <span class="n">embeddings</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="n">use_rna_pcs</span> <span class="ow">and</span> <span class="n">rna_altexp</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">embeddings</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;rna&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">use_adt_pcs</span> <span class="ow">and</span> <span class="n">adt_altexp</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">embeddings</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;adt&quot;</span><span class="p">)</span>

    <span class="n">target_embedding</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">embeddings</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;at least one &#39;use_*&#39; must be true&quot;</span><span class="p">)</span>

    <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">embeddings</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">embeddings</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;rna&quot;</span><span class="p">:</span>
            <span class="n">target_embedding</span> <span class="o">=</span> <span class="n">_define_single_target_embedding</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">rna_altexp</span><span class="p">,</span> <span class="n">rna_pca_output_name</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">embeddings</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;adt&quot;</span><span class="p">:</span>
            <span class="n">target_embedding</span> <span class="o">=</span> <span class="n">_define_single_target_embedding</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">adt_altexp</span><span class="p">,</span> <span class="n">adt_pca_output_name</span><span class="p">)</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="n">main_reddims</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">altexp_reddims</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">if</span> <span class="s2">&quot;rna&quot;</span> <span class="ow">in</span> <span class="n">embeddings</span><span class="p">:</span>
            <span class="n">_add_source_embedding_to_scale</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">rna_altexp</span><span class="p">,</span> <span class="n">rna_pca_output_name</span><span class="p">,</span> <span class="n">main_reddims</span><span class="p">,</span> <span class="n">altexp_reddims</span><span class="p">)</span>
        <span class="k">if</span> <span class="s2">&quot;adt&quot;</span> <span class="ow">in</span> <span class="n">embeddings</span><span class="p">:</span>
            <span class="n">_add_source_embedding_to_scale</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">adt_altexp</span><span class="p">,</span> <span class="n">adt_pca_output_name</span><span class="p">,</span> <span class="n">main_reddims</span><span class="p">,</span> <span class="n">altexp_reddims</span><span class="p">)</span>

        <span class="n">x</span> <span class="o">=</span> <span class="n">scale_by_neighbors_se</span><span class="p">(</span> 
            <span class="n">x</span><span class="p">,</span>
            <span class="n">output_name</span><span class="o">=</span><span class="n">scale_output_name</span><span class="p">,</span>
            <span class="n">main_reddims</span><span class="o">=</span><span class="n">main_reddims</span><span class="p">,</span>
            <span class="n">altexp_reddims</span><span class="o">=</span><span class="n">altexp_reddims</span><span class="p">,</span>
            <span class="n">block</span><span class="o">=</span><span class="n">block</span><span class="p">,</span>
            <span class="n">num_threads</span><span class="o">=</span><span class="n">num_threads</span><span class="p">,</span>
            <span class="n">nn_parameters</span><span class="o">=</span><span class="n">nn_parameters</span><span class="p">,</span>
            <span class="o">**</span><span class="n">more_scale_args</span>
        <span class="p">)</span>
        <span class="n">target_embedding</span> <span class="o">=</span> <span class="n">scale_output_name</span>

    <span class="c1">############ Performing batch correction ⸜(｡˃ ᵕ ˂ )⸝ #############</span>

    <span class="k">if</span> <span class="n">block</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">x</span> <span class="o">=</span> <span class="n">correct_mnn_se</span><span class="p">(</span>
            <span class="n">x</span><span class="p">,</span>
            <span class="n">output_name</span><span class="o">=</span><span class="n">mnn_output_name</span><span class="p">,</span>
            <span class="n">reddim_type</span><span class="o">=</span><span class="n">target_embedding</span><span class="p">,</span>
            <span class="n">block</span><span class="o">=</span><span class="n">block</span><span class="p">,</span>
            <span class="n">num_threads</span><span class="o">=</span><span class="n">num_threads</span><span class="p">,</span>
            <span class="n">nn_parameters</span><span class="o">=</span><span class="n">nn_parameters</span><span class="p">,</span>
            <span class="o">**</span><span class="n">more_mnn_args</span>
        <span class="p">)</span>
        <span class="n">target_embedding</span> <span class="o">=</span> <span class="n">mnn_output_name</span>

        <span class="k">if</span> <span class="n">block_name</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">()</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">set_column</span><span class="p">(</span><span class="n">block_name</span><span class="p">,</span> <span class="n">block</span><span class="p">)</span>
            <span class="n">x</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">set_column_data</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>

    <span class="c1">############ Assorted neighbor-related stuff ⸜(⸝⸝⸝´꒳`⸝⸝⸝)⸝ #############</span>

    <span class="n">x</span> <span class="o">=</span> <span class="n">run_all_neighbor_steps_se</span><span class="p">(</span>
        <span class="n">x</span><span class="p">,</span>
        <span class="n">reddim_type</span><span class="o">=</span><span class="n">target_embedding</span><span class="p">,</span>
        <span class="n">cluster_output_name</span><span class="o">=</span><span class="n">cluster_graph_output_name</span><span class="p">,</span>
        <span class="n">more_umap_args</span><span class="o">=</span><span class="n">more_umap_args</span><span class="p">,</span> 
        <span class="n">more_tsne_args</span><span class="o">=</span><span class="n">more_tsne_args</span><span class="p">,</span> 
        <span class="n">more_build_graph_args</span><span class="o">=</span><span class="n">more_build_graph_args</span><span class="p">,</span>
        <span class="n">more_cluster_graph_args</span><span class="o">=</span><span class="n">more_cluster_graph_args</span><span class="p">,</span>
        <span class="n">more_neighbor_args</span><span class="o">=</span><span class="n">more_neighbor_args</span><span class="p">,</span>
        <span class="n">nn_parameters</span><span class="o">=</span><span class="n">nn_parameters</span><span class="p">,</span>
        <span class="n">num_threads</span><span class="o">=</span><span class="n">num_threads</span>
    <span class="p">)</span>

    <span class="c1">############ Maybe some k-means clustering ⸜(⸝⸝⸝´꒳`⸝⸝⸝)⸝ #############</span>

    <span class="k">if</span> <span class="n">kmeans_clusters</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">x</span> <span class="o">=</span> <span class="n">cluster_kmeans_se</span><span class="p">(</span>
            <span class="n">x</span><span class="p">,</span>
            <span class="n">k</span><span class="o">=</span><span class="n">kmeans_clusters</span><span class="p">,</span>
            <span class="n">reddim_type</span><span class="o">=</span><span class="n">target_embedding</span><span class="p">,</span>
            <span class="n">output_name</span><span class="o">=</span><span class="n">kmeans_clusters_output_name</span><span class="p">,</span>
            <span class="n">num_threads</span><span class="o">=</span><span class="n">num_threads</span><span class="p">,</span>
            <span class="o">**</span><span class="n">more_kmeans_args</span>
        <span class="p">)</span>

    <span class="n">chosen_clusters</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">clusters_for_markers</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">c</span> <span class="o">==</span> <span class="s2">&quot;graph&quot;</span><span class="p">:</span>
            <span class="n">name</span> <span class="o">=</span> <span class="n">cluster_graph_output_name</span>
        <span class="k">elif</span> <span class="n">c</span> <span class="o">==</span> <span class="s2">&quot;kmeans&quot;</span><span class="p">:</span>
            <span class="n">name</span> <span class="o">=</span> <span class="n">kmeans_clusters_output_name</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">()</span> 
        <span class="k">if</span> <span class="n">name</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">x</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">()</span><span class="o">.</span><span class="n">get_column_names</span><span class="p">():</span>
            <span class="n">chosen_clusters</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">()[</span><span class="n">name</span><span class="p">]</span>
            <span class="k">break</span>

    <span class="c1">############ Finding markers (˶˃ ᵕ ˂˶) #############</span>

    <span class="n">markers</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="k">if</span> <span class="n">chosen_clusters</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">markers</span> <span class="o">=</span> <span class="n">biocutils</span><span class="o">.</span><span class="n">NamedList</span><span class="p">([],</span> <span class="p">[])</span>

        <span class="k">if</span> <span class="n">rna_altexp</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">markers</span><span class="p">[</span><span class="s2">&quot;rna&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">score_markers_se</span><span class="p">(</span>
                <span class="n">_get_modality</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">rna_altexp</span><span class="p">),</span>
                <span class="n">groups</span><span class="o">=</span><span class="n">chosen_clusters</span><span class="p">,</span>
                <span class="n">assay_type</span><span class="o">=</span><span class="n">rna_norm_output_name</span><span class="p">,</span>
                <span class="n">num_threads</span><span class="o">=</span><span class="n">num_threads</span><span class="p">,</span>
                <span class="n">block</span><span class="o">=</span><span class="n">block</span><span class="p">,</span>
                <span class="o">**</span><span class="n">more_rna_marker_args</span>
            <span class="p">)</span>

        <span class="k">if</span> <span class="n">adt_altexp</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">markers</span><span class="p">[</span><span class="s2">&quot;adt&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">score_markers_se</span><span class="p">(</span>
                <span class="n">_get_modality</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">adt_altexp</span><span class="p">),</span>
                <span class="n">groups</span><span class="o">=</span><span class="n">chosen_clusters</span><span class="p">,</span>
                <span class="n">assay_type</span><span class="o">=</span><span class="n">adt_norm_output_name</span><span class="p">,</span>
                <span class="n">num_threads</span><span class="o">=</span><span class="n">num_threads</span><span class="p">,</span>
                <span class="n">block</span><span class="o">=</span><span class="n">block</span><span class="p">,</span>
                <span class="o">**</span><span class="n">more_adt_marker_args</span>
            <span class="p">)</span>

        <span class="k">if</span> <span class="n">crispr_altexp</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">markers</span><span class="p">[</span><span class="s2">&quot;crispr&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">score_markers_se</span><span class="p">(</span>
                <span class="n">_get_modality</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">crispr_altexp</span><span class="p">),</span>
                <span class="n">groups</span><span class="o">=</span><span class="n">chosen_clusters</span><span class="p">,</span>
                <span class="n">assay_type</span><span class="o">=</span><span class="n">crispr_norm_output_name</span><span class="p">,</span>
                <span class="n">num_threads</span><span class="o">=</span><span class="n">num_threads</span><span class="p">,</span>
                <span class="n">block</span><span class="o">=</span><span class="n">block</span><span class="p">,</span>
                <span class="o">**</span><span class="n">more_crispr_marker_args</span>
            <span class="p">)</span>

    <span class="k">return</span> <span class="n">biocutils</span><span class="o">.</span><span class="n">NamedList</span><span class="p">([</span><span class="n">x</span><span class="p">,</span> <span class="n">markers</span><span class="p">],</span> <span class="p">[</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="s2">&quot;markers&quot;</span><span class="p">])</span></div>


<span class="k">def</span><span class="w"> </span><span class="nf">_delayify_assays</span><span class="p">(</span><span class="n">x</span><span class="p">:</span> <span class="n">summarizedexperiment</span><span class="o">.</span><span class="n">SummarizedExperiment</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">summarizedexperiment</span><span class="o">.</span><span class="n">SummarizedExperiment</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">x</span><span class="o">.</span><span class="n">get_assay_names</span><span class="p">():</span>
        <span class="n">x</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">set_assay</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">delayedarray</span><span class="o">.</span><span class="n">DelayedArray</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">get_assay</span><span class="p">(</span><span class="n">i</span><span class="p">)))</span>

    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">singlecellexperiment</span><span class="o">.</span><span class="n">SingleCellExperiment</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">x</span><span class="o">.</span><span class="n">get_alternative_experiment_names</span><span class="p">():</span>
            <span class="n">x</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">set_alternative_experiment</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">_delayify_assays</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">get_alternative_experiment</span><span class="p">(</span><span class="n">i</span><span class="p">)))</span>

    <span class="k">return</span> <span class="n">x</span>


<span class="k">def</span><span class="w"> </span><span class="nf">_maybe_concat</span><span class="p">(</span><span class="n">prefix</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">prefix</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">name</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">prefix</span> <span class="o">+</span> <span class="n">name</span>


<span class="k">def</span><span class="w"> </span><span class="nf">_use_main_experiment</span><span class="p">(</span><span class="n">altexp</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">str</span><span class="p">,</span> <span class="nb">bool</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">altexp</span><span class="p">,</span> <span class="nb">bool</span><span class="p">):</span>
        <span class="k">return</span> <span class="kc">False</span>
    <span class="k">if</span> <span class="n">altexp</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;boolean &#39;*_altexp&#39; must be False to use the main experiment&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="kc">True</span>


<span class="k">def</span><span class="w"> </span><span class="nf">_get_modality</span><span class="p">(</span><span class="n">x</span><span class="p">:</span> <span class="n">summarizedexperiment</span><span class="o">.</span><span class="n">SummarizedExperiment</span><span class="p">,</span> <span class="n">altexp</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">str</span><span class="p">,</span> <span class="nb">bool</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">summarizedexperiment</span><span class="o">.</span><span class="n">SummarizedExperiment</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">_use_main_experiment</span><span class="p">(</span><span class="n">altexp</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">x</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">x</span><span class="o">.</span><span class="n">get_alternative_experiment</span><span class="p">(</span><span class="n">altexp</span><span class="p">)</span>


<span class="k">def</span><span class="w"> </span><span class="nf">_set_modality</span><span class="p">(</span>
    <span class="n">x</span><span class="p">:</span> <span class="n">summarizedexperiment</span><span class="o">.</span><span class="n">SummarizedExperiment</span><span class="p">,</span>
    <span class="n">altexp</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">str</span><span class="p">,</span> <span class="nb">bool</span><span class="p">],</span>
    <span class="n">replacement</span><span class="p">:</span> <span class="n">summarizedexperiment</span><span class="o">.</span><span class="n">SummarizedExperiment</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">summarizedexperiment</span><span class="o">.</span><span class="n">SummarizedExperiment</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">_use_main_experiment</span><span class="p">(</span><span class="n">altexp</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">replacement</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">x</span><span class="o">.</span><span class="n">set_alternative_experiment</span><span class="p">(</span><span class="n">altexp</span><span class="p">,</span> <span class="n">replacement</span><span class="p">)</span>


<span class="k">def</span><span class="w"> </span><span class="nf">_define_single_target_embedding</span><span class="p">(</span>
    <span class="n">x</span><span class="p">:</span> <span class="n">summarizedexperiment</span><span class="o">.</span><span class="n">SummarizedExperiment</span><span class="p">,</span>
    <span class="n">altexp</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">str</span><span class="p">,</span> <span class="nb">bool</span><span class="p">],</span>
    <span class="n">output_name</span><span class="p">:</span> <span class="nb">str</span>
<span class="p">):</span>
    <span class="k">if</span> <span class="n">_use_main_experiment</span><span class="p">(</span><span class="n">altexp</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">output_name</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">altexp</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
        <span class="n">altexp</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">get_alternative_experiment_names</span><span class="p">()[</span><span class="n">altexp</span><span class="p">]</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">altexp</span><span class="p">,</span> <span class="n">output_name</span><span class="p">)</span>


<span class="k">def</span><span class="w"> </span><span class="nf">_add_source_embedding_to_scale</span><span class="p">(</span>
    <span class="n">x</span><span class="p">:</span> <span class="n">summarizedexperiment</span><span class="o">.</span><span class="n">SummarizedExperiment</span><span class="p">,</span>
    <span class="n">altexp</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">str</span><span class="p">,</span> <span class="nb">bool</span><span class="p">],</span>
    <span class="n">output_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">main_reddim</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span>
    <span class="n">altexp_reddim</span><span class="p">:</span> <span class="nb">dict</span> 
<span class="p">):</span>
    <span class="k">if</span> <span class="n">_use_main_experiment</span><span class="p">(</span><span class="n">altexp</span><span class="p">):</span>
        <span class="n">main_reddim</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">output_name</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">altexp</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
            <span class="n">altexp</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">get_alternative_experiment_names</span><span class="p">()[</span><span class="n">altexp</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">altexp</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">altexp_reddim</span><span class="p">:</span> 
            <span class="n">altexp_reddim</span><span class="p">[</span><span class="n">altexp</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">altexp_reddim</span><span class="p">[</span><span class="n">altexp</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">output_name</span><span class="p">)</span>
</pre></div>
        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          
          
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2023, Jayaram Kancherla
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer no-toc">
      
      
      
    </aside>
  </div>
</div><script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js?v=c02300dd"></script>
    <script src="../../_static/doctools.js?v=888ff710"></script>
    <script src="../../_static/sphinx_highlight.js?v=4825356b"></script>
    <script src="../../_static/scripts/furo.js?v=46bd48cc"></script>
    </body>
</html>