<!doctype html>
<html class="no-js" lang="en" data-content_root="">
  <head><meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="index" title="Index" href="../genindex.html"><link rel="search" title="Search" href="../search.html"><link rel="prev" title="scranpy" href="modules.html">

    <!-- Generated with Sphinx 7.1.2 and Furo 2025.12.19 -->
        <title>scranpy package - scranpy 0.3.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo.css?v=7bdb33bb" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo-extensions.css?v=8dab3a3b" />
    
    


<style>
  body {
    --color-code-background: #eeffcc;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle site navigation sidebar">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc" aria-label="Toggle table of contents sidebar">
<label class="overlay sidebar-overlay" for="__navigation"></label>
<label class="overlay toc-overlay" for="__toc"></label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <span class="icon"><svg><use href="#svg-menu"></use></svg></span>
      </label>
    </div>
    <div class="header-center">
      <a href="../index.html"><div class="brand">scranpy 0.3.0 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle" aria-label="Toggle Light / Dark / Auto color theme">
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <span class="icon"><svg><use href="#svg-toc"></use></svg></span>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../index.html">
  
  <span class="sidebar-brand-text">scranpy 0.3.0 documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../readme.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contributing.html">Contributions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="../authors.html">Authors</a></li>
<li class="toctree-l1"><a class="reference internal" href="../changelog.html">Changelog</a></li>
<li class="toctree-l1 current has-children"><a class="reference internal" href="modules.html">Module Reference</a><input aria-label="Toggle navigation of Module Reference" checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><span class="icon"><svg><use href="#svg-arrow-right"></use></svg></span></label><ul class="current">
<li class="toctree-l2 current current-page"><a class="current reference internal" href="#">scranpy package</a></li>
</ul>
</li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="view-this-page">
  <a class="muted-link" href="../_sources/api/scranpy.rst.txt" title="View this page">
    <svg><use href="#svg-eye"></use></svg>
    <span class="visually-hidden">View this page</span>
  </a>
</div>
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle" aria-label="Toggle Light / Dark / Auto color theme">
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <span class="icon"><svg><use href="#svg-toc"></use></svg></span>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <section id="scranpy-package">
<h1>scranpy package<a class="headerlink" href="#scranpy-package" title="Permalink to this heading">¶</a></h1>
<section id="module-scranpy">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-scranpy" title="Permalink to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="scranpy.aggregate_across_cells">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">aggregate_across_cells</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">factors</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_aggregate_across_cells.html#aggregate_across_cells"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.aggregate_across_cells" title="Permalink to this definition">¶</a></dt>
<dd><p>Aggregate expression values across cells based on one or more grouping factors.
This is primarily used to create pseudo-bulk profiles for each cluster/sample combination.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Any" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a></span>) – A matrix-like object where rows correspond to genes or genomic features and columns correspond to cells.
Values are expected to be counts.</p></li>
<li><p><strong>factors</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>, <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a>, <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a>]</span>) – One or more grouping factors, see <a class="reference internal" href="#scranpy.combine_factors" title="scranpy.combine_factors"><code class="xref py py-func docutils literal notranslate"><span class="pre">combine_factors()</span></code></a>.
Each entry should be a sequence of length equal to the number of columns in <code class="docutils literal notranslate"><span class="pre">x</span></code>.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use for aggregation.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><p>A <code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code> containing the following entries.</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">sum</span></code>: double-precision NumPy matrix where each row corresponds to a gene and each column corresponds to a unique combination of grouping levels.
Each matrix entry contains the summed expression across all cells with that combination.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">detected</span></code>: integer NumPy matrix where each row corresponds to a gene and each column corresponds to a unique combination of grouping levels.
Each matrix entry contains the number of cells with detected expression in that combination.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">combinations</span></code>: a <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a> containing all unique combinations of levels across <code class="docutils literal notranslate"><span class="pre">factors</span></code>.
Each column corresponds to an entry of <code class="docutils literal notranslate"><span class="pre">factors</span></code> while each row corresponds to a combination.
Specifically, the <code class="docutils literal notranslate"><span class="pre">i</span></code>-th combination is defined as the <code class="docutils literal notranslate"><span class="pre">i</span></code>-th elements of all columns.
Combinations are in the same order as the columns of <code class="xref py py-attr docutils literal notranslate"><span class="pre">sum</span></code> and <code class="xref py py-attr docutils literal notranslate"><span class="pre">detected</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">counts</span></code>: an integer NumPy array containing the number of cells associated with each combination in <code class="docutils literal notranslate"><span class="pre">combinations</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">index</span></code>: an Integer NumPy array of length equal to the number of cells.
This specifies the combination <code class="docutils literal notranslate"><span class="pre">combinations</span></code> associated with each cell in <code class="docutils literal notranslate"><span class="pre">x</span></code>.</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The <code class="docutils literal notranslate"><span class="pre">aggregate_across_cells</span></code> function in the <a class="reference external" href="https://libscran.github.io/scran_aggregate">scran_aggregate</a> C++ library.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mat</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="mi">100</span><span class="p">,</span> <span class="mi">20</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">clusters</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="s2">&quot;D&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">blocks</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="mi">4</span> <span class="o">+</span> <span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">*</span> <span class="mi">4</span> <span class="o">+</span> <span class="p">[</span><span class="mi">3</span><span class="p">]</span> <span class="o">*</span> <span class="mi">4</span> <span class="o">+</span> <span class="p">[</span><span class="mi">4</span><span class="p">]</span> <span class="o">*</span> <span class="mi">4</span> <span class="o">+</span> <span class="p">[</span><span class="mi">5</span><span class="p">]</span> <span class="o">*</span> <span class="mi">4</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aggr</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">aggregate_across_cells</span><span class="p">(</span><span class="n">mat</span><span class="p">,</span> <span class="p">{</span> <span class="s2">&quot;clusters&quot;</span><span class="p">:</span> <span class="n">clusters</span><span class="p">,</span> <span class="s2">&quot;blocks&quot;</span><span class="p">:</span> <span class="n">blocks</span> <span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aggr</span><span class="p">[</span><span class="s2">&quot;sum&quot;</span><span class="p">][:</span><span class="mi">5</span><span class="p">,]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">aggr</span><span class="p">[</span><span class="s2">&quot;combinations&quot;</span><span class="p">])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.aggregate_across_cells_se">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">aggregate_across_cells_se</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">factors</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">more_aggr_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assay_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'counts'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'factor_'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">counts_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'counts'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">meta_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'aggregated'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_coldata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">more_coldata_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">altexps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy_altexps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.aggregate_across_cells_se" title="Permalink to this definition">¶</a></dt>
<dd><p>Aggregate expression values across groups of cells for each gene, storing the result in a <a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>.
This calls <a class="reference internal" href="#scranpy.aggregate_across_cells" title="scranpy.aggregate_across_cells"><code class="xref py py-func docutils literal notranslate"><span class="pre">aggregate_across_cells()</span></code></a> along with <a class="reference internal" href="#scranpy.aggregate_column_data" title="scranpy.aggregate_column_data"><code class="xref py py-func docutils literal notranslate"><span class="pre">aggregate_column_data()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> – A <a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a> object or one of its subclasses.
Rows correspond to genes and columns correspond to cells.</p></li>
<li><p><strong>factors</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>, <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a>, <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a>]</span>) – One or more grouping factors, see the argument of the same name in <a class="reference internal" href="#scranpy.aggregate_across_cells" title="scranpy.aggregate_across_cells"><code class="xref py py-func docutils literal notranslate"><span class="pre">aggregate_across_cells()</span></code></a>.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Passed to <a class="reference internal" href="#scranpy.aggregate_across_cells" title="scranpy.aggregate_across_cells"><code class="xref py py-func docutils literal notranslate"><span class="pre">aggregate_across_cells()</span></code></a>.</p></li>
<li><p><strong>more_aggr_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Further arguments to pass to <a class="reference internal" href="#scranpy.aggregate_across_cells" title="scranpy.aggregate_across_cells"><code class="xref py py-func docutils literal notranslate"><span class="pre">aggregate_across_cells()</span></code></a>.</p></li>
<li><p><strong>assay_type</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>) – Name or index of the assay of <code class="docutils literal notranslate"><span class="pre">x</span></code> to be aggregated.</p></li>
<li><p><strong>output_prefix</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – Prefix to add to the names of the columns containing the factor combinations in the column data of the output object.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, no prefix is added.</p></li>
<li><p><strong>counts_name</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></span>) – Name of the column in which to store the cell count for each factor combination in the column data of the output object.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, the cell counts are not reported.</p></li>
<li><p><strong>meta_name</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – Name of the metadata entry in which to store additional information like the combination indices in the output object.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, additional outputs are not reported.</p></li>
<li><p><strong>include_coldata</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to add the aggregated column data from <code class="docutils literal notranslate"><span class="pre">x</span></code> to the output.
If <code class="docutils literal notranslate"><span class="pre">True</span></code>, this adds the output of <a class="reference internal" href="#scranpy.aggregate_column_data" title="scranpy.aggregate_column_data"><code class="xref py py-func docutils literal notranslate"><span class="pre">aggregate_column_data()</span></code></a> to the column data of the output object.</p></li>
<li><p><strong>more_coldata_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Additional arguments to pass to <a class="reference internal" href="#scranpy.aggregate_column_data" title="scranpy.aggregate_column_data"><code class="xref py py-func docutils literal notranslate"><span class="pre">aggregate_column_data()</span></code></a>.
Only relevant if <code class="docutils literal notranslate"><span class="pre">include_coldata</span> <span class="pre">=</span> <span class="pre">True</span></code>.</p></li>
<li><p><strong>altexps</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>, <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/constants.html#None" title="(in Python v3.14)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]</span>) – <p>List of integers or strings, containing the indices or names of alternative experiments of <code class="docutils literal notranslate"><span class="pre">x</span></code> to aggregate.
The aggregated assay from each alternative experiment is determined by <code class="docutils literal notranslate"><span class="pre">assay_type</span></code>.</p>
<p>Alternatively, this may be a single integer or string containing the index or name of one alternative experiment to aggregate.
Again, the aggregated assay from each alternative experiment is determined by <code class="docutils literal notranslate"><span class="pre">assay_type</span></code>.</p>
<p>Alternatively, this may be a dictionary where keys are strings and values are integers or strings.
Each key should be the name of an alternative experiment while each value is the index/name of the assay to aggregate from that experiment.</p>
<p>Alternatively, a <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a> of integers or strings.
If named, this is treated as a dictionary, otherwise it is treated as a list.</p>
<p>Only relevant if <code class="docutils literal notranslate"><span class="pre">x</span></code> is a <a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a> or one of its subclasses.</p>
</p></li>
<li><p><strong>copy_altexps</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to copy the column data and metadata of the output <code class="docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code> into each of its alternative experiments.
Only relevant if <code class="docutils literal notranslate"><span class="pre">x</span></code> is a <code class="docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code> or one of its subclasses.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><p>A <a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a> where each column corresponds to a factor combination.
Each row corresponds to a gene in <code class="docutils literal notranslate"><span class="pre">x</span></code>, and the row data is taken directly from <code class="docutils literal notranslate"><span class="pre">x</span></code>.
The assays contain the sum of counts (<code class="docutils literal notranslate"><span class="pre">sum</span></code>) and the number of detected cells (<code class="docutils literal notranslate"><span class="pre">detected</span></code>) in each combination for each gene.
The column data contains:</p>
<ul class="simple">
<li><p>The factor combinations, with column names prefixed by <code class="docutils literal notranslate"><span class="pre">output_prefix</span></code>.</p></li>
<li><p>The cell count for each combination, named by <code class="docutils literal notranslate"><span class="pre">counts_name</span></code>.</p></li>
<li><p>Additional column data from <code class="docutils literal notranslate"><span class="pre">x</span></code> if <code class="docutils literal notranslate"><span class="pre">include_coldata</span> <span class="pre">=</span> <span class="pre">True</span></code>.
This is aggregated with <code class="xref py py-func docutils literal notranslate"><span class="pre">aggregate_column_data`()</span></code> on the combination indices.</p></li>
</ul>
<p>The metadata contains a <code class="docutils literal notranslate"><span class="pre">meta_name</span></code> entry, which is a list with an <code class="docutils literal notranslate"><span class="pre">index</span></code> integer vector of length equal to the number of cells in <code class="docutils literal notranslate"><span class="pre">x</span></code>.
Each entry of this vector is an index of the factor combination (i.e., column of the output object) to which the corresponding cell was assigned.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">altexps</span></code> is specified, a <a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a> is returned instead.
The same aggregation described above for the main experiment is applied to each alternative experiment.
If <code class="docutils literal notranslate"><span class="pre">copy_altexps</span> <span class="pre">=</span> <span class="pre">True</span></code>, the column data for each alternative experiment will contain a copy of the factor combinations and cell counts,
and the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> will contain a copy of the index vector.</p>
</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">get_test_rna_data_se</span><span class="p">(</span><span class="s2">&quot;start&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aggr</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">aggregate_across_cells_se</span><span class="p">(</span><span class="n">sce</span><span class="p">,</span> <span class="n">factors</span><span class="o">=</span><span class="p">[</span><span class="n">sce</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">()[</span><span class="s2">&quot;level1class&quot;</span><span class="p">]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aggr</span><span class="o">.</span><span class="n">get_assay</span><span class="p">(</span><span class="mi">0</span><span class="p">)[:</span><span class="mi">5</span><span class="p">,]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">aggr</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># We can also aggregate within alternative experiments.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aggr2</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">aggregate_across_cells_se</span><span class="p">(</span><span class="n">sce</span><span class="p">,</span> <span class="n">factors</span><span class="o">=</span><span class="p">[</span><span class="n">sce</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">()[</span><span class="s2">&quot;level1class&quot;</span><span class="p">]],</span> <span class="n">altexps</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;ERCC&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aggr2</span><span class="o">.</span><span class="n">get_alternative_experiment</span><span class="p">(</span><span class="s2">&quot;ERCC&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">get_assay</span><span class="p">(</span><span class="mi">0</span><span class="p">)[:</span><span class="mi">5</span><span class="p">,]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.aggregate_across_genes">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">aggregate_across_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sets</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">row_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">average</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_aggregate_across_genes.html#aggregate_across_genes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.aggregate_across_genes" title="Permalink to this definition">¶</a></dt>
<dd><p>Aggregate expression values across genes, potentially with weights.
This is typically used to summarize expression values for gene sets into a single per-cell score.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Any" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a></span>) – Matrix-like object where rows correspond to genes or genomic features and columns correspond to cells.
Values are expected to be log-expression values.</p></li>
<li><p><strong>sets</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>, <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a>]</span>) – <p>Sequence of gene sets.
Each gene set may be represented by:</p>
<ul>
<li><p>A sequence of integers, specifying the row indices of the genes in that set.</p></li>
<li><p>A sequence of strings, specifying the row names of the genes in that set.
If any strings are present, <code class="docutils literal notranslate"><span class="pre">row_names</span></code> should also be supplied.</p></li>
<li><p>A tuple of length 2, containing a sequence of strings/integers (row names/indices) and a numeric array (weights).</p></li>
<li><p>A <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a> where each row corresponds to a gene.
The first column contains the row names/indices and the second column contains the weights.</p></li>
</ul>
<p>Alternatively, a dictionary may be supplied where each key is the name of a gene set and each value is a sequence/tuple as described above.
The keys will be used to name the output <code class="docutils literal notranslate"><span class="pre">NamedList</span></code>.</p>
<p>Alternatively, a <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a> where each entry is a gene set represented by a sequence/tuple as described above.
If names are available, they will be used to name the output <code class="docutils literal notranslate"><span class="pre">NamedList</span></code>.</p>
</p></li>
<li><p><strong>row_names</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Sequence of strings of length equal to the number of rows of <code class="docutils literal notranslate"><span class="pre">x</span></code>, containing the name of each gene.
Duplicate names are allowed but only the first occurrence will be used.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, rows are assumed to be unnamed.</p></li>
<li><p><strong>average</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to compute the average rather than the sum.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to be used for aggregation.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a> of length equal to that of <code class="docutils literal notranslate"><span class="pre">sets</span></code>.
Each entry is a numeric vector of length equal to the number of columns in <code class="docutils literal notranslate"><span class="pre">x</span></code>,
containing the (weighted) sum/mean of expression values for the corresponding set across all cells.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The <code class="docutils literal notranslate"><span class="pre">aggregate_across_genes</span></code> function in the <a class="reference external" href="https://libscran.github.io/scran_aggregate">scran_aggregate</a> C++ library, which implements the aggregation.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mat</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="mi">100</span><span class="p">,</span> <span class="mi">20</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sets</span> <span class="o">=</span> <span class="p">{</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="s2">&quot;foo&quot;</span><span class="p">:</span> <span class="p">[</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="mi">9</span> <span class="p">],</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="s2">&quot;bar&quot;</span><span class="p">:</span> <span class="nb">range</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">40</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aggr</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">aggregate_across_genes</span><span class="p">(</span><span class="n">mat</span><span class="p">,</span> <span class="n">sets</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">aggr</span><span class="o">.</span><span class="n">get_names</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">aggr</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.aggregate_across_genes_se">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">aggregate_across_genes_se</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sets</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">more_aggr_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assay_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'logcounts'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_se_aggregate_across_genes.html#aggregate_across_genes_se"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.aggregate_across_genes_se" title="Permalink to this definition">¶</a></dt>
<dd><p>Aggregate expression values across sets of genes for each cell.
This calls <a class="reference internal" href="#scranpy.aggregate_across_genes" title="scranpy.aggregate_across_genes"><code class="xref py py-func docutils literal notranslate"><span class="pre">aggregate_across_genes()</span></code></a> on an assay from a <a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a></span>) – A <a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a> object or one of its subclasses.
Rows correspond to genes and columns correspond to cells.</p></li>
<li><p><strong>sets</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a></span>) – Sequence of gene sets, see <a class="reference internal" href="#scranpy.aggregate_across_genes" title="scranpy.aggregate_across_genes"><code class="xref py py-func docutils literal notranslate"><span class="pre">aggregate_across_genes()</span></code></a> for details.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Passed to <a class="reference internal" href="#scranpy.aggregate_across_genes" title="scranpy.aggregate_across_genes"><code class="xref py py-func docutils literal notranslate"><span class="pre">aggregate_across_genes()</span></code></a>.</p></li>
<li><p><strong>more_aggr_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Further arguments to pass to <a class="reference internal" href="#scranpy.aggregate_across_genes" title="scranpy.aggregate_across_genes"><code class="xref py py-func docutils literal notranslate"><span class="pre">aggregate_across_genes()</span></code></a>.</p></li>
<li><p><strong>assay_type</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>) – Name or index of the assay of <code class="docutils literal notranslate"><span class="pre">x</span></code> to be aggregated across genes.</p></li>
<li><p><strong>output_name</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – String specifying the assay name of the aggregated values in the output object.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, it defaults to <code class="docutils literal notranslate"><span class="pre">assay_type</span></code> if that argument is a string, otherwise it is set to <code class="docutils literal notranslate"><span class="pre">&quot;aggregated&quot;</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A <a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a> with number of rows equal to the number of gene sets.
The lone assay contains the aggregated values for each gene set for all cells.
The column data is the same as that of <code class="docutils literal notranslate"><span class="pre">x</span></code>.
If <code class="docutils literal notranslate"><span class="pre">sets</span></code> is named, the names are used as the row names of the output.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">get_test_rna_data_se</span><span class="p">(</span><span class="s2">&quot;norm&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sets</span> <span class="o">=</span> <span class="p">{</span> <span class="s2">&quot;foo&quot;</span><span class="p">:</span> <span class="p">[</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">10</span> <span class="p">],</span> <span class="s2">&quot;bar&quot;</span><span class="p">:</span> <span class="p">[</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">11</span><span class="p">,</span> <span class="mi">17</span><span class="p">,</span> <span class="mi">23</span> <span class="p">]</span> <span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aggregated</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">aggregate_across_genes_se</span><span class="p">(</span><span class="n">sce</span><span class="p">,</span> <span class="n">sets</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aggregated</span><span class="o">.</span><span class="n">get_assay</span><span class="p">(</span><span class="mi">0</span><span class="p">)[:,:</span><span class="mi">10</span><span class="p">]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.aggregate_column_data">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">aggregate_column_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">coldata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">index</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">number</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">only_simple</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">placeholder</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.aggregate_column_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Aggregate the column data from a <a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a> for groups of cells.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>coldata</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a></span>) – A <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a> containing the column data for a <code class="docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code>.
Each row should correspond to a cell.</p></li>
<li><p><strong>index</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a></span>) – Vector of length equal to the number of cells.
Each entry should be the index of the factor combination to which each cell in <code class="docutils literal notranslate"><span class="pre">coldata</span></code> was assigned,
e.g., the index vector produced by <a class="reference internal" href="#scranpy.combine_factors" title="scranpy.combine_factors"><code class="xref py py-func docutils literal notranslate"><span class="pre">combine_factors()</span></code></a>.</p></li>
<li><p><strong>number</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Total number of unique factor combinations.
All elements of <code class="docutils literal notranslate"><span class="pre">index</span></code> should be less than <code class="docutils literal notranslate"><span class="pre">number</span></code>.</p></li>
<li><p><strong>only_simple</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to skip a column of <code class="docutils literal notranslate"><span class="pre">coldata</span></code> that is not a list, NumPy array, <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a> or <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.Factor.Factor" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Factor</span></code></a>.</p></li>
<li><p><strong>placeholder</strong> – Placeholder value to store in the output column when a factor combination does not have a single unique value.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a> with number of rows equal to <code class="docutils literal notranslate"><span class="pre">number</span></code>.
Each “simple” column in <code class="docutils literal notranslate"><span class="pre">coldata</span></code> (i.e., list, NumPy array, <code class="docutils literal notranslate"><span class="pre">NamedList</span></code> or <code class="docutils literal notranslate"><span class="pre">Factor</span></code>) is represented by a column in the output <code class="docutils literal notranslate"><span class="pre">BiocFrame</span></code>.
In each column, the <code class="docutils literal notranslate"><span class="pre">j</span></code>-th entry is equal to the unique value of all rows where <code class="docutils literal notranslate"><span class="pre">index</span> <span class="pre">==</span> <span class="pre">j</span></code>, or <code class="docutils literal notranslate"><span class="pre">placeholder</span></code> if there is not exactly one unique value.
If <code class="docutils literal notranslate"><span class="pre">only_simple</span> <span class="pre">=</span> <span class="pre">False</span></code>, any non-simple columns of <code class="docutils literal notranslate"><span class="pre">coldata</span></code> are represented in the output <code class="docutils literal notranslate"><span class="pre">BiocFrame</span></code> by a list of <code class="docutils literal notranslate"><span class="pre">placeholder</span></code> values.
Otherwise, if <code class="docutils literal notranslate"><span class="pre">only_simple</span> <span class="pre">=</span> <span class="pre">True</span></code>, any non-simple columns of <code class="docutils literal notranslate"><span class="pre">coldata</span></code> are skipped.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">biocframe</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">biocframe</span><span class="o">.</span><span class="n">BiocFrame</span><span class="p">({</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="s2">&quot;X&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;a&quot;</span><span class="p">,</span> <span class="s2">&quot;a&quot;</span><span class="p">,</span> <span class="s2">&quot;b&quot;</span><span class="p">,</span> <span class="s2">&quot;b&quot;</span><span class="p">,</span> <span class="s2">&quot;c&quot;</span><span class="p">,</span> <span class="s2">&quot;c&quot;</span><span class="p">],</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="s2">&quot;Y&quot;</span><span class="p">:</span> <span class="p">[</span>  <span class="mi">1</span><span class="p">,</span>   <span class="mi">1</span><span class="p">,</span>   <span class="mi">1</span><span class="p">,</span>   <span class="mi">2</span><span class="p">,</span>   <span class="mi">2</span><span class="p">,</span>   <span class="mi">2</span><span class="p">],</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="s2">&quot;Z&quot;</span><span class="p">:</span> <span class="p">[</span><span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">scranpy</span><span class="o">.</span><span class="n">aggregate_column_data</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="mi">3</span><span class="p">))</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.analyze">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">analyze</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_analyze.html#analyze"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.analyze" title="Permalink to this definition">¶</a></dt>
<dd><p>Deprecated, use <a class="reference internal" href="#scranpy.analyze_se" title="scranpy.analyze_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">analyze_se()</span></code></a> instead.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a></span></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.analyze_se">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">analyze_se</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">x,</span> <span class="pre">rna_altexp=False,</span> <span class="pre">adt_altexp=None,</span> <span class="pre">crispr_altexp=None,</span> <span class="pre">rna_assay_type='counts',</span> <span class="pre">adt_assay_type='counts',</span> <span class="pre">crispr_assay_type='counts',</span> <span class="pre">block=None,</span> <span class="pre">block_name='block',</span> <span class="pre">rna_qc_subsets=[],</span> <span class="pre">rna_qc_output_prefix=None,</span> <span class="pre">more_rna_qc_args={},</span> <span class="pre">adt_qc_subsets=[],</span> <span class="pre">adt_qc_output_prefix=None,</span> <span class="pre">more_adt_qc_args={},</span> <span class="pre">crispr_qc_output_prefix=None,</span> <span class="pre">more_crispr_qc_args={},</span> <span class="pre">filter_cells=True,</span> <span class="pre">rna_norm_output_name='logcounts',</span> <span class="pre">more_rna_norm_args={},</span> <span class="pre">adt_norm_output_name='logcounts',</span> <span class="pre">more_adt_norm_args={},</span> <span class="pre">crispr_norm_output_name='logcounts',</span> <span class="pre">more_crispr_norm_args={},</span> <span class="pre">rna_hvg_output_prefix=None,</span> <span class="pre">more_rna_hvg_args={},</span> <span class="pre">rna_pca_output_name='PCA',</span> <span class="pre">more_rna_pca_args={},</span> <span class="pre">adt_pca_output_name='PCA',</span> <span class="pre">more_adt_pca_args={},</span> <span class="pre">use_rna_pcs=True,</span> <span class="pre">use_adt_pcs=True,</span> <span class="pre">scale_output_name='combined',</span> <span class="pre">more_scale_args={},</span> <span class="pre">mnn_output_name='MNN',</span> <span class="pre">more_mnn_args={},</span> <span class="pre">more_umap_args={},</span> <span class="pre">more_tsne_args={},</span> <span class="pre">cluster_graph_output_name='graph_cluster',</span> <span class="pre">more_build_graph_args={},</span> <span class="pre">more_cluster_graph_args={},</span> <span class="pre">more_neighbor_args={},</span> <span class="pre">kmeans_clusters=None,</span> <span class="pre">kmeans_clusters_output_name='kmeans_cluster',</span> <span class="pre">more_kmeans_args={},</span> <span class="pre">clusters_for_markers=['graph',</span> <span class="pre">'kmeans'],</span> <span class="pre">more_rna_marker_args={},</span> <span class="pre">more_adt_marker_args={},</span> <span class="pre">more_crispr_marker_args={},</span> <span class="pre">nn_parameters=&lt;knncolle.annoy.AnnoyParameters</span> <span class="pre">object&gt;,</span> <span class="pre">num_threads=3</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_se_analyze.html#analyze_se"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.analyze_se" title="Permalink to this definition">¶</a></dt>
<dd><p>Execute a simple single-cell analysis pipeline, starting from a count matrix and ending with clusters, visualizations and markers.
This is equivalent to:</p>
<ul class="simple">
<li><p>Running <a class="reference internal" href="#scranpy.quick_rna_qc_se" title="scranpy.quick_rna_qc_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">quick_rna_qc_se()</span></code></a>,
<a class="reference internal" href="#scranpy.quick_adt_qc_se" title="scranpy.quick_adt_qc_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">quick_adt_qc_se()</span></code></a>
and/or <a class="reference internal" href="#scranpy.quick_crispr_qc_se" title="scranpy.quick_crispr_qc_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">quick_crispr_qc_se()</span></code></a>,
for quality control.</p></li>
<li><p>Subsetting <code class="docutils literal notranslate"><span class="pre">x</span></code> to only retain the high-quality cells in all modalities.</p></li>
<li><p>Running <a class="reference internal" href="#scranpy.normalize_rna_counts_se" title="scranpy.normalize_rna_counts_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">normalize_rna_counts_se()</span></code></a>,
<a class="reference internal" href="#scranpy.normalize_adt_counts_se" title="scranpy.normalize_adt_counts_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">normalize_adt_counts_se()</span></code></a>
and/or <a class="reference internal" href="#scranpy.normalize_crispr_counts_se" title="scranpy.normalize_crispr_counts_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">normalize_crispr_counts_se()</span></code></a>,
for normalization.</p></li>
<li><p>Running <a class="reference internal" href="#scranpy.choose_rna_hvgs_se" title="scranpy.choose_rna_hvgs_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">choose_rna_hvgs_se()</span></code></a> to identify highly variable genes.</p></li>
<li><p>Running <a class="reference internal" href="#scranpy.run_pca_se" title="scranpy.run_pca_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_pca_se()</span></code></a> on the RNA and/or ADT data.</p></li>
<li><p>Running <a class="reference internal" href="#scranpy.scale_by_neighbors_se" title="scranpy.scale_by_neighbors_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">scale_by_neighbors_se()</span></code></a> if multiple modalities are present.</p></li>
<li><p>Running <a class="reference internal" href="#scranpy.correct_mnn_se" title="scranpy.correct_mnn_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">correct_mnn_se()</span></code></a> if multiple batches are present.</p></li>
<li><p>Running <a class="reference internal" href="#scranpy.run_all_neighbor_steps_se" title="scranpy.run_all_neighbor_steps_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_all_neighbor_steps_se()</span></code></a> to obtain t-SNE and UMAP coordinates, and to perform graph-based clustering.</p></li>
<li><p>Running <a class="reference internal" href="#scranpy.cluster_kmeans_se" title="scranpy.cluster_kmeans_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">cluster_kmeans_se()</span></code></a> to perform k-means clustering, if requested.</p></li>
<li><p>Running <a class="reference internal" href="#scranpy.score_markers_se" title="scranpy.score_markers_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">score_markers_se()</span></code></a> to compute markers for each modality based on one of the clusterings.</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a></span>) – A <code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code> object or one of its subclasses.
Rows correspond to genomic features (genes, ADTs, or CRISPR guides) and columns correspond to cells.</p></li>
<li><p><strong>rna_altexp</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/constants.html#None" title="(in Python v3.14)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]</span>) – Name or index of the alternative experiment of <code class="docutils literal notranslate"><span class="pre">x</span></code> containing the RNA data.
Alternatively <code class="docutils literal notranslate"><span class="pre">False</span></code>, in which case the main experiment is assumed to contain the RNA data.
Alternatively <code class="docutils literal notranslate"><span class="pre">None</span></code>, in which case it is assumed that no RNA data is available.</p></li>
<li><p><strong>adt_altexp</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/constants.html#None" title="(in Python v3.14)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]</span>) – Name or index of the alternative experiment of <code class="docutils literal notranslate"><span class="pre">x</span></code> containing the ADT data.
Alternatively <code class="docutils literal notranslate"><span class="pre">False</span></code>, in which case the main experiment is assumed to contain the ADT data.
Alternatively <code class="docutils literal notranslate"><span class="pre">None</span></code>, in which case it is assumed that no ADT data is available.</p></li>
<li><p><strong>crispr_altexp</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/constants.html#None" title="(in Python v3.14)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]</span>) – Name or index of the alternative experiment of <code class="docutils literal notranslate"><span class="pre">x</span></code> containing the CRISPR data.
Alternatively <code class="docutils literal notranslate"><span class="pre">False</span></code>, in which case the main experiment is assumed to contain the CRISPR data.
Alternatively <code class="docutils literal notranslate"><span class="pre">None</span></code>, in which case it is assumed that no CRISPR data is available.</p></li>
<li><p><strong>rna_assay_type</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>) – Name or index of the assay containing the RNA count data.
Only used if <code class="docutils literal notranslate"><span class="pre">rna_altexp</span></code> is not <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>adt_assay_type</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>) – Name or index of the assay containing the ADT count data.
Only used if <code class="docutils literal notranslate"><span class="pre">rna_altexp</span></code> is not <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>crispr_assay_type</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>) – Name or index of the assay containing the CRISPR count data.
Only used if <code class="docutils literal notranslate"><span class="pre">rna_altexp</span></code> is not <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Blocking factor, as a sequence containing the block of origin (e.g., batch, sample) for each cell in <code class="docutils literal notranslate"><span class="pre">x</span></code>.
Alternatively <code class="docutils literal notranslate"><span class="pre">None</span></code>, if all cells are from the same block.</p></li>
<li><p><strong>block_name</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – Name of the column in which to store the blocking factor in the column data of the output <code class="docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code>.
Only used if <code class="docutils literal notranslate"><span class="pre">block</span></code> is not <code class="docutils literal notranslate"><span class="pre">None</span></code>.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, the blocking factor is not stored in the output.</p></li>
<li><p><strong>rna_qc_subsets</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>, <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a>]</span>) – Passed to <a class="reference internal" href="#scranpy.quick_rna_qc_se" title="scranpy.quick_rna_qc_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">quick_rna_qc_se()</span></code></a> as the <code class="docutils literal notranslate"><span class="pre">subsets</span></code> argument.
Only used if <code class="docutils literal notranslate"><span class="pre">rna_altexp</span></code> is not <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>rna_qc_output_prefix</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – Passed to <a class="reference internal" href="#scranpy.quick_rna_qc_se" title="scranpy.quick_rna_qc_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">quick_rna_qc_se()</span></code></a> as the <code class="docutils literal notranslate"><span class="pre">output_prefix</span></code> argument.
Only used if <code class="docutils literal notranslate"><span class="pre">rna_altexp</span></code> is not <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>more_rna_qc_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Additional arguments to pass to <a class="reference internal" href="#scranpy.quick_rna_qc_se" title="scranpy.quick_rna_qc_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">quick_rna_qc_se()</span></code></a>.
Only used if <code class="docutils literal notranslate"><span class="pre">rna_altexp</span></code> is not <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>adt_qc_subsets</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>, <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a>]</span>) – Passed to <a class="reference internal" href="#scranpy.quick_adt_qc_se" title="scranpy.quick_adt_qc_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">quick_adt_qc_se()</span></code></a> as the <code class="docutils literal notranslate"><span class="pre">subsets</span></code> argument.
Only used if <code class="docutils literal notranslate"><span class="pre">adt_altexp</span></code> is not <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>adt_qc_output_prefix</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – Passed to <a class="reference internal" href="#scranpy.quick_adt_qc_se" title="scranpy.quick_adt_qc_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">quick_adt_qc_se()</span></code></a> as the <code class="docutils literal notranslate"><span class="pre">output_prefix</span></code> argument.
Only used if <code class="docutils literal notranslate"><span class="pre">adt_altexp</span></code> is not <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>more_adt_qc_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Additional arguments to pass to <a class="reference internal" href="#scranpy.quick_adt_qc_se" title="scranpy.quick_adt_qc_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">quick_adt_qc_se()</span></code></a>.
Only used if <code class="docutils literal notranslate"><span class="pre">adt_altexp</span></code> is not <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>crispr_qc_output_prefix</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – Passed to <a class="reference internal" href="#scranpy.quick_crispr_qc_se" title="scranpy.quick_crispr_qc_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">quick_crispr_qc_se()</span></code></a> as the <code class="docutils literal notranslate"><span class="pre">output_prefix</span></code> argument.
Only used if <code class="docutils literal notranslate"><span class="pre">crispr_altexp</span></code> is not <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>more_crispr_qc_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Additional arguments to pass to <a class="reference internal" href="#scranpy.quick_crispr_qc_se" title="scranpy.quick_crispr_qc_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">quick_crispr_qc_se()</span></code></a>.
Only used if <code class="docutils literal notranslate"><span class="pre">crispr_altexp</span></code> is not <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>filter_cells</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to filter <code class="docutils literal notranslate"><span class="pre">x</span></code> to only retain high-quality cells in all modalities.
If <code class="docutils literal notranslate"><span class="pre">False</span></code> QC metrics and thresholds are still computed but are not used to filter the count matrices.</p></li>
<li><p><strong>rna_norm_output_name</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></span>) – Passed to <a class="reference internal" href="#scranpy.normalize_rna_counts_se" title="scranpy.normalize_rna_counts_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">normalize_rna_counts_se()</span></code></a> as the <code class="docutils literal notranslate"><span class="pre">output_name</span></code> argument.
Only used if <code class="docutils literal notranslate"><span class="pre">rna_altexp</span></code> is not <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>more_rna_norm_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Additional arguments to pass to <a class="reference internal" href="#scranpy.normalize_rna_counts_se" title="scranpy.normalize_rna_counts_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">normalize_rna_counts_se()</span></code></a>.
Only used if <code class="docutils literal notranslate"><span class="pre">rna_altexp</span></code> is not <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>adt_norm_output_name</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></span>) – Passed to <a class="reference internal" href="#scranpy.normalize_adt_counts_se" title="scranpy.normalize_adt_counts_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">normalize_adt_counts_se()</span></code></a> as the <code class="docutils literal notranslate"><span class="pre">output_name</span></code> argument.
Only used if <code class="docutils literal notranslate"><span class="pre">adt_altexp</span></code> is not <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>more_adt_norm_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Additional arguments to pass to <a class="reference internal" href="#scranpy.normalize_adt_counts_se" title="scranpy.normalize_adt_counts_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">normalize_adt_counts_se()</span></code></a>.
Only used if <code class="docutils literal notranslate"><span class="pre">adt_altexp</span></code> is not <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>crispr_norm_output_name</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></span>) – Passed to <a class="reference internal" href="#scranpy.normalize_crispr_counts_se" title="scranpy.normalize_crispr_counts_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">normalize_crispr_counts_se()</span></code></a> as the <code class="docutils literal notranslate"><span class="pre">output_name</span></code> argument.
Only used if <code class="docutils literal notranslate"><span class="pre">crispr_altexp</span></code> is not <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>more_crispr_norm_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Additional arguments to pass to <a class="reference internal" href="#scranpy.normalize_crispr_counts_se" title="scranpy.normalize_crispr_counts_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">normalize_crispr_counts_se()</span></code></a>.
Only used if <code class="docutils literal notranslate"><span class="pre">crispr_altexp</span></code> is not <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>rna_hvg_output_prefix</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – Passed to <a class="reference internal" href="#scranpy.choose_rna_hvgs_se" title="scranpy.choose_rna_hvgs_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">choose_rna_hvgs_se()</span></code></a> as the <code class="docutils literal notranslate"><span class="pre">output_prefix</span></code> argument.
Only used if <code class="docutils literal notranslate"><span class="pre">rna_altexp</span></code> is not <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>more_rna_hvg_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Additional arguments to pass to <a class="reference internal" href="#scranpy.choose_rna_hvgs_se" title="scranpy.choose_rna_hvgs_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">choose_rna_hvgs_se()</span></code></a>.
Only used if <code class="docutils literal notranslate"><span class="pre">rna_altexp</span></code> is not <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>rna_pca_output_name</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></span>) – Passed to <a class="reference internal" href="#scranpy.run_pca_se" title="scranpy.run_pca_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_pca_se()</span></code></a> as the <code class="docutils literal notranslate"><span class="pre">output_name</span></code> argument.
Only used if <code class="docutils literal notranslate"><span class="pre">rna_altexp</span></code> is not <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>more_rna_pca_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Additional arguments to pass to <a class="reference internal" href="#scranpy.run_pca_se" title="scranpy.run_pca_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_pca_se()</span></code></a>.
Only used if <code class="docutils literal notranslate"><span class="pre">rna_altexp</span></code> is not <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>adt_pca_output_name</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></span>) – Passed to <a class="reference internal" href="#scranpy.run_pca_se" title="scranpy.run_pca_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_pca_se()</span></code></a> as the <code class="docutils literal notranslate"><span class="pre">output_name</span></code> argument.
Only used if <code class="docutils literal notranslate"><span class="pre">adt_altexp</span></code> is not <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>more_adt_pca_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Additional arguments to pass to <a class="reference internal" href="#scranpy.run_pca_se" title="scranpy.run_pca_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_pca_se()</span></code></a>.
Only used if <code class="docutils literal notranslate"><span class="pre">adt_altexp</span></code> is not <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>use_rna_pcs</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to use the RNA-derived PCs for downstream steps (i.e., clustering, visualization).
Only used if <code class="docutils literal notranslate"><span class="pre">rna_altexp</span></code> is not <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>use_adt_pcs</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to use the ADT-derived PCs for downstream steps (i.e., clustering, visualization).
Only used if <code class="docutils literal notranslate"><span class="pre">adt_altexp</span></code> is not <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>scale_output_name</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></span>) – Passed to <a class="reference internal" href="#scranpy.scale_by_neighbors_se" title="scranpy.scale_by_neighbors_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">scale_by_neighbors_se()</span></code></a> as the <code class="docutils literal notranslate"><span class="pre">output_name</span></code> argument.
Only used if multiple modalities are available and their corresponding <code class="docutils literal notranslate"><span class="pre">use_*_pcs</span></code> arguments are <code class="docutils literal notranslate"><span class="pre">True</span></code>.</p></li>
<li><p><strong>more_scale_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Additional arguments to pass to <a class="reference internal" href="#scranpy.scale_by_neighbors_se" title="scranpy.scale_by_neighbors_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">scale_by_neighbors_se()</span></code></a>.
Only used if multiple modalities are available and their corresponding <code class="docutils literal notranslate"><span class="pre">use_*_pcs</span></code> arguments are <code class="docutils literal notranslate"><span class="pre">True</span></code>.</p></li>
<li><p><strong>mnn_output_name</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></span>) – Passed to <a class="reference internal" href="#scranpy.correct_mnn_se" title="scranpy.correct_mnn_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">correct_mnn_se()</span></code></a> as the <code class="docutils literal notranslate"><span class="pre">output_name</span></code> argument.
Only used if <code class="docutils literal notranslate"><span class="pre">block</span></code> is not <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>more_mnn_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Additional arguments to pass to <a class="reference internal" href="#scranpy.correct_mnn_se" title="scranpy.correct_mnn_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">correct_mnn_se()</span></code></a>.
Only used if <code class="docutils literal notranslate"><span class="pre">block</span></code> is not <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>more_tsne_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Passed to <a class="reference internal" href="#scranpy.run_all_neighbor_steps_se" title="scranpy.run_all_neighbor_steps_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_all_neighbor_steps_se()</span></code></a>.</p></li>
<li><p><strong>more_umap_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Passed to <a class="reference internal" href="#scranpy.run_all_neighbor_steps_se" title="scranpy.run_all_neighbor_steps_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_all_neighbor_steps_se()</span></code></a>.</p></li>
<li><p><strong>more_build_graph_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Passed to <a class="reference internal" href="#scranpy.run_all_neighbor_steps_se" title="scranpy.run_all_neighbor_steps_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_all_neighbor_steps_se()</span></code></a>.</p></li>
<li><p><strong>cluster_graph_output_name</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></span>) – Passed to <a class="reference internal" href="#scranpy.run_all_neighbor_steps_se" title="scranpy.run_all_neighbor_steps_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_all_neighbor_steps_se()</span></code></a> as <code class="docutils literal notranslate"><span class="pre">cluster_output_name</span></code>.</p></li>
<li><p><strong>more_cluster_graph_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Passed to <a class="reference internal" href="#scranpy.run_all_neighbor_steps_se" title="scranpy.run_all_neighbor_steps_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_all_neighbor_steps_se()</span></code></a>.</p></li>
<li><p><strong>more_neighbor_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Passed to <a class="reference internal" href="#scranpy.run_all_neighbor_steps_se" title="scranpy.run_all_neighbor_steps_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_all_neighbor_steps_se()</span></code></a>.</p></li>
<li><p><strong>kmeans_clusters</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>) – Passed to <a class="reference internal" href="#scranpy.cluster_kmeans_se" title="scranpy.cluster_kmeans_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">cluster_kmeans_se()</span></code></a> as the <code class="docutils literal notranslate"><span class="pre">k</span></code> argument.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, k-means clustering is not performed.</p></li>
<li><p><strong>kmeans_clusters_output_name</strong> – Passed to <a class="reference internal" href="#scranpy.cluster_kmeans_se" title="scranpy.cluster_kmeans_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">cluster_kmeans_se()</span></code></a> as the <code class="docutils literal notranslate"><span class="pre">output_name</span></code> argument.
Ignored if <code class="docutils literal notranslate"><span class="pre">kmeans_clusters</span></code> is <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>more_kmeans_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Additional arguments to pass to <a class="reference internal" href="#scranpy.cluster_kmeans_se" title="scranpy.cluster_kmeans_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">cluster_kmeans_se()</span></code></a>.
Ignored if <code class="docutils literal notranslate"><span class="pre">kmeans_clusters</span></code> is <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>clusters_for_markers</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Literal" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'graph'</span></code>, <code class="docutils literal notranslate"><span class="pre">'kmeans'</span></code>]]</span>) – List of clustering algorithms, specifying the clusters to be used for marker detection.
The first available clustering will be chosen.
If no clustering is available from the list, markers will not be computed.</p></li>
<li><p><strong>more_rna_marker_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Additional arguments to pass to <a class="reference internal" href="#scranpy.score_markers_se" title="scranpy.score_markers_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">score_markers_se()</span></code></a> for the RNA data.
Ignored if no suitable clusterings are available or if <code class="docutils literal notranslate"><span class="pre">rna_altexp</span></code> is <code class="docutils literal notranslate"><span class="pre">None</span></code></p></li>
<li><p><strong>more_adt_marker_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Additional arguments to pass to <a class="reference internal" href="#scranpy.score_markers_se" title="scranpy.score_markers_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">score_markers_se()</span></code></a> for the ADT data.
Ignored if no suitable clusterings are available or if <code class="docutils literal notranslate"><span class="pre">adt_altexp</span></code> is <code class="docutils literal notranslate"><span class="pre">None</span></code></p></li>
<li><p><strong>more_crispr_marker_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Additional arguments to pass to <a class="reference internal" href="#scranpy.score_markers_se" title="scranpy.score_markers_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">score_markers_se()</span></code></a> for the CRISPR data.
Ignored if no suitable clusterings are available or if <code class="docutils literal notranslate"><span class="pre">crispr_altexp</span></code> is <code class="docutils literal notranslate"><span class="pre">None</span></code></p></li>
<li><p><strong>nn_parameters</strong> – Parameters for the nearest-neighbor search.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use in each step.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">x</span></code>: a <a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a> that is a copy of the input <code class="docutils literal notranslate"><span class="pre">x</span></code>.
It is also decorated with the results of each analysis step.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">markers</span></code>: a list of list of <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a> objects containing the marker statistics for each modality.
Each inner list corresponds to a modality (RNA, ADT, etc.) while each <code class="docutils literal notranslate"><span class="pre">BiocFrame</span></code> corresponds to a group.</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>A <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a> containing</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">get_test_rna_data_se</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">is_mito</span> <span class="o">=</span> <span class="p">[</span> <span class="n">s</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;mt-&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">sce</span><span class="o">.</span><span class="n">get_row_names</span><span class="p">()</span> <span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">analyze_se</span><span class="p">(</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">sce</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">rna_qc_subsets</span><span class="o">=</span><span class="p">{</span> <span class="s2">&quot;mito&quot;</span><span class="p">:</span> <span class="n">is_mito</span> <span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res</span><span class="p">[</span><span class="s2">&quot;x&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">get_assay_names</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res</span><span class="p">[</span><span class="s2">&quot;x&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">get_reduced_dimension_names</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="s2">&quot;x&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">biocutils</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">biocutils</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="s2">&quot;x&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">()[</span><span class="s2">&quot;graph_cluster&quot;</span><span class="p">]))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">scranpy</span><span class="o">.</span><span class="n">preview_markers</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="s2">&quot;markers&quot;</span><span class="p">][</span><span class="s2">&quot;rna&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]))</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.build_snn_graph">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">build_snn_graph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_neighbors=10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weight_scheme='ranked'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads=1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nn_parameters=&lt;knncolle.annoy.AnnoyParameters</span> <span class="pre">object&gt;</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.build_snn_graph" title="Permalink to this definition">¶</a></dt>
<dd><p>Build a shared nearest neighbor (SNN) graph where each node is a cell.
Edges are formed between cells that share one or more nearest neighbors, weighted by the number or importance of those shared neighbors.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a>, <a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.find_knn.FindKnnResults" title="(in knncolle v0.0.post1.dev1+g951fea368)"><code class="xref py py-class docutils literal notranslate"><span class="pre">FindKnnResults</span></code></a>, <a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.classes.Index" title="(in knncolle v0.0.post1.dev1+g951fea368)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Index</span></code></a>]</span>) – <p>Numeric matrix where rows are dimensions and columns are cells,
typically containing a low-dimensional representation from, e.g., <a class="reference internal" href="#scranpy.run_pca" title="scranpy.run_pca"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_pca()</span></code></a>.</p>
<p>Alternatively, a <a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.find_knn.FindKnnResults" title="(in knncolle v0.0.post1.dev1+g951fea368)"><code class="xref py py-class docutils literal notranslate"><span class="pre">FindKnnResults</span></code></a> object containing existing neighbor search results.</p>
<p>Alternatively, a <a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.classes.Index" title="(in knncolle v0.0.post1.dev1+g951fea368)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Index</span></code></a> object.</p>
</p></li>
<li><p><strong>num_neighbors</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of neighbors in the nearest-neighbor graph.
Larger values generally result in broader clusters during community detection.
Ignored if <code class="docutils literal notranslate"><span class="pre">x</span></code> is a <a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.find_knn.FindKnnResults" title="(in knncolle v0.0.post1.dev1+g951fea368)"><code class="xref py py-class docutils literal notranslate"><span class="pre">FindKnnResults</span></code></a> object.</p></li>
<li><p><strong>weight_scheme</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Literal" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'ranked'</span></code>, <code class="docutils literal notranslate"><span class="pre">'number'</span></code>, <code class="docutils literal notranslate"><span class="pre">'jaccard'</span></code>]</span>) – Weighting scheme to use for the edges of the SNN graph, based on the number or ranking of the shared nearest neighbors.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.</p></li>
<li><p><strong>nn_parameters</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.classes.Parameters" title="(in knncolle v0.0.post1.dev1+g951fea368)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Parameters</span></code></a></span>) – The algorithm to use for the nearest-neighbor search.
Only used if <code class="docutils literal notranslate"><span class="pre">x</span></code> is not a pre-built nearest-neighbor search index or a list of existing nearest-neighbor search results.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><p>A <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a> containing the components of a (possibly weighted) graph.</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">vertices</span></code>: integer specifying the number of vertices (i.e., cells) in the graph.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">edges</span></code>: integer NumPy array containing the graph edges.
Pairs of values represent the endpoints of an (undirected) edge, i.e., <code class="docutils literal notranslate"><span class="pre">edges[0:2]</span></code> form the first edge, <code class="docutils literal notranslate"><span class="pre">edges[2:4]</span></code> form the second edge and so on.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">weights</span></code>: double-precision NumPy array of edge weights.
This has length equal to half the length of <code class="docutils literal notranslate"><span class="pre">edges</span></code>; the first weight corresponds to the first edge, and so on.</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The <code class="docutils literal notranslate"><span class="pre">build_snn_graph</span></code> function in the <a class="reference external" href="https://libscran.github.io/scran_graph_cluster">scran_graph_cluster</a> C++ library.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pcs</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">200</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">graph</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">build_snn_graph</span><span class="p">(</span><span class="n">pcs</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">graph</span><span class="o">.</span><span class="n">get_names</span><span class="p">())</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.center_size_factors">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">center_size_factors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">size_factors</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'lowest'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">in_place</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.center_size_factors" title="Permalink to this definition">¶</a></dt>
<dd><p>Center size factors before computing normalized values from the count matrix.
This ensures that the normalized values are on the same scale as the original counts for easier interpretation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>size_factors</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span>) – Floating-point array containing size factors for all cells.</p></li>
<li><p><strong>block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Block assignment for each cell.
If provided, this should have length equal to the number of cells, where cells have the same value if and only if they are in the same block.
Defaults to <code class="docutils literal notranslate"><span class="pre">None</span></code>, where all cells are treated as being part of the same block.</p></li>
<li><p><strong>mode</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Literal" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'lowest'</span></code>, <code class="docutils literal notranslate"><span class="pre">'per-block'</span></code>]</span>) – How to scale size factors across blocks.
<code class="docutils literal notranslate"><span class="pre">lowest</span></code> will scale all size factors by the lowest per-block average.
<code class="docutils literal notranslate"><span class="pre">per-block</span></code> will center the size factors in each block separately.
This argument is only used if <code class="docutils literal notranslate"><span class="pre">block</span></code> is provided.</p></li>
<li><p><strong>in_place</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to modify <code class="docutils literal notranslate"><span class="pre">size_factors</span></code> in place.
If <code class="docutils literal notranslate"><span class="pre">False</span></code>, a new array is returned.
This argument only used if <code class="docutils literal notranslate"><span class="pre">size_factors</span></code> is double-precision, otherwise a new array is always returned.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Double-precision NumPy array containing centered size factors.
If <code class="docutils literal notranslate"><span class="pre">in_place</span> <span class="pre">=</span> <span class="pre">True</span></code>, this is a reference to <code class="docutils literal notranslate"><span class="pre">size_factors</span></code>.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The <code class="docutils literal notranslate"><span class="pre">center_size_factors</span></code> and <code class="docutils literal notranslate"><span class="pre">center_size_factors_blocked</span></code> functions in the <a class="reference external" href="https://libscran.github.io/scran_norm">scran_norm</a> C++ library,
which describes the rationale behind centering.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sf</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sf2</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">center_size_factors</span><span class="p">(</span><span class="n">sf</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sf2</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.choose_highly_variable_genes">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">choose_highly_variable_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">stats</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">top</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">larger</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_ties</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_choose_highly_variable_genes.html#choose_highly_variable_genes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.choose_highly_variable_genes" title="Permalink to this definition">¶</a></dt>
<dd><p>Choose highly variable genes (HVGs), typically based on a variance-related statistic.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>stats</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span>) – Array of variances (or a related statistic) across all genes.
Typically the residuals from <a class="reference internal" href="#scranpy.model_gene_variances" title="scranpy.model_gene_variances"><code class="xref py py-func docutils literal notranslate"><span class="pre">model_gene_variances()</span></code></a> used here.</p></li>
<li><p><strong>top</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of top genes to retain.
Note that the actual number of retained genes may not be equal to <code class="docutils literal notranslate"><span class="pre">top</span></code>, depending on the other options.</p></li>
<li><p><strong>larger</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether larger values of <code class="docutils literal notranslate"><span class="pre">stats</span></code> represent more variable genes.
If true, HVGs are defined from the largest values of <code class="docutils literal notranslate"><span class="pre">stats</span></code>.</p></li>
<li><p><strong>keep_ties</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to keep ties at the <code class="docutils literal notranslate"><span class="pre">top</span></code>-th most variable gene.
This avoids arbitrary breaking of tied values.</p></li>
<li><p><strong>bound</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a>]</span>) – The lower bound (if <code class="docutils literal notranslate"><span class="pre">larger</span> <span class="pre">=</span> <span class="pre">True</span></code>) or upper bound (otherwise) to be applied to <code class="docutils literal notranslate"><span class="pre">stats</span></code>.
Genes are not considered to be HVGs if they do not pass this bound, even if they are within the <code class="docutils literal notranslate"><span class="pre">top</span></code> genes.
Ignored if <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Integer NumPy array containing the indices of genes in <code class="docutils literal notranslate"><span class="pre">stats</span></code> that are considered to be highly variable.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The <code class="docutils literal notranslate"><span class="pre">choose_highly_variable_genes</span></code> function from the <a class="reference external" href="https://libscran.github.io/scran_variances">scran_variances</a> library,
which provides the underlying implementation.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">resids</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="mi">1000</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">chosen</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">choose_highly_variable_genes</span><span class="p">(</span><span class="n">resids</span><span class="p">,</span> <span class="n">top</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">chosen</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.choose_pseudo_count">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">choose_pseudo_count</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">size_factors</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quantile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_bias</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_value</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_choose_pseudo_count.html#choose_pseudo_count"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.choose_pseudo_count" title="Permalink to this definition">¶</a></dt>
<dd><p>Choose a suitable pseudo-count to control the bias introduced by log-transformation of normalized counts.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>size_factors</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span>) – Floating-point array of size factors for all cells.</p></li>
<li><p><strong>quantile</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Quantile to use for defining extreme size factors.</p></li>
<li><p><strong>max_bias</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Maximum allowed bias in the log-fold changes between cells.</p></li>
<li><p><strong>min_value</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Minimum value for the pseudo-count.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Choice of pseudo-count, for use in <a class="reference internal" href="#scranpy.normalize_counts" title="scranpy.normalize_counts"><code class="xref py py-func docutils literal notranslate"><span class="pre">normalize_counts()</span></code></a>.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The <code class="docutils literal notranslate"><span class="pre">choose_pseudo_count</span></code> function in the <a class="reference external" href="https://libscran.github.io/scran_norm">scran_norm</a> C++ library, which describes the rationale behind the choice of pseudo-count.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sf</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="mi">1000</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sf</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">center_size_factors</span><span class="p">(</span><span class="n">sf</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pseudo</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">choose_pseudo_count</span><span class="p">(</span><span class="n">sf</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">pseudo</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.choose_rna_hvgs_se">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">choose_rna_hvgs_se</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">more_var_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">top</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">more_choose_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assay_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'logcounts'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_per_block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.choose_rna_hvgs_se" title="Permalink to this definition">¶</a></dt>
<dd><p>Model the mean-variance relationship across genes and choose highly variable genes (HVGs) based on the residuals of the fitted trend.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a></span>) – A <a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a> object or any of its subclasses.
Rows correspond to genes and columns correspond to cells.</p></li>
<li><p><strong>block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Block assignment for each cell in <code class="docutils literal notranslate"><span class="pre">x</span></code>, passed to <a class="reference internal" href="#scranpy.model_gene_variances" title="scranpy.model_gene_variances"><code class="xref py py-func docutils literal notranslate"><span class="pre">model_gene_variances()</span></code></a>.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads, passed to <a class="reference internal" href="#scranpy.model_gene_variances" title="scranpy.model_gene_variances"><code class="xref py py-func docutils literal notranslate"><span class="pre">model_gene_variances()</span></code></a>.</p></li>
<li><p><strong>more_var_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Additional arguments to pass to <a class="reference internal" href="#scranpy.model_gene_variances" title="scranpy.model_gene_variances"><code class="xref py py-func docutils literal notranslate"><span class="pre">model_gene_variances()</span></code></a>.</p></li>
<li><p><strong>top</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of top genes to use as HVGs, passed to <a class="reference internal" href="#scranpy.choose_highly_variable_genes" title="scranpy.choose_highly_variable_genes"><code class="xref py py-func docutils literal notranslate"><span class="pre">choose_highly_variable_genes()</span></code></a>.</p></li>
<li><p><strong>more_choose_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Additional arguments to pass to <a class="reference internal" href="#scranpy.choose_highly_variable_genes" title="scranpy.choose_highly_variable_genes"><code class="xref py py-func docutils literal notranslate"><span class="pre">choose_highly_variable_genes()</span></code></a>.</p></li>
<li><p><strong>assay_type</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>) – Name or index of the assay of <code class="docutils literal notranslate"><span class="pre">x</span></code> containing the log-expression data to use for computing variances.</p></li>
<li><p><strong>output_prefix</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – Prefix to add to the names of the columns of the row data in which to store the output statistics.</p></li>
<li><p><strong>include_per_block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether the per-block statistics should be stored in the row data of the output object.
Only relevant if <code class="docutils literal notranslate"><span class="pre">block</span></code> is specified.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A copy of <code class="docutils literal notranslate"><span class="pre">x</span></code>, with per-gene variance modelling statistics added to its row data.
An <code class="docutils literal notranslate"><span class="pre">hvg</span></code> column indicates the genes that were chosen as HVGs.
If <code class="docutils literal notranslate"><span class="pre">include_per_block</span> <span class="pre">=</span> <span class="pre">True</span></code> and <code class="docutils literal notranslate"><span class="pre">block</span></code> is specified,
the per-block statistics are stored as a nested <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a> in the <code class="docutils literal notranslate"><span class="pre">per_block</span></code> column.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">get_test_rna_data_se</span><span class="p">(</span><span class="s2">&quot;norm&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">choose_rna_hvgs_se</span><span class="p">(</span><span class="n">sce</span><span class="p">,</span> <span class="n">more_var_args</span><span class="o">=</span><span class="p">{</span> <span class="s2">&quot;use_min_width&quot;</span><span class="p">:</span> <span class="kc">True</span> <span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">biocutils</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">biocutils</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">sce</span><span class="o">.</span><span class="n">get_row_data</span><span class="p">()[</span><span class="s2">&quot;hvg&quot;</span><span class="p">]))</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.cluster_graph">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">cluster_graph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'multilevel'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">multilevel_resolution</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">leiden_resolution</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">leiden_objective</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'modularity'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">walktrap_steps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">42</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_cluster_graph.html#cluster_graph"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.cluster_graph" title="Permalink to this definition">¶</a></dt>
<dd><p>Identify clusters of cells using a variety of community detection methods from a graph where similar cells are connected.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a></span>) – Components of the graph to be clustered, typically produced by <code class="xref py py-func docutils literal notranslate"><span class="pre">build_snn_graph()</span></code>.
Each node of the graph should be a cell.</p></li>
<li><p><strong>method</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Literal" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'multilevel'</span></code>, <code class="docutils literal notranslate"><span class="pre">'leiden'</span></code>, <code class="docutils literal notranslate"><span class="pre">'walktrap'</span></code>]</span>) – Community detection algorithm to use.</p></li>
<li><p><strong>multilevel_resolution</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Resolution of the clustering when <code class="docutils literal notranslate"><span class="pre">method</span> <span class="pre">=</span> <span class="pre">&quot;multilevel&quot;</span></code>.
Larger values result in finer clusters.</p></li>
<li><p><strong>leiden_resolution</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Resolution of the clustering when <code class="docutils literal notranslate"><span class="pre">method</span> <span class="pre">=</span> <span class="pre">&quot;leiden&quot;</span></code>.
Larger values result in finer clusters.</p></li>
<li><p><strong>leiden_objective</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Literal" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'modularity'</span></code>, <code class="docutils literal notranslate"><span class="pre">'cpm'</span></code>, <code class="docutils literal notranslate"><span class="pre">'er'</span></code>]</span>) – Objective function to use when <code class="docutils literal notranslate"><span class="pre">method</span> <span class="pre">=</span> <span class="pre">&quot;leiden&quot;</span></code>.</p></li>
<li><p><strong>walktrap_steps</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of steps to use when <code class="docutils literal notranslate"><span class="pre">method</span> <span class="pre">=</span> <span class="pre">&quot;walktrap&quot;</span></code>.</p></li>
<li><p><strong>seed</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Random seed to use for <code class="docutils literal notranslate"><span class="pre">method</span> <span class="pre">=</span> <span class="pre">&quot;multilevel&quot;</span></code> or <code class="docutils literal notranslate"><span class="pre">&quot;leiden&quot;</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><p>A <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a> containing the following entries.</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">membership</span></code>: an integer NumPy array containing the cluster assignment for each vertex, i.e., cell.
All values are in [0, N) where N is the total number of clusters.</p></li>
</ul>
<p>For <code class="docutils literal notranslate"><span class="pre">method</span> <span class="pre">=</span> <span class="pre">&quot;multilevel&quot;</span></code>, the output also contains:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">levels</span></code>: a list containing the clustering at each level of the algorithm.
Each entry corresponds to one level and is an integer NumPy array that contains the cluster assignment for each cell at that level.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">modularity</span></code>: a double-precision NumPy array containing the modularity at each level.
This has length equal to that of <code class="docutils literal notranslate"><span class="pre">levels</span></code>, and the largest value corresponds to the assignments reported in <code class="docutils literal notranslate"><span class="pre">membership</span></code>.</p></li>
</ul>
<p>For <code class="docutils literal notranslate"><span class="pre">method</span> <span class="pre">=</span> <span class="pre">&quot;walktrap&quot;</span></code>, the output also contains:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">merges</span></code>: an integer NumPy matrix with two columns.
Each row  corresponds to a merge step and specifies the pair of cells or clusters that were merged at that step.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">modularity</span></code>: a double-precision NumPy array that contains the modularity score at each merge step.</p></li>
</ul>
<p>For <code class="docutils literal notranslate"><span class="pre">method</span> <span class="pre">=</span> <span class="pre">&quot;leiden&quot;</span></code>, the output also contains:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">quality</span></code>: quality of the clustering.
This is the same as the modularity if <code class="docutils literal notranslate"><span class="pre">leiden_objective</span> <span class="pre">=</span> <span class="pre">&quot;modularity&quot;</span></code>.</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">References</p>
<p><a class="reference external" href="https://igraph.org/c/html/latest/igraph-Community.html">https://igraph.org/c/html/latest/igraph-Community.html</a>, for the underlying implementation of each clustering method.</p>
<p>The various <code class="docutils literal notranslate"><span class="pre">cluster_*</span></code> functions in the <a class="reference external" href="https://libscran.github.io/scran_graph_cluster">scran_graph_cluster</a> C++ library.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pcs</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">200</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">graph</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">build_snn_graph</span><span class="p">(</span><span class="n">pcs</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">clust</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">cluster_graph</span><span class="p">(</span><span class="n">graph</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">biocutils</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">biocutils</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">clust</span><span class="p">[</span><span class="s2">&quot;membership&quot;</span><span class="p">]))</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.cluster_graph_se">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">cluster_graph_se</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_neighbors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">more_build_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'multilevel'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">resolution</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">more_cluster_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reddim_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'PCA'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'clusters'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">meta_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_se_cluster_graph.html#cluster_graph_se"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.cluster_graph_se" title="Permalink to this definition">¶</a></dt>
<dd><p>Construct a shared-nearest neighbor (SNN) graph from an existing low-dimensional embedding,
and apply community detection algorithms to obtain clusters of cells.
This calls <a class="reference internal" href="#scranpy.build_snn_graph" title="scranpy.build_snn_graph"><code class="xref py py-func docutils literal notranslate"><span class="pre">build_snn_graph()</span></code></a> followed by <a class="reference internal" href="#scranpy.cluster_graph" title="scranpy.cluster_graph"><code class="xref py py-func docutils literal notranslate"><span class="pre">cluster_graph()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a></span>) – A <a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a> or one of its subclasses.
Rows correspond to genomic features and columns correspond to cells.</p></li>
<li><p><strong>num_neighbors</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of neighbors for constructing the SNN graph, see <a class="reference internal" href="#scranpy.build_snn_graph" title="scranpy.build_snn_graph"><code class="xref py py-func docutils literal notranslate"><span class="pre">build_snn_graph()</span></code></a> for details.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads for the neighbor search when constructing the SNN graph, see <a class="reference internal" href="#scranpy.build_snn_graph" title="scranpy.build_snn_graph"><code class="xref py py-func docutils literal notranslate"><span class="pre">build_snn_graph()</span></code></a> for details.</p></li>
<li><p><strong>more_build_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Additional arguments to be passed to <a class="reference internal" href="#scranpy.build_snn_graph" title="scranpy.build_snn_graph"><code class="xref py py-func docutils literal notranslate"><span class="pre">build_snn_graph()</span></code></a>.</p></li>
<li><p><strong>method</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Literal" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'multilevel'</span></code>, <code class="docutils literal notranslate"><span class="pre">'leiden'</span></code>, <code class="docutils literal notranslate"><span class="pre">'walktrap'</span></code>]</span>) – Community detection method to be used by <a class="reference internal" href="#scranpy.cluster_graph" title="scranpy.cluster_graph"><code class="xref py py-func docutils literal notranslate"><span class="pre">cluster_graph()</span></code></a>.</p></li>
<li><p><strong>resolution</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a>]</span>) – Clustering resolution to be used by <a class="reference internal" href="#scranpy.cluster_graph" title="scranpy.cluster_graph"><code class="xref py py-func docutils literal notranslate"><span class="pre">cluster_graph()</span></code></a>.
This is either passed as <code class="docutils literal notranslate"><span class="pre">multilevel_resolution</span></code> or <code class="docutils literal notranslate"><span class="pre">leiden_resolution</span></code>, depending on <code class="docutils literal notranslate"><span class="pre">method</span></code>.</p></li>
<li><p><strong>more_cluster_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Additional arguments to be passed to <a class="reference internal" href="#scranpy.cluster_graph" title="scranpy.cluster_graph"><code class="xref py py-func docutils literal notranslate"><span class="pre">cluster_graph()</span></code></a>.</p></li>
<li><p><strong>reddim_type</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#tuple" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tuple</span></code></a>]</span>) – Name or index of the existing reduced dimension of <code class="docutils literal notranslate"><span class="pre">x</span></code> to be used for clustering.
Alternatively a tuple, where the first element contains the name of an alternative experiment of <code class="docutils literal notranslate"><span class="pre">x</span></code>,
and the second element contains the name/index of an embedding in that alternative experiment.</p></li>
<li><p><strong>output_name</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></span>) – Name of the column of the column data in which to store the cluster assignments.</p></li>
<li><p><strong>meta_name</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – Name of the metadta entry in which to store extra clustering output.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, no extra clustering output is stored.</p></li>
<li><p><strong>graph_name</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – Name of the metadta entry in which to store the SNN graph.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, the graph is not stored.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><code class="docutils literal notranslate"><span class="pre">x</span></code> is returned with the cluster assignment for each cell stored in its column data.
Additional clustering output is stored in its metadata.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">get_test_rna_data_se</span><span class="p">(</span><span class="s2">&quot;pca&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">cluster_graph_se</span><span class="p">(</span><span class="n">sce</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">biocutils</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">biocutils</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">sce</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">()[</span><span class="s2">&quot;clusters&quot;</span><span class="p">]))</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.cluster_kmeans">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">cluster_kmeans</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">k</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">init_method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'var-part'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">refine_method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'hartigan-wong'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">var_part_optimize_partition</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">var_part_size_adjustment</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lloyd_iterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hartigan_wong_iterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hartigan_wong_quick_transfer_iterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hartigan_wong_quit_quick_transfer_failure</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5489</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_cluster_kmeans.html#cluster_kmeans"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.cluster_kmeans" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform k-means clustering with a variety of different initialization and refinement algorithms.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span>) – Input data matrix where rows are dimensions and columns are observations (i.e., cells).</p></li>
<li><p><strong>k</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of clusters.</p></li>
<li><p><strong>init_method</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Literal" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'var-part'</span></code>, <code class="docutils literal notranslate"><span class="pre">'kmeans++'</span></code>, <code class="docutils literal notranslate"><span class="pre">'random'</span></code>]</span>) – Initialization method for defining the initial centroid coordinates.
Choices are variance partitioning (<code class="docutils literal notranslate"><span class="pre">var-part</span></code>), kmeans++ (<code class="docutils literal notranslate"><span class="pre">kmeans++</span></code>) or random initialization (<code class="docutils literal notranslate"><span class="pre">random</span></code>).</p></li>
<li><p><strong>refine_method</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Literal" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'hartigan-wong'</span></code>, <code class="docutils literal notranslate"><span class="pre">'lloyd'</span></code>]</span>) – Method to use to refine the cluster assignments and centroid coordinates.
Choices are Lloyd’s algorithm (<code class="docutils literal notranslate"><span class="pre">lloyd</span></code>) or the Hartigan-Wong algorithm (<code class="docutils literal notranslate"><span class="pre">hartigan-wong</span></code>).</p></li>
<li><p><strong>var_part_optimize_partition</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether each partition boundary should be optimized to reduce the sum of squares in the child partitions.
Only used if <code class="docutils literal notranslate"><span class="pre">init_method</span> <span class="pre">=</span> <span class="pre">&quot;var-part&quot;</span></code>.</p></li>
<li><p><strong>var_part_size_adjustment</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Floating-point value between 0 and 1, specifying the adjustment to the cluster size when prioritizing the next cluster to partition.
Setting this to 0 will ignore the cluster size while setting this to 1 will generally favor larger clusters.
Only used if <code class="docutils literal notranslate"><span class="pre">init_method</span> <span class="pre">=</span> <span class="pre">&quot;var-part&quot;</span></code>.</p></li>
<li><p><strong>lloyd_iterations</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Maximmum number of iterations for the Lloyd algorithm.</p></li>
<li><p><strong>hartigan_wong_iterations</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Maximmum number of iterations for the Hartigan-Wong algorithm.</p></li>
<li><p><strong>hartigan_wong_quick_transfer_iterations</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Maximmum number of quick transfer iterations for the Hartigan-Wong algorithm.</p></li>
<li><p><strong>hartigan_wong_quit_quick_transfer_failure</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to quit the Hartigan-Wong algorithm upon convergence failure during quick transfer iterations.</p></li>
<li><p><strong>seed</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Seed to use for random or kmeans++ initialization.</p></li>
<li><p><strong>num.threads</strong> – Number of threads to use.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><p>A <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a> containing the following entries.</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">clusters</span></code>: an integer NumPy array containing the cluster assignment for each cell.
Values are integers in [0, N) where N is the total number of clusters.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">centers</span></code>: a double-precision NumPy matrix containing the coordinates of the cluster centroids.
Dimensions are in the rows while centers are in the columns.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">iterations</span></code>: integer specifying the number of refinement iterations that were performed.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">status</span></code>: convergence status.
Any non-zero value indicates a convergence failure though the exact meaning depends on the choice of <code class="docutils literal notranslate"><span class="pre">refine_method</span></code>.</p>
<ul>
<li><p>For Lloyd, a value of 2 indicates convergence failure.</p></li>
<li><p>For Hartigan-Wong, a value of 2 indicates convergence failure in the optimal transfer iterations.
A value of 4 indicates convergence failure in the quick transfer iterations when <code class="docutils literal notranslate"><span class="pre">hartigan_wong_quit_quick_transfer_failure</span> <span class="pre">=</span> <span class="pre">True</span></code>.</p></li>
</ul>
</li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">References</p>
<p><a class="reference external" href="https://ltla.github.io/CppKmeans">https://ltla.github.io/CppKmeans</a>, which describes the various initialization and refinement algorithms in more detail.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pcs</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">200</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">clust</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">cluster_kmeans</span><span class="p">(</span><span class="n">pcs</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">biocutils</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">biocutils</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">clust</span><span class="p">[</span><span class="s2">&quot;clusters&quot;</span><span class="p">]))</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.cluster_kmeans_se">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">cluster_kmeans_se</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">k</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">more_kmeans_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reddim_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'PCA'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'clusters'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">meta_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_se_cluster_kmeans.html#cluster_kmeans_se"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.cluster_kmeans_se" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform k-means clustering on an existing low-dimensional embedding.
This calls <a class="reference internal" href="#scranpy.cluster_kmeans" title="scranpy.cluster_kmeans"><code class="xref py py-func docutils literal notranslate"><span class="pre">cluster_kmeans()</span></code></a> on reduced dimensions from a <a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a></span>) – A <a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a> object or one of its subclasses.
Rows correspond to genomic features and columns correspond to cells.</p></li>
<li><p><strong>k</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of clusters, to be passed to <a class="reference internal" href="#scranpy.cluster_kmeans" title="scranpy.cluster_kmeans"><code class="xref py py-func docutils literal notranslate"><span class="pre">cluster_kmeans()</span></code></a>.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads, to be passed to <a class="reference internal" href="#scranpy.cluster_kmeans" title="scranpy.cluster_kmeans"><code class="xref py py-func docutils literal notranslate"><span class="pre">cluster_kmeans()</span></code></a>.</p></li>
<li><p><strong>more_kmeans_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Additional arguments to be passed to <a class="reference internal" href="#scranpy.cluster_kmeans" title="scranpy.cluster_kmeans"><code class="xref py py-func docutils literal notranslate"><span class="pre">cluster_kmeans()</span></code></a>.</p></li>
<li><p><strong>reddim_type</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#tuple" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tuple</span></code></a>]</span>) – Name or index of an existing embedding in <code class="docutils literal notranslate"><span class="pre">x</span></code>, on which to perform the clustering.
Alternatively a tuple, where the first element contains the name of an alternative experiment of <code class="docutils literal notranslate"><span class="pre">x</span></code>,
and the second element contains the name/index of an embedding in that alternative experiment.</p></li>
<li><p><strong>output_name</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></span>) – Name of the column of the column data in which to store the cluster assignments.</p></li>
<li><p><strong>meta_name</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – Name of the metadta entry in which to store extra clustering output.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, no extra clustering output is stored.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A copy of <code class="docutils literal notranslate"><span class="pre">x</span></code>, with the cluster assignment for each cell stored in its column data.
Additional clustering output is stored in its metadata.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">get_test_rna_data_se</span><span class="p">(</span><span class="s2">&quot;pca&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">cluster_kmeans_se</span><span class="p">(</span><span class="n">sce</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">biocutils</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">biocutils</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">sce</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">()[</span><span class="s2">&quot;clusters&quot;</span><span class="p">]))</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.combine_factors">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">combine_factors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">factors</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_unused</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.combine_factors" title="Permalink to this definition">¶</a></dt>
<dd><p>Combine multiple categorical factors based on the unique combinations of levels from each factor.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>factors</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>, <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a>, <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a>]</span>) – <p>Sequence of factors of interest.
Each entry corresponds to a factor and should be a sequence of the same length.
Corresponding elements across all factors represent the combination of levels for a single observation.</p>
<p>Alternatively, a dictionary where each entry corresponds to a factor.
Each key is the name of the factor and value is a sequence containing the factor.</p>
<p>Alternatively, a <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a> where each entry is a factor.
This may or may not be named.</p>
<p>Alternatively, a <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a> where each column is a factor.</p>
</p></li>
<li><p><strong>keep_unused</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to report unused combinations of levels.
If any entry of <code class="docutils literal notranslate"><span class="pre">factors</span></code> is a <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.Factor.Factor" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Factor</span></code></a> object, any unused levels will also be preserved.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><p><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a> containing the following entries.</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">levels</span></code>: a <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a> containing the sorted and unique combinations of levels as a tuple.
Each column corresponds to a factor in <code class="docutils literal notranslate"><span class="pre">factors</span></code> while each row represents a unique combination.
Corresponding elements of each column define a single combination, i.e., the <code class="docutils literal notranslate"><span class="pre">i</span></code>-th combination is defined by taking the <code class="docutils literal notranslate"><span class="pre">i</span></code>-th element of each column.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">index</span></code>: an integer NumPy array specifying the index into <code class="docutils literal notranslate"><span class="pre">levels</span></code> for each observation.</p></li>
</ul>
<p>For observation <code class="docutils literal notranslate"><span class="pre">i</span></code> and factor <code class="docutils literal notranslate"><span class="pre">j</span></code>, <code class="docutils literal notranslate"><span class="pre">levels[j][index[i]]</span></code> will recover <code class="docutils literal notranslate"><span class="pre">factors[j][i]</span></code>.</p>
</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The <code class="docutils literal notranslate"><span class="pre">combine_factors</span></code> function in the <a class="reference external" href="https://libscran.github.io/scran_aggregate">scran_aggregate</a> library.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">random</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">choices</span><span class="p">([</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">],</span> <span class="n">k</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">y</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">choices</span><span class="p">([</span><span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">],</span> <span class="n">k</span> <span class="o">=</span> <span class="mi">20</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">combined</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">combine_factors</span><span class="p">({</span> <span class="s2">&quot;foo&quot;</span><span class="p">:</span> <span class="n">x</span><span class="p">,</span> <span class="s2">&quot;bar&quot;</span><span class="p">:</span>  <span class="n">y</span> <span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">combined</span><span class="p">[</span><span class="s2">&quot;levels&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">biocutils</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">biocutils</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">combined</span><span class="p">[</span><span class="s2">&quot;index&quot;</span><span class="p">]))</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.compute_adt_qc_metrics">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">compute_adt_qc_metrics</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subsets</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">row_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_adt_quality_control.html#compute_adt_qc_metrics"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.compute_adt_qc_metrics" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute quality control metrics from ADT count data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Any" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a></span>) – A matrix-like object containing ADT counts.
Each row should correspond to an ADT while each column should correspond to a cell.</p></li>
<li><p><strong>subsets</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>, <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a>]</span>) – <p>Subsets of ADTs corresponding to control features like IgGs.
This may be either:</p>
<ul>
<li><p>A list of arrays.
Each array corresponds to an ADT subset and can either contain boolean or integer values.</p>
<ul>
<li><p>For booleans, the sequence should be of length equal to the number of rows, and values should be truthy for rows that belong in the subset.
If the sequence contains booleans, it should not contain any other type.</p></li>
<li><p>For integers, the value is the row index of an ADT in the subset.</p></li>
<li><p>For strings, the value is the name of an ADT in the subset.
This should match at least one element in <code class="docutils literal notranslate"><span class="pre">row_names</span></code>.</p></li>
</ul>
</li>
<li><p>A dictionary where keys are the names of each ADT subset and the values are arrays as described above.</p></li>
<li><p>A <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a> where each element is an array as described above, possibly with names.</p></li>
</ul>
</p></li>
<li><p><strong>row_names</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Sequence of strings of length equal to the number of rows of <code class="docutils literal notranslate"><span class="pre">x</span></code>, containing the name of each ADT.
Duplicate names are allowed but only the first occurrence will be used.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, rows are assumed to be unnamed.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><p>A <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a> with number of rows equal to the number of cells (i.e., columns) in <code class="docutils literal notranslate"><span class="pre">x</span></code>.
It contains the following columns:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">sum</span></code>, a double-precision NumPy array containing the sum of counts across all ADTs for each cell.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">detected</span></code>, an integer NumPy array containing the number of detected ADTs in each cell.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">subset_sum</span></code>, a nested <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a> with one column per subset in <code class="docutils literal notranslate"><span class="pre">subsets</span></code>.
Each column is a double-precision NumPy array that contains the sum of counts for the corresponding subset in each cell.</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The <code class="docutils literal notranslate"><span class="pre">compute_adt_qc_metrics</span></code> function in the <a class="reference external" href="https://github.com/libscran/scran_qc">scran_qc</a> C++ library, which describes the rationale behind these QC metrics.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mat</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="n">lam</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">1000</span><span class="p">),</span> <span class="p">(</span><span class="mi">50</span><span class="p">,</span> <span class="mi">20</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">compute_adt_qc_metrics</span><span class="p">(</span><span class="n">mat</span><span class="p">,</span> <span class="p">{</span> <span class="s2">&quot;IgG&quot;</span><span class="p">:</span> <span class="p">[</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">40</span> <span class="p">]</span> <span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.compute_clrm1_factors">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">compute_clrm1_factors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_compute_clrm1_factors.html#compute_clrm1_factors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.compute_clrm1_factors" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute size factors from an ADT count matrix using the CLRm1 method.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Any" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a></span>) – A matrix-like object containing ADT count data.
Rows correspond to tags and columns correspond to cells.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Double-precision NumPy array containing the CLRm1 size factor for each cell.
Note that these size factors are not centered and should be passed through, e.g., <a class="reference internal" href="#scranpy.center_size_factors" title="scranpy.center_size_factors"><code class="xref py py-func docutils literal notranslate"><span class="pre">center_size_factors()</span></code></a> before normalization.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p><a class="reference external" href="https://github.com/libscran/clrm1">https://github.com/libscran/clrm1</a>, for a description of the CLRm1 method.</p>
<dl>
<dt>Examples</dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="n">lam</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">1000</span><span class="p">)</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="mi">50</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sf</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">compute_clrm1_factors</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">scranpy</span><span class="o">.</span><span class="n">center_size_factors</span><span class="p">(</span><span class="n">sf</span><span class="p">))</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.compute_crispr_qc_metrics">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">compute_crispr_qc_metrics</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_crispr_quality_control.html#compute_crispr_qc_metrics"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.compute_crispr_qc_metrics" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute quality control metrics from CRISPR count data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Any" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a></span>) – A matrix-like object containing CRISPR counts.
Each row should correspond to a guide while each column should correspond to a cell.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><p>A <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a> with number of rows equal to the number of cells (i.e., columns) in <code class="docutils literal notranslate"><span class="pre">x</span></code>.
It contains the following columns.</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">sum</span></code>, a double-precision NumPy array containing the sum of counts across all guides for each cell.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">detected</span></code>, an integer NumPy array containing the number of guides with non-zero counts in each cell.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">max_value</span></code>, a double-precision NumPy array containing the maximum count for each cell.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">max_index</span></code>, an integer NumPy array containing the row index of the guide with the maximum count in each cell.</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The <code class="docutils literal notranslate"><span class="pre">compute_crispr_qc_metrics</span></code> function in the <a class="reference external" href="https://github.com/libscran/scran_qc">scran_qc</a> C++ library, which describes the rationale behind these QC metrics.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mat</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="n">lam</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">1000</span><span class="p">),</span> <span class="p">(</span><span class="mi">50</span><span class="p">,</span> <span class="mi">20</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">compute_crispr_qc_metrics</span><span class="p">(</span><span class="n">mat</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.compute_rna_qc_metrics">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">compute_rna_qc_metrics</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subsets</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">row_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_rna_quality_control.html#compute_rna_qc_metrics"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.compute_rna_qc_metrics" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute quality control metrics from RNA count data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Any" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a></span>) – A matrix-like object containing RNA counts.
Each row should correspond to a gene while each column should correspond to a cell.</p></li>
<li><p><strong>subsets</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>, <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a>]</span>) – <p>Subsets of genes corresponding to “control” features like mitochondrial genes.
This may be either:</p>
<ul>
<li><p>A list of sequences.
Each sequence corresponds to a gene subset and can contain booleans, integers or strings.</p>
<ul>
<li><p>For booleans, the sequence should be of length equal to the number of rows, and values should be truthy for rows that belong in the subset.
If the sequence contains booleans, it should not contain any other type.</p></li>
<li><p>For integers, the value is the row index of a gene in the subset.</p></li>
<li><p>For strings, the value is the name of a gene in the subset.
This should match at least one element in <code class="docutils literal notranslate"><span class="pre">row_names</span></code>.</p></li>
</ul>
</li>
<li><p>A dictionary where keys are the names of each gene subset and the values are arrays as described above.</p></li>
<li><p>A <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a> where each element is an array as described above, possibly with names.</p></li>
</ul>
</p></li>
<li><p><strong>row_names</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Sequence of strings of length equal to the number of rows of <code class="docutils literal notranslate"><span class="pre">x</span></code>, containing the name of each gene.
Duplicate names are allowed but only the first occurrence will be used.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, rows are assumed to be unnamed.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><p>A <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a> with number of rows equal to the number of cells (i.e., columns) in <code class="docutils literal notranslate"><span class="pre">x</span></code>.
It contains the following columns:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">sum</span></code>: a double-precision NumPy array containing the sum of counts across all genes for each cell.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">detected</span></code>: an integer NumPy array containing the number of genes with non-zero expression in each cell.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">subset_proportion</span></code>: a nested <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a> with one column per subset in <code class="docutils literal notranslate"><span class="pre">subsets</span></code>.
Each column is a double-precision NumPy array that contains the proportion of counts in the corresponding subset in each cell.</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The <code class="docutils literal notranslate"><span class="pre">compute_rna_qc_metrics</span></code> function in the <a class="reference external" href="https://github.com/libscran/scran_qc">scran_qc</a> C++ library, which describes the rationale behind these QC metrics.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mat</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="n">lam</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">1000</span><span class="p">),</span> <span class="p">(</span><span class="mi">50</span><span class="p">,</span> <span class="mi">20</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">compute_rna_qc_metrics</span><span class="p">(</span><span class="n">mat</span><span class="p">,</span> <span class="p">{</span> <span class="s2">&quot;mito&quot;</span><span class="p">:</span> <span class="p">[</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">40</span> <span class="p">]</span> <span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.compute_rna_qc_metrics_with_altexps">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">compute_rna_qc_metrics_with_altexps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subsets</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">altexp_proportions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assay_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'counts'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.compute_rna_qc_metrics_with_altexps" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute RNA-related QC metrics for the main and alternative experiments of a <a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a>.
This calls <a class="reference internal" href="#scranpy.compute_rna_qc_metrics" title="scranpy.compute_rna_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_rna_qc_metrics()</span></code></a> on each experiment.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a></span>) – A <a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a> object or one of its subclasses.
Rows correspond to genes and columns correspond to cells.</p></li>
<li><p><strong>subsets</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Mapping" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Mapping</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – List of subsets of control genes, passed to <a class="reference internal" href="#scranpy.compute_rna_qc_metrics" title="scranpy.compute_rna_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_rna_qc_metrics()</span></code></a> for the main experiment.</p></li>
<li><p><strong>altexp_proportions</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>, <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/constants.html#None" title="(in Python v3.14)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]</span>) – <p>Indices or names of alternative experiments for which to compute QC metrics.
This is typically used to refer to alternative experiments holding spike-in data.
An additional proportion is computed for each alternative experiment relative to the main experiment,
i.e., X/(X+Y) where X is the alternative experiment’s total and Y is the main experiment’s total.
These proportions will be used for filtering in the same manner as the proportions computed from <code class="docutils literal notranslate"><span class="pre">subsets</span></code>.</p>
<p>If a list or unnamed <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a> is supplied,
it should contain the indices/names of alternative experiments for which to compute QC metrics.
The assay to use from each alternative experiment is determined by <code class="docutils literal notranslate"><span class="pre">assay_type</span></code>.</p>
<p>If a string or integer is supplied, it is assumed to be a name/index of one alternative experiment for which to compute QC metrics.
Again, the assay to use from each alternative experiment is determined by <code class="docutils literal notranslate"><span class="pre">assay_type</span></code>.</p>
<p>If a dictionary or named <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a> is supplied,
each key is a string specifying the name of an alternative experiment for which to compute QC metrics,
while each value is an integer/string specifying the index/name of the assay to use from that experiment.</p>
<p>This option is only relevant if <code class="docutils literal notranslate"><span class="pre">x</span></code> is a <a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a>.</p>
</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads, passed to <a class="reference internal" href="#scranpy.compute_rna_qc_metrics" title="scranpy.compute_rna_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_rna_qc_metrics()</span></code></a>.</p></li>
<li><p><strong>assay_type</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – Index or name of the assay of <code class="docutils literal notranslate"><span class="pre">x</span></code> containing the ADT count matrix.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">main</span></code>: a <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a> containing QC statistics for the main experiment,
see <a class="reference internal" href="#scranpy.compute_rna_qc_metrics" title="scranpy.compute_rna_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_rna_qc_metrics()</span></code></a> for details.
The proportion of counts for each alternative experiment in <code class="docutils literal notranslate"><span class="pre">altexp_proportions</span></code> is stored in the <code class="docutils literal notranslate"><span class="pre">subset_proportions</span></code> column.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">altexps</span></code>: a <code class="docutils literal notranslate"><span class="pre">NamedList</span></code> with one entry per alternative experiment listed in <code class="docutils literal notranslate"><span class="pre">altexp_proportions</span></code>.
Each entry is named after its corresponding alternative experiment and is a <code class="docutils literal notranslate"><span class="pre">BiocFrame</span></code> of QC statistics for that experiment.</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>A <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a> containing</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">get_test_rna_data_se</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">is_mito</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">y</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;mt-&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">sce</span><span class="o">.</span><span class="n">get_row_names</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">compute_rna_qc_metrics_with_altexps</span><span class="p">(</span><span class="n">sce</span><span class="p">,</span> <span class="n">subsets</span><span class="o">=</span><span class="p">{</span> <span class="s2">&quot;mito&quot;</span><span class="p">:</span> <span class="n">is_mito</span> <span class="p">},</span> <span class="n">altexp_proportions</span><span class="o">=</span><span class="s2">&quot;ERCC&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">][</span><span class="s2">&quot;subset_proportion&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="s2">&quot;altexps&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.correct_mnn">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">correct_mnn</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_neighbors=15</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_steps=1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">merge_policy='rss'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_mads=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">robust_iterations=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">robust_trim=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mass_cap=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_policy=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nn_parameters=&lt;knncolle.annoy.AnnoyParameters</span> <span class="pre">object&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads=1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_correct_mnn.html#correct_mnn"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.correct_mnn" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform mutual nearest neighbor (MNN) correction to remove batch effects from a matrix of low-dimensional embeddings.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span>) – Matrix of coordinates where rows are dimensions and columns are cells, typically generated by <a class="reference internal" href="#scranpy.run_pca" title="scranpy.run_pca"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_pca()</span></code></a>.</p></li>
<li><p><strong>block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a></span>) – Factor specifying the block of origin (e.g., batch, sample) for each cell. Length should equal the number of columns in <code class="docutils literal notranslate"><span class="pre">x</span></code>.</p></li>
<li><p><strong>num_neighbors</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of neighbors to use when identifying MNN pairs.</p></li>
<li><p><strong>num_steps</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of steps for the recursive neighbor search to compute the center of mass.
Larger values mitigate the kissing effect but increase the risk of including inappropriately distant subpopulations into the center of mass.</p></li>
<li><p><strong>merge_policy</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Literal" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'rss'</span></code>, <code class="docutils literal notranslate"><span class="pre">'size'</span></code>, <code class="docutils literal notranslate"><span class="pre">'variance'</span></code>, <code class="docutils literal notranslate"><span class="pre">'input'</span></code>]</span>) – <p>Policy for choosing the merge order.</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">input</span></code> uses the input order of the batches.
Observations in the last batch are corrected first, and then the second-last batch, and so on.
This allows users to control the merge order by simply changing the inputs.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">size</span></code> merges batches in order of increasing size (i.e., the number of observations).
So, the smallest batch is corrected first while the largest batch is unchanged.
The aim is to lower compute time by reducing the number of observations that need to be reprocessed in later merge steps.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">variance</span></code> merges batches in order of increasing variance between observations.
So, the batch with the lowest variance is corrected first while the batch with the highest variance is unchanged.
The aim is to lower compute time by encouraging more observations to be corrected to the most variable batch, thus avoid reprocessing in later merge steps.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">rss</span></code> merges batches in order of increasing residual sum of squares (RSS).
This is effectively a compromise between <code class="docutils literal notranslate"><span class="pre">variance</span></code> and <code class="docutils literal notranslate"><span class="pre">size</span></code>.</p></li>
</ul>
</p></li>
<li><p><strong>num_mads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>) – Deprecated and ignored.</p></li>
<li><p><strong>robust_iterations</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>) – Deprecated and ignored.</p></li>
<li><p><strong>robust_trim</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a>]</span>) – Deprecated and ignored.</p></li>
<li><p><strong>mass_cap</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>) – Deprecated and ignored.</p></li>
<li><p><strong>order</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Deprecated and ignored.
Now the merge order is always determinted automatically.</p></li>
<li><p><strong>reference_policy</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – Deprecated, see <code class="docutils literal notranslate"><span class="pre">merge_policy</span></code> instead.</p></li>
<li><p><strong>nn_parameters</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.classes.Parameters" title="(in knncolle v0.0.post1.dev1+g951fea368)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Parameters</span></code></a></span>) – The nearest-neighbor algorithm to use.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><p>A <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a> containing the following entries.</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">corrected</span></code>, a double-precision NumPy array of the same dimensions as <code class="docutils literal notranslate"><span class="pre">x</span></code>, containing the corrected values.</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">References</p>
<p><a class="reference external" href="https://libscran.github.io/mnncorrect">https://libscran.github.io/mnncorrect</a>, which describes the MNN correction algorithm in more detail.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pcs</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">200</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">block</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">100</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mnn_out</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">correct_mnn</span><span class="p">(</span><span class="n">pcs</span><span class="p">,</span> <span class="n">block</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mnn_out</span><span class="p">[</span><span class="s2">&quot;corrected&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.correct_mnn_se">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">correct_mnn_se</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nn_parameters=&lt;knncolle.annoy.AnnoyParameters</span> <span class="pre">object&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads=1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">more_mnn_args={}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reddim_type='PCA'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_name='MNN'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_se_correct_mnn.html#correct_mnn_se"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.correct_mnn_se" title="Permalink to this definition">¶</a></dt>
<dd><p>Correct batch effects from an existing embedding with mutual nearest neighbors (MNNs).
This calls <a class="reference internal" href="#scranpy.correct_mnn" title="scranpy.correct_mnn"><code class="xref py py-func docutils literal notranslate"><span class="pre">correct_mnn()</span></code></a> on reduced dimensions from a <a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a></span>) – A <a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a> object or one of its subclasses.
Rows correspond to genomic features and columns correspond to cells.</p></li>
<li><p><strong>block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a></span>) – Block assignment of each cell, see <a class="reference internal" href="#scranpy.correct_mnn" title="scranpy.correct_mnn"><code class="xref py py-func docutils literal notranslate"><span class="pre">correct_mnn()</span></code></a> for more details.</p></li>
<li><p><strong>nn_parameters</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.classes.Parameters" title="(in knncolle v0.0.post1.dev1+g951fea368)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Parameters</span></code></a></span>) – Algorithm for nearest neighbor search, see <a class="reference internal" href="#scranpy.correct_mnn" title="scranpy.correct_mnn"><code class="xref py py-func docutils literal notranslate"><span class="pre">correct_mnn()</span></code></a> for more details.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads, see <a class="reference internal" href="#scranpy.correct_mnn" title="scranpy.correct_mnn"><code class="xref py py-func docutils literal notranslate"><span class="pre">correct_mnn()</span></code></a> for more details.</p></li>
<li><p><strong>more_mnn_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Additional arguments to pass to <a class="reference internal" href="#scranpy.correct_mnn" title="scranpy.correct_mnn"><code class="xref py py-func docutils literal notranslate"><span class="pre">correct_mnn()</span></code></a>.</p></li>
<li><p><strong>reddim_type</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#tuple" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tuple</span></code></a>]</span>) – Name or index of an existing embedding in <code class="docutils literal notranslate"><span class="pre">x</span></code>, on which to perform the batch correction.
Alternatively a tuple, where the first element contains the name of an alternative experiment of <code class="docutils literal notranslate"><span class="pre">x</span></code>,
and the second element contains the name/index of an embedding in that alternative experiment.</p></li>
<li><p><strong>output_name</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></span>) – Name of the reduced dimension entry which to store the corrected coordinates.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A copy of <code class="docutils literal notranslate"><span class="pre">x</span></code> , with the corrected embedding stored as a reduced dimension entry.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">get_test_rna_data_se</span><span class="p">(</span><span class="s2">&quot;pca&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Treating the tissue of origin as the batch.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">correct_mnn_se</span><span class="p">(</span><span class="n">sce</span><span class="p">,</span> <span class="n">block</span><span class="o">=</span><span class="n">sce</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">()[</span><span class="s2">&quot;tissue&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span><span class="o">.</span><span class="n">get_reduced_dimension_names</span><span class="p">()</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.filter_adt_qc_metrics">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">filter_adt_qc_metrics</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">thresholds</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metrics</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_adt_quality_control.html#filter_adt_qc_metrics"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.filter_adt_qc_metrics" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter for high-quality cells based on ADT-derived QC metrics.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>thresholds</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a></span>) – Filter thresholds on the QC metrics, typically computed with <a class="reference internal" href="#scranpy.suggest_adt_qc_thresholds" title="scranpy.suggest_adt_qc_thresholds"><code class="xref py py-func docutils literal notranslate"><span class="pre">suggest_adt_qc_thresholds()</span></code></a>.</p></li>
<li><p><strong>metrics</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a></span>) – ADT-derived QC metrics, typically computed with <a class="reference internal" href="#scranpy.compute_adt_qc_metrics" title="scranpy.compute_adt_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_adt_qc_metrics()</span></code></a>.</p></li>
<li><p><strong>block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Blocking factor specifying the block of origin (e.g., batch, sample) for each cell in <code class="docutils literal notranslate"><span class="pre">metrics</span></code>.
The levels should be a subset of those used in <a class="reference internal" href="#scranpy.suggest_adt_qc_thresholds" title="scranpy.suggest_adt_qc_thresholds"><code class="xref py py-func docutils literal notranslate"><span class="pre">suggest_adt_qc_thresholds()</span></code></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A boolean NumPy vector of length equal to the number of cells in <code class="docutils literal notranslate"><span class="pre">metrics</span></code>, containing truthy values for putative high-quality cells.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The <code class="docutils literal notranslate"><span class="pre">AdtQcFilters</span></code> and <code class="docutils literal notranslate"><span class="pre">AdtQcBlockedFilters</span></code> functions in the <a class="reference external" href="https://github.com/libscran/scran_qc">scran_qc</a> C++ library.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mat</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="n">lam</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">1000</span><span class="p">),</span> <span class="p">(</span><span class="mi">50</span><span class="p">,</span> <span class="mi">20</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">compute_adt_qc_metrics</span><span class="p">(</span><span class="n">mat</span><span class="p">,</span> <span class="p">{</span> <span class="s2">&quot;IgG&quot;</span><span class="p">:</span> <span class="p">[</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">40</span> <span class="p">]</span> <span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">suggest_adt_qc_thresholds</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">keep</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">filter_adt_qc_metrics</span><span class="p">(</span><span class="n">filt</span><span class="p">,</span> <span class="n">res</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">biocutils</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">biocutils</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">keep</span><span class="p">))</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.filter_crispr_qc_metrics">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">filter_crispr_qc_metrics</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">thresholds</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metrics</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_crispr_quality_control.html#filter_crispr_qc_metrics"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.filter_crispr_qc_metrics" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter for high-quality cells based on CRISPR-derived QC metrics.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>thresholds</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a></span>) – Filter thresholds on the QC metrics, typically computed with <a class="reference internal" href="#scranpy.suggest_crispr_qc_thresholds" title="scranpy.suggest_crispr_qc_thresholds"><code class="xref py py-func docutils literal notranslate"><span class="pre">suggest_crispr_qc_thresholds()</span></code></a>.</p></li>
<li><p><strong>metrics</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a></span>) – CRISPR-derived QC metrics, typically computed with <a class="reference internal" href="#scranpy.compute_crispr_qc_metrics" title="scranpy.compute_crispr_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_crispr_qc_metrics()</span></code></a>.</p></li>
<li><p><strong>block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Blocking factor specifying the block of origin (e.g., batch, sample) for each cell in <code class="docutils literal notranslate"><span class="pre">metrics</span></code>.
The levels should be a subset of those used in <a class="reference internal" href="#scranpy.suggest_crispr_qc_thresholds" title="scranpy.suggest_crispr_qc_thresholds"><code class="xref py py-func docutils literal notranslate"><span class="pre">suggest_crispr_qc_thresholds()</span></code></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A boolean NumPy vector of length equal to the number of cells in <code class="docutils literal notranslate"><span class="pre">metrics</span></code>, containing truthy values for putative high-quality cells.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The <code class="docutils literal notranslate"><span class="pre">CrisprQcFilters</span></code> and <code class="docutils literal notranslate"><span class="pre">CrisprQcBlockedFilters</span></code> functions in the <a class="reference external" href="https://github.com/libscran/scran_qc">scran_qc</a> C++ library.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mat</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="n">lam</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">1000</span><span class="p">),</span> <span class="p">(</span><span class="mi">50</span><span class="p">,</span> <span class="mi">20</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">compute_crispr_qc_metrics</span><span class="p">(</span><span class="n">mat</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">suggest_crispr_qc_thresholds</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">keep</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">filter_crispr_qc_metrics</span><span class="p">(</span><span class="n">filt</span><span class="p">,</span> <span class="n">res</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">keep</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.filter_rna_qc_metrics">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">filter_rna_qc_metrics</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">thresholds</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metrics</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_rna_quality_control.html#filter_rna_qc_metrics"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.filter_rna_qc_metrics" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter for high-quality cells based on RNA-derived QC metrics.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>thresholds</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a></span>) – Filter thresholds on the QC metrics, typically computed with <a class="reference internal" href="#scranpy.suggest_rna_qc_thresholds" title="scranpy.suggest_rna_qc_thresholds"><code class="xref py py-func docutils literal notranslate"><span class="pre">suggest_rna_qc_thresholds()</span></code></a>.</p></li>
<li><p><strong>metrics</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a></span>) – RNA-derived QC metrics, typically computed with <a class="reference internal" href="#scranpy.compute_rna_qc_metrics" title="scranpy.compute_rna_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_rna_qc_metrics()</span></code></a>.</p></li>
<li><p><strong>block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Blocking factor specifying the block of origin (e.g., batch, sample) for each cell in <code class="docutils literal notranslate"><span class="pre">metrics</span></code>.
The levels should be a subset of those used in <a class="reference internal" href="#scranpy.suggest_rna_qc_thresholds" title="scranpy.suggest_rna_qc_thresholds"><code class="xref py py-func docutils literal notranslate"><span class="pre">suggest_rna_qc_thresholds()</span></code></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A boolean NumPy vector of length equal to the number of cells in <code class="docutils literal notranslate"><span class="pre">metrics</span></code>, containing truthy values for putative high-quality cells.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The <code class="docutils literal notranslate"><span class="pre">RnaQcFilters</span></code> and <code class="docutils literal notranslate"><span class="pre">RnaQcBlockedFilters</span></code> functions in the <a class="reference external" href="https://github.com/libscran/scran_qc">scran_qc</a> C++ library.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mat</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="n">lam</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">1000</span><span class="p">),</span> <span class="p">(</span><span class="mi">50</span><span class="p">,</span> <span class="mi">20</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">compute_rna_qc_metrics</span><span class="p">(</span><span class="n">mat</span><span class="p">,</span> <span class="p">{</span> <span class="s2">&quot;mito&quot;</span><span class="p">:</span> <span class="p">[</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">40</span> <span class="p">]</span> <span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">suggest_rna_qc_thresholds</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">keep</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">filter_rna_qc_metrics</span><span class="p">(</span><span class="n">filt</span><span class="p">,</span> <span class="n">res</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">biocutils</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">keep</span><span class="p">))</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.fit_variance_trend">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">fit_variance_trend</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mean</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variance</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mean_filter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_mean</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transform</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">span</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_min_width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_window_count</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">200</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_fit_variance_trend.html#fit_variance_trend"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.fit_variance_trend" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit a trend to the per-gene variances with respect to the mean.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mean</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span>) – Array containing the mean (log-)expression for each gene.</p></li>
<li><p><strong>variance</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span>) – Array containing the variance in the (log-)expression for each gene.
This should have length equal to <code class="docutils literal notranslate"><span class="pre">mean</span></code>.</p></li>
<li><p><strong>mean_filter</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to filter on the means before trend fitting.</p></li>
<li><p><strong>min_mean</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – The minimum mean of genes to use in trend fitting.
Only used if <code class="docutils literal notranslate"><span class="pre">mean_filter</span> <span class="pre">=</span> <span class="pre">True</span></code>.</p></li>
<li><p><strong>transform</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether a quarter-root transformation should be applied before trend fitting.</p></li>
<li><p><strong>span</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Span of the LOWESS smoother.
Ignored if <code class="docutils literal notranslate"><span class="pre">use_min_width</span> <span class="pre">=</span> <span class="pre">True</span></code>.</p></li>
<li><p><strong>use_min_width</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether a minimum width constraint should be applied to the LOWESS smoother.
This is useful to avoid overfitting in high-density intervals.</p></li>
<li><p><strong>min_width</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Minimum width of the window to use when <code class="docutils literal notranslate"><span class="pre">use_min_width</span> <span class="pre">=</span> <span class="pre">True</span></code>.</p></li>
<li><p><strong>min_window_count</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Minimum number of observations in each window.
Only used if <code class="docutils literal notranslate"><span class="pre">use_min_width</span> <span class="pre">=</span> <span class="pre">True</span></code>.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><p>A <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a> containing the following entries.</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">fitted</span></code>: a double-precision NumPy array of length equal to <code class="docutils literal notranslate"><span class="pre">mean</span></code>, containing the fitted value of the trend for each gene.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">residual</span></code>: a double-precision NumPy array of length equal to <code class="docutils literal notranslate"><span class="pre">mean</span></code>, containing the residual from the trend for each gene.</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The <code class="docutils literal notranslate"><span class="pre">fit_variance_trend</span></code> function in the <a class="reference external" href="https://libscran.github.io/scran_variances">scran_variances</a> C++ library, for the underlying implementation.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Mock up a typical mean-variance relationship for count data.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mean</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">exponential</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="mi">1000</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">variance</span> <span class="o">=</span> <span class="n">mean</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">mean</span><span class="o">**</span><span class="mi">2</span><span class="p">)</span> <span class="o">*</span> <span class="mi">2</span><span class="o">**</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">mean</span><span class="p">))</span> <span class="o">/</span> <span class="mi">10</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fit</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">fit_variance_trend</span><span class="p">(</span><span class="n">mean</span><span class="p">,</span> <span class="n">variance</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">fit</span><span class="p">[</span><span class="s2">&quot;fitted&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">fit</span><span class="p">[</span><span class="s2">&quot;residual&quot;</span><span class="p">])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.format_compute_adt_qc_metrics_result">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">format_compute_adt_qc_metrics_result</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flatten</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.format_compute_adt_qc_metrics_result" title="Permalink to this definition">¶</a></dt>
<dd><p>Pretty-format the results of <a class="reference internal" href="#scranpy.compute_adt_qc_metrics" title="scranpy.compute_adt_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_adt_qc_metrics()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a></span>) – Result of <a class="reference internal" href="#scranpy.compute_adt_qc_metrics" title="scranpy.compute_adt_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_adt_qc_metrics()</span></code></a>.</p></li>
<li><p><strong>flatten</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to flatten the nested <code class="docutils literal notranslate"><span class="pre">BiocFrame</span></code> of subset sums.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A <code class="docutils literal notranslate"><span class="pre">BiocFrame</span></code> containing per-cell QC statistics.
If <code class="docutils literal notranslate"><span class="pre">flatten</span> <span class="pre">=</span> <span class="pre">True</span></code>, the subset sums are stored as top-level columns with name <code class="docutils literal notranslate"><span class="pre">subset_sum_&lt;SUBSET&gt;</span></code> where <code class="docutils literal notranslate"><span class="pre">&lt;SUBSET&gt;</span></code> is the name of the subset.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">get_test_adt_data_se</span><span class="p">()</span><span class="o">.</span><span class="n">get_alternative_experiment</span><span class="p">(</span><span class="s2">&quot;ADT&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">is_igg</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">y</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s2">&quot;IgG&quot;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">sce</span><span class="o">.</span><span class="n">get_row_names</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">qc</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">compute_adt_qc_metrics</span><span class="p">(</span><span class="n">sce</span><span class="o">.</span><span class="n">get_assay</span><span class="p">(</span><span class="mi">0</span><span class="p">),</span> <span class="n">subsets</span><span class="o">=</span><span class="p">{</span> <span class="s2">&quot;igg&quot;</span><span class="p">:</span> <span class="n">is_igg</span> <span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">scranpy</span><span class="o">.</span><span class="n">format_compute_adt_qc_metrics_result</span><span class="p">(</span><span class="n">qc</span><span class="p">))</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.format_compute_crispr_qc_metrics_result">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">format_compute_crispr_qc_metrics_result</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.format_compute_crispr_qc_metrics_result" title="Permalink to this definition">¶</a></dt>
<dd><p>Pretty-format the results of <a class="reference internal" href="#scranpy.compute_crispr_qc_metrics" title="scranpy.compute_crispr_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_crispr_qc_metrics()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>df</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a></span>) – Result of <a class="reference internal" href="#scranpy.compute_crispr_qc_metrics" title="scranpy.compute_crispr_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_crispr_qc_metrics()</span></code></a>.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A <code class="docutils literal notranslate"><span class="pre">BiocFrame</span></code> containing per-cell QC statistics.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">get_test_crispr_data_se</span><span class="p">()</span><span class="o">.</span><span class="n">get_alternative_experiment</span><span class="p">(</span><span class="s2">&quot;CRISPR Guide Capture&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">compute_crispr_qc_metrics</span><span class="p">(</span><span class="n">sce</span><span class="o">.</span><span class="n">get_assay</span><span class="p">(</span><span class="mi">0</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">scranpy</span><span class="o">.</span><span class="n">format_compute_crispr_qc_metrics_result</span><span class="p">(</span><span class="n">df</span><span class="p">))</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.format_compute_rna_qc_metrics_result">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">format_compute_rna_qc_metrics_result</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flatten</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.format_compute_rna_qc_metrics_result" title="Permalink to this definition">¶</a></dt>
<dd><p>Pretty-format the results of <a class="reference internal" href="#scranpy.compute_rna_qc_metrics" title="scranpy.compute_rna_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_rna_qc_metrics()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a></span>) – Result of <a class="reference internal" href="#scranpy.compute_rna_qc_metrics" title="scranpy.compute_rna_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_rna_qc_metrics()</span></code></a>.</p></li>
<li><p><strong>flatten</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to flatten the nested <code class="docutils literal notranslate"><span class="pre">BiocFrame</span></code> of subset proportions.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A <code class="docutils literal notranslate"><span class="pre">BiocFrame</span></code> containing per-cell QC statistics.
If <code class="docutils literal notranslate"><span class="pre">flatten</span> <span class="pre">=</span> <span class="pre">True</span></code>, the subset proportions are stored as top-level columns with name <code class="docutils literal notranslate"><span class="pre">subset_proportion_&lt;SUBSET&gt;</span></code> where <code class="docutils literal notranslate"><span class="pre">&lt;SUBSET&gt;</span></code> is the name of the subset.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">get_test_rna_data_se</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">is_mito</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">y</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;mt-&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">sce</span><span class="o">.</span><span class="n">get_row_names</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">qc</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">compute_rna_qc_metrics</span><span class="p">(</span><span class="n">sce</span><span class="o">.</span><span class="n">get_assay</span><span class="p">(</span><span class="mi">0</span><span class="p">),</span> <span class="n">subsets</span><span class="o">=</span><span class="p">{</span> <span class="s2">&quot;mito&quot;</span><span class="p">:</span> <span class="n">is_mito</span> <span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">scranpy</span><span class="o">.</span><span class="n">format_compute_rna_qc_metrics_result</span><span class="p">(</span><span class="n">qc</span><span class="p">))</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.format_score_markers_result">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">format_score_markers_result</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">res</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extra_columns</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order_by</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">row_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.format_score_markers_result" title="Permalink to this definition">¶</a></dt>
<dd><p>Reformat the output of <a class="reference internal" href="#scranpy.score_markers" title="scranpy.score_markers"><code class="xref py py-func docutils literal notranslate"><span class="pre">score_markers()</span></code></a> into a list of per-group <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a> objects.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>res</strong> – Results of <a class="reference internal" href="#scranpy.score_markers" title="scranpy.score_markers"><code class="xref py py-func docutils literal notranslate"><span class="pre">score_markers()</span></code></a>.</p></li>
<li><p><strong>extra_columns</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>, <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/constants.html#None" title="(in Python v3.14)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]</span>) – A <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a> with the same number of rows as <code class="docutils literal notranslate"><span class="pre">x</span></code>, containing extra columns to add each <code class="docutils literal notranslate"><span class="pre">BiocFrame</span></code>.</p></li>
<li><p><strong>order_by</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/constants.html#None" title="(in Python v3.14)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]</span>) – Name of the column to use for ordering the rows of each output <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a>.
Alternatively <code class="docutils literal notranslate"><span class="pre">True</span></code>, in which case a column is automatically chosen from the effect size summaries.
If <code class="docutils literal notranslate"><span class="pre">None</span></code> or <code class="docutils literal notranslate"><span class="pre">False</span></code>, no ordering is performed.</p></li>
<li><p><strong>row_names</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Sequence of strings containing the row names to add to each <code class="docutils literal notranslate"><span class="pre">BiocFrame</span></code>.
This should correspond to the gene names corresponding to the rows of <code class="docutils literal notranslate"><span class="pre">x</span></code> used in <a class="reference internal" href="#scranpy.score_markers" title="scranpy.score_markers"><code class="xref py py-func docutils literal notranslate"><span class="pre">score_markers()</span></code></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><p>A <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a> of <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a> objects.
Each <code class="docutils literal notranslate"><span class="pre">BiocFrame</span></code> corresponds to a unique group in <code class="docutils literal notranslate"><span class="pre">groups</span></code>.
Each row of each <code class="docutils literal notranslate"><span class="pre">BiocFrame</span></code> contains statistics for a gene in <code class="docutils literal notranslate"><span class="pre">x</span></code>.
Each <code class="docutils literal notranslate"><span class="pre">BiocFrame</span></code> contains the following columns:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">mean</span></code>, the mean expression in the current group.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">detected</span></code>, the proportion of cells with detected expression in the current group.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&lt;effect&gt;_&lt;summary&gt;</span></code>, a summary statistic for an effect size,
e.g., <code class="docutils literal notranslate"><span class="pre">cohens_d_mean</span></code> contains the mean Cohen’s d across comparisons involving the current group.</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">get_test_rna_data_se</span><span class="p">(</span><span class="s2">&quot;cluster&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">raw_markers</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">score_markers</span><span class="p">(</span><span class="n">sce</span><span class="o">.</span><span class="n">get_assay</span><span class="p">(</span><span class="s2">&quot;logcounts&quot;</span><span class="p">),</span> <span class="n">sce</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">()[</span><span class="s2">&quot;clusters&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">formatted</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">format_score_markers_result</span><span class="p">(</span><span class="n">raw_markers</span><span class="p">,</span> <span class="n">row_names</span><span class="o">=</span><span class="n">sce</span><span class="o">.</span><span class="n">get_row_names</span><span class="p">())</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.get_test_adt_data_se">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">get_test_adt_data_se</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">at</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'start'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.get_test_adt_data_se" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a CITE-seq dataset with varying levels of processing.
This contains human PBMCs obtained with <code class="docutils literal notranslate"><span class="pre">scrnaseq.fetch_dataset(&quot;kotliarov-pbmc-2020&quot;,</span> <span class="pre">&quot;2024-04-18&quot;)</span></code>.
The main experiment contains RNA counts and the alternative experiment contains ADT counts.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>at</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Literal" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'start'</span></code>, <code class="docutils literal notranslate"><span class="pre">'qc'</span></code>, <code class="docutils literal notranslate"><span class="pre">'norm'</span></code>, <code class="docutils literal notranslate"><span class="pre">'hvg'</span></code>, <code class="docutils literal notranslate"><span class="pre">'pca'</span></code>]</span>) – The desired level of processing.
For <code class="docutils literal notranslate"><span class="pre">start</span></code>, no processing is performed.
Otherwise, the dataset is returned after quality control (<code class="docutils literal notranslate"><span class="pre">qc</span></code>), normalization (<code class="docutils literal notranslate"><span class="pre">norm</span></code>), feature selection (<code class="docutils literal notranslate"><span class="pre">hvg</span></code>) or PCA (<code class="docutils literal notranslate"><span class="pre">pca</span></code>).</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The dataset as a <a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a>.
This uses caching to avoid recomputation.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">get_test_adt_data_se</span><span class="p">()</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.get_test_crispr_data_se">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">get_test_crispr_data_se</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">at</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'start'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.get_test_crispr_data_se" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a Perturb-seq dataset with varying levels of processing.
This contains a pancreatic beta cell line obtained with <code class="docutils literal notranslate"><span class="pre">scrnaseq.fetch_dataset(&quot;cao-pancreas-2025&quot;,</span> <span class="pre">&quot;2025-10-10&quot;,</span> <span class="pre">&quot;rqc&quot;)</span></code>.
The main experiment contains RNA counts and the alternative experiment contains CRISPR guide counts.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>at</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Literal" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'start'</span></code>, <code class="docutils literal notranslate"><span class="pre">'qc'</span></code>]</span>) – The desired level of processing.
For <code class="docutils literal notranslate"><span class="pre">start</span></code>, no processing is performed.
Otherwise, the dataset is returned after quality control (<code class="docutils literal notranslate"><span class="pre">qc</span></code>), normalization (<code class="docutils literal notranslate"><span class="pre">norm</span></code>), feature selection (<code class="docutils literal notranslate"><span class="pre">hvg</span></code>) or PCA (<code class="docutils literal notranslate"><span class="pre">pca</span></code>).</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The dataset as a <a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a>.
This uses caching to avoid recomputation.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">get_test_crispr_data_se</span><span class="p">()</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.get_test_rna_data_se">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">get_test_rna_data_se</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">at</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'start'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.get_test_rna_data_se" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a single-cell RNA-seq dataset with varying levels of processing.
This contains cells from the mouse brain, obtained with <code class="docutils literal notranslate"><span class="pre">scrnaseq.fetch_dataset(&quot;zeisel-brain-2015&quot;,</span> <span class="pre">&quot;2023-12-14&quot;)</span></code>.
The main experiment contains RNA counts and the alternative experiments contain ERCC and repeat element counts.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>at</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Literal" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'start'</span></code>, <code class="docutils literal notranslate"><span class="pre">'qc'</span></code>, <code class="docutils literal notranslate"><span class="pre">'norm'</span></code>, <code class="docutils literal notranslate"><span class="pre">'hvg'</span></code>, <code class="docutils literal notranslate"><span class="pre">'pca'</span></code>, <code class="docutils literal notranslate"><span class="pre">'cluster'</span></code>]</span>) – The desired level of processing.
For <code class="docutils literal notranslate"><span class="pre">start</span></code>, no processing is performed.
Otherwise, the dataset is returned after quality control (<code class="docutils literal notranslate"><span class="pre">qc</span></code>), normalization (<code class="docutils literal notranslate"><span class="pre">norm</span></code>), feature selection (<code class="docutils literal notranslate"><span class="pre">hvg</span></code>), PCA (<code class="docutils literal notranslate"><span class="pre">pca</span></code>) or clustering (<code class="docutils literal notranslate"><span class="pre">cluster</span></code>).</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The dataset as a <a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a>.
This uses caching to avoid recomputation.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">get_test_rna_data_se</span><span class="p">()</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.model_gene_variances">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">model_gene_variances</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block_average_policy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'mean'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block_weight_policy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'variable'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variable_block_weight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(0,</span> <span class="pre">1000)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block_quantile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mean_filter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_mean</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transform</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">span</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_min_width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_window_count</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">200</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_model_gene_variances.html#model_gene_variances"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.model_gene_variances" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the variance in (log-)expression values for each gene, and model the trend in the variances with respect to the mean.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Any" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a></span>) – A matrix-like object where rows correspond to genes or genomic features and columns correspond to cells.
It is typically expected to contain log-expression values, e.g., from <a class="reference internal" href="#scranpy.normalize_counts" title="scranpy.normalize_counts"><code class="xref py py-func docutils literal notranslate"><span class="pre">normalize_counts()</span></code></a>.</p></li>
<li><p><strong>block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Array of length equal to the number of columns of <code class="docutils literal notranslate"><span class="pre">x</span></code>, containing the block of origin (e.g., batch, sample) for each cell.
Alternatively <code class="docutils literal notranslate"><span class="pre">None</span></code>, if all cells are from the same block.</p></li>
<li><p><strong>block_average_policy</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Literal" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'mean'</span></code>, <code class="docutils literal notranslate"><span class="pre">'quantile'</span></code>]</span>) – Policy for averaging statistics across blocks.
This can either use a (weighted) <code class="docutils literal notranslate"><span class="pre">mean</span></code> or a <code class="docutils literal notranslate"><span class="pre">quantile</span></code>.</p></li>
<li><p><strong>block_weight_policy</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Literal" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'variable'</span></code>, <code class="docutils literal notranslate"><span class="pre">'equal'</span></code>, <code class="docutils literal notranslate"><span class="pre">'none'</span></code>]</span>) – Policy for weighting different blocks when computing the weighted mean across blocks for each statistic.
Only used if <code class="docutils literal notranslate"><span class="pre">block</span></code> is provided and <code class="docutils literal notranslate"><span class="pre">block_average_policy</span> <span class="pre">=</span> <span class="pre">&quot;mean&quot;</span></code>.</p></li>
<li><p><strong>variable_block_weight</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Tuple" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code></a></span>) – Parameters for variable block weighting.
This should be a tuple of length 2 where the first and second values are used as the lower and upper bounds, respectively, for the variable weight calculation.
Only used if <code class="docutils literal notranslate"><span class="pre">block</span></code> is provided, <code class="docutils literal notranslate"><span class="pre">block_average_policy</span> <span class="pre">=</span> <span class="pre">&quot;mean&quot;</span></code>, and <code class="docutils literal notranslate"><span class="pre">block_weight_policy</span> <span class="pre">=</span> <span class="pre">&quot;variable&quot;</span></code>.</p></li>
<li><p><strong>block_quantile</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Probability for computing the quantile across blocks.
Defaults to 0.5, i.e., the median of per-block statistics.
Only used if <code class="docutils literal notranslate"><span class="pre">block</span></code> is provided and <code class="docutils literal notranslate"><span class="pre">block_average_policy</span> <span class="pre">=</span> <span class="pre">&quot;quantile&quot;</span></code>.</p></li>
<li><p><strong>mean_filter</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to filter on the means before trend fitting.</p></li>
<li><p><strong>min_mean</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – The minimum mean of genes to use in trend fitting.
Only used if <code class="docutils literal notranslate"><span class="pre">mean_filter</span> <span class="pre">=</span> <span class="pre">True</span></code>.</p></li>
<li><p><strong>transform</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether a quarter-root transformation should be applied before trend fitting.</p></li>
<li><p><strong>span</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Span of the LOWESS smoother for trend fitting, see <a class="reference internal" href="#scranpy.fit_variance_trend" title="scranpy.fit_variance_trend"><code class="xref py py-func docutils literal notranslate"><span class="pre">fit_variance_trend()</span></code></a>.</p></li>
<li><p><strong>use_min_width</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether a minimum width constraint should be applied during trend fitting, see <a class="reference internal" href="#scranpy.fit_variance_trend" title="scranpy.fit_variance_trend"><code class="xref py py-func docutils literal notranslate"><span class="pre">fit_variance_trend()</span></code></a>.</p></li>
<li><p><strong>min_width</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Minimum width of the smoothing window for trend fitting, see <a class="reference internal" href="#scranpy.fit_variance_trend" title="scranpy.fit_variance_trend"><code class="xref py py-func docutils literal notranslate"><span class="pre">fit_variance_trend()</span></code></a>.</p></li>
<li><p><strong>min_window_count</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Minimum number of observations in each smoothing window for trend fitting, see <a class="reference internal" href="#scranpy.fit_variance_trend" title="scranpy.fit_variance_trend"><code class="xref py py-func docutils literal notranslate"><span class="pre">fit_variance_trend()</span></code></a>.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><p>A <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a> containing <code class="docutils literal notranslate"><span class="pre">statistics</span></code>.
This is a <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a> with one row per gene and the following columns:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">mean</span></code>: a double-precision NumPy array containing the mean (log-)expression for each gene.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">variance</span></code>: a double-precision NumPy array containing the mean (log-)expression for each gene.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">fitted</span></code>: a double-precision NumPy array containing the fitted value of the mean-variance trend for each gene.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">residual</span></code>: a double-precision NumPy array containing the residual from the mean-variance trend for each gene.</p></li>
</ul>
<p>If <code class="docutils literal notranslate"><span class="pre">block</span></code> is supplied, the <code class="docutils literal notranslate"><span class="pre">NamedList</span></code> will also contain:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">per_block</span></code>: a <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a> containing the per-block statistics.
Each entry is a <code class="docutils literal notranslate"><span class="pre">BiocFrame</span></code> that contains the <code class="docutils literal notranslate"><span class="pre">mean</span></code>, <code class="docutils literal notranslate"><span class="pre">variance</span></code>, <code class="docutils literal notranslate"><span class="pre">fitted</span></code> and <code class="docutils literal notranslate"><span class="pre">residual</span></code> for each block.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">block_ids</span></code>: a list containing the identities of the blocks.
This corresponds to the entries of <code class="docutils literal notranslate"><span class="pre">per_block</span></code>.</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The <code class="docutils literal notranslate"><span class="pre">model_gene_variances</span></code> function in the <a class="reference external" href="https://libscran.github.io/scran_variances">scran_variances</a> C++ library.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># mock up some log-normalized data with an interesting mean-variance relationship.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mu</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="mi">200</span><span class="p">)</span> <span class="o">*</span> <span class="mi">5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">normed</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">((</span><span class="mi">200</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">10</span><span class="p">):</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">normed</span><span class="p">[:,</span><span class="n">c</span><span class="p">]</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="n">lam</span><span class="o">=</span><span class="n">mu</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">200</span><span class="p">))</span> <span class="o">/</span> <span class="n">numpy</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">model_gene_variances</span><span class="p">(</span><span class="n">normed</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="s2">&quot;statistics&quot;</span><span class="p">])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.normalize_adt_counts_se">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">normalize_adt_counts_se</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size_factors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">center</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'lowest'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pseudo_count</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assay_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'counts'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'logcounts'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">factor_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'size_factor'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_se_normalize_adt_counts.html#normalize_adt_counts_se"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.normalize_adt_counts_se" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute (log-)normalized expression values after performing scaling normalization of an ADT count matrix.
This calls <a class="reference internal" href="#scranpy.compute_clrm1_factors" title="scranpy.compute_clrm1_factors"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_clrm1_factors()</span></code></a> to compute CLRm1 size factors,
<a class="reference internal" href="#scranpy.center_size_factors" title="scranpy.center_size_factors"><code class="xref py py-func docutils literal notranslate"><span class="pre">center_size_factors()</span></code></a> to center the size factors,
and then <a class="reference internal" href="#scranpy.normalize_counts" title="scranpy.normalize_counts"><code class="xref py py-func docutils literal notranslate"><span class="pre">normalize_counts()</span></code></a> to compute normalized log-expression values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a></span>) – A <a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a> object or one of its subclasses.
Rows correspond to antibody-derived tags (ADTs) and columns correspond to cells.</p></li>
<li><p><strong>size_factors</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Size factor for each cell in <code class="docutils literal notranslate"><span class="pre">x</span></code>, to be passed to <a class="reference internal" href="#scranpy.normalize_counts" title="scranpy.normalize_counts"><code class="xref py py-func docutils literal notranslate"><span class="pre">normalize_counts()</span></code></a>.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, size factors are computed with <a class="reference internal" href="#scranpy.compute_clrm1_factors" title="scranpy.compute_clrm1_factors"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_clrm1_factors()</span></code></a>.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads, to be passed to <a class="reference internal" href="#scranpy.normalize_counts" title="scranpy.normalize_counts"><code class="xref py py-func docutils literal notranslate"><span class="pre">normalize_counts()</span></code></a>.</p></li>
<li><p><strong>center</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to center the <code class="docutils literal notranslate"><span class="pre">size_factors</span></code> by passing them to <a class="reference internal" href="#scranpy.center_size_factors" title="scranpy.center_size_factors"><code class="xref py py-func docutils literal notranslate"><span class="pre">center_size_factors()</span></code></a>.</p></li>
<li><p><strong>block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Block assignment for each cell in <code class="docutils literal notranslate"><span class="pre">x</span></code>, to be passed to <a class="reference internal" href="#scranpy.center_size_factors" title="scranpy.center_size_factors"><code class="xref py py-func docutils literal notranslate"><span class="pre">center_size_factors()</span></code></a>.</p></li>
<li><p><strong>mode</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Literal" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'lowest'</span></code>, <code class="docutils literal notranslate"><span class="pre">'per-block'</span></code>]</span>) – How to scale the size factors across blocks, see <a class="reference internal" href="#scranpy.center_size_factors" title="scranpy.center_size_factors"><code class="xref py py-func docutils literal notranslate"><span class="pre">center_size_factors()</span></code></a>.
Only relevant if <code class="docutils literal notranslate"><span class="pre">block</span></code> is provided.</p></li>
<li><p><strong>log</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to log-transform the normalized expression values, see <a class="reference internal" href="#scranpy.normalize_counts" title="scranpy.normalize_counts"><code class="xref py py-func docutils literal notranslate"><span class="pre">normalize_counts()</span></code></a>.</p></li>
<li><p><strong>pseudo_count</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Pseudo-count for the log-transformation, see <a class="reference internal" href="#scranpy.normalize_counts" title="scranpy.normalize_counts"><code class="xref py py-func docutils literal notranslate"><span class="pre">normalize_counts()</span></code></a>.
Only relevant if <code class="docutils literal notranslate"><span class="pre">log</span> <span class="pre">=</span> <span class="pre">True</span></code>.</p></li>
<li><p><strong>assay_type</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – Name or index specifying the assay of <code class="docutils literal notranslate"><span class="pre">x</span></code> that contains the count matrix to be normalized.</p></li>
<li><p><strong>output_name</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></span>) – Name of the assay in which to store the normalized matrix in the output object.</p></li>
<li><p><strong>factor_name</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – Name of the column of the column data in which to store the size factors in the output object.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, the size factors are not stored.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A copy of <code class="docutils literal notranslate"><span class="pre">x</span></code>, with an additional assay containing the (log-)normalized matrix.
Size factors are also stored in the column data.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">get_test_adt_data_se</span><span class="p">(</span><span class="s2">&quot;qc&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">get_alternative_experiment</span><span class="p">(</span><span class="s2">&quot;ADT&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">normalize_adt_counts_se</span><span class="p">(</span><span class="n">sce</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span><span class="o">.</span><span class="n">get_assay_names</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">()[</span><span class="s2">&quot;size_factor&quot;</span><span class="p">]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.normalize_counts">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">normalize_counts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size_factors</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pseudo_count</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_base</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">preserve_sparsity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">delayed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.normalize_counts" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a matrix of (log-transformed) normalized expression values.
The normalization removes uninteresting per-cell differences due to sequencing efficiency and library size.
The log-transformation ensures that any differences represent log-fold changes in downstream analysis steps; such relative changes in expression are more relevant than absolute changes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Any" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a></span>) – <p>Matrix-like object containing cells in columns and features in rows, typically with count data.</p>
<p>Alternatively, a <a class="reference external" href="https://tatami-inc.github.io/mattress/api/mattress.html#mattress.InitializedMatrix.InitializedMatrix" title="(in mattress v0.4.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">InitializedMatrix</span></code></a> representing a count matrix, typically created by <a class="reference external" href="https://tatami-inc.github.io/mattress/api/mattress.html#mattress.initialize.initialize" title="(in mattress v0.4.2)"><code class="xref py py-func docutils literal notranslate"><span class="pre">initialize()</span></code></a>.</p>
</p></li>
<li><p><strong>size_factors</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a></span>) – Size factor for each cell. This should have length equal to the number of columns in <code class="docutils literal notranslate"><span class="pre">x</span></code>.</p></li>
<li><p><strong>log</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether log-transformation should be performed.</p></li>
<li><p><strong>pseudo_count</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Positive pseudo-count to add before log-transformation.
Ignored if <code class="docutils literal notranslate"><span class="pre">log</span> <span class="pre">=</span> <span class="pre">False</span></code>.</p></li>
<li><p><strong>log_base</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Base of the log-transformation, ignored if <code class="docutils literal notranslate"><span class="pre">log</span> <span class="pre">=</span> <span class="pre">False</span></code>.</p></li>
<li><p><strong>preserve_sparsity</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to preserve sparsity when <code class="docutils literal notranslate"><span class="pre">pseudo_count</span> <span class="pre">!=</span> <span class="pre">1</span></code>.
If <code class="docutils literal notranslate"><span class="pre">True</span></code>, users should manually add <code class="docutils literal notranslate"><span class="pre">log(pseudo_count,</span> <span class="pre">log_base)</span></code> to the returned matrix to obtain the desired log-transformed expression values.
Ignored if <code class="docutils literal notranslate"><span class="pre">log</span> <span class="pre">=</span> <span class="pre">False</span></code> or <code class="docutils literal notranslate"><span class="pre">pseudo_count</span> <span class="pre">=</span> <span class="pre">1</span></code>.</p></li>
<li><p><strong>delayed</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether operations on a matrix-like <code class="docutils literal notranslate"><span class="pre">x</span></code> should be delayed.
This improves memory efficiency at the cost of some speed in downstream operations.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://biocpy.github.io/DelayedArray/api/delayedarray.html#delayedarray.DelayedArray.DelayedArray" title="(in DelayedArray v0.6.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">DelayedArray</span></code></a>, <a class="reference external" href="https://tatami-inc.github.io/mattress/api/mattress.html#mattress.InitializedMatrix.InitializedMatrix" title="(in mattress v0.4.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">InitializedMatrix</span></code></a>]</span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><p>If <code class="docutils literal notranslate"><span class="pre">x</span></code> is a matrix-like object and <code class="docutils literal notranslate"><span class="pre">delayed</span> <span class="pre">=</span> <span class="pre">True</span></code>, a <a class="reference external" href="https://biocpy.github.io/DelayedArray/api/delayedarray.html#delayedarray.DelayedArray.DelayedArray" title="(in DelayedArray v0.6.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">DelayedArray</span></code></a> is returned containing the (log-transformed) normalized expression matrix.
If <code class="docutils literal notranslate"><span class="pre">delayed</span> <span class="pre">=</span> <span class="pre">False</span></code>, the type of the (log-)normalized matrix will depend on the operations applied to <code class="docutils literal notranslate"><span class="pre">x</span></code>.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">x</span></code> is an <code class="docutils literal notranslate"><span class="pre">InitializedMatrix</span></code>, a new <code class="docutils literal notranslate"><span class="pre">InitializedMatrix</span></code> is returned containing the normalized expression matrix.</p>
</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The <code class="docutils literal notranslate"><span class="pre">normalize_counts</span></code> function in the <a class="reference external" href="https://libscran.github.io/scran_norm">scran_norm</a> C++ library, for the rationale behind normalization and log-transformation.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="n">lam</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">1000</span><span class="p">)</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="mi">50</span><span class="p">,</span> <span class="mi">20</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sf</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">center_size_factors</span><span class="p">(</span><span class="n">counts</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">normed</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">normalize_counts</span><span class="p">(</span><span class="n">counts</span><span class="p">,</span> <span class="n">sf</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">normed</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.normalize_crispr_counts_se">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">normalize_crispr_counts_se</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size_factors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">center</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'lowest'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pseudo_count</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assay_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'counts'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'logcounts'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">factor_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'size_factor'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_se_normalize_crispr_counts.html#normalize_crispr_counts_se"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.normalize_crispr_counts_se" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute (log-)normalized expression values after performing scaling normalization of a CRISPR guide count matrix.
This calls <a class="reference internal" href="#scranpy.center_size_factors" title="scranpy.center_size_factors"><code class="xref py py-func docutils literal notranslate"><span class="pre">center_size_factors()</span></code></a> to center the library sizes,
and then <a class="reference internal" href="#scranpy.normalize_counts" title="scranpy.normalize_counts"><code class="xref py py-func docutils literal notranslate"><span class="pre">normalize_counts()</span></code></a> to compute normalized log-expression values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a></span>) – A <a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a> object or one of its subclasses.
Rows correspond to guides and columns correspond to cells.</p></li>
<li><p><strong>size_factors</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Size factor for each cell in <code class="docutils literal notranslate"><span class="pre">x</span></code>, to be passed to <a class="reference internal" href="#scranpy.normalize_counts" title="scranpy.normalize_counts"><code class="xref py py-func docutils literal notranslate"><span class="pre">normalize_counts()</span></code></a>.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, this defaults to the column sums of the count matrix in <code class="docutils literal notranslate"><span class="pre">x</span></code>.</p></li>
<li><p><strong>center</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to center the <code class="docutils literal notranslate"><span class="pre">size_factors</span></code> by passing them to <a class="reference internal" href="#scranpy.center_size_factors" title="scranpy.center_size_factors"><code class="xref py py-func docutils literal notranslate"><span class="pre">center_size_factors()</span></code></a>.</p></li>
<li><p><strong>block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Block assignment for each cell in <code class="docutils literal notranslate"><span class="pre">x</span></code>, to be passed to <a class="reference internal" href="#scranpy.center_size_factors" title="scranpy.center_size_factors"><code class="xref py py-func docutils literal notranslate"><span class="pre">center_size_factors()</span></code></a>.</p></li>
<li><p><strong>mode</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Literal" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'lowest'</span></code>, <code class="docutils literal notranslate"><span class="pre">'per-block'</span></code>]</span>) – How to scale the size factors across blocks, see <a class="reference internal" href="#scranpy.center_size_factors" title="scranpy.center_size_factors"><code class="xref py py-func docutils literal notranslate"><span class="pre">center_size_factors()</span></code></a>.
Only relevant if <code class="docutils literal notranslate"><span class="pre">block</span></code> is provided.</p></li>
<li><p><strong>log</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to log-transform the normalized expression values, see <a class="reference internal" href="#scranpy.normalize_counts" title="scranpy.normalize_counts"><code class="xref py py-func docutils literal notranslate"><span class="pre">normalize_counts()</span></code></a>.</p></li>
<li><p><strong>pseudo_count</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Pseudo-count for the log-transformation, see <a class="reference internal" href="#scranpy.normalize_counts" title="scranpy.normalize_counts"><code class="xref py py-func docutils literal notranslate"><span class="pre">normalize_counts()</span></code></a>.
Only relevant if <code class="docutils literal notranslate"><span class="pre">log</span> <span class="pre">=</span> <span class="pre">True</span></code>.</p></li>
<li><p><strong>assay_type</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>) – Name or index specifying the assay of <code class="docutils literal notranslate"><span class="pre">x</span></code> that contains the count matrix to be normalized.</p></li>
<li><p><strong>output_name</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></span>) – Name of the assay in which to store the normalized matrix in the output object.</p></li>
<li><p><strong>factor_name</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – Name of the column of the column data in which to store the size factors in the output object.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, the size factors are not stored.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A copy of <code class="docutils literal notranslate"><span class="pre">x</span></code>, with an additional assay containing the (log-)normalized matrix.
Size factors are also stored in the column data.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">get_test_crispr_data_se</span><span class="p">(</span><span class="s2">&quot;qc&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">get_alternative_experiment</span><span class="p">(</span><span class="s2">&quot;CRISPR Guide Capture&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">normalize_crispr_counts_se</span><span class="p">(</span><span class="n">sce</span><span class="p">,</span> <span class="n">size_factors</span><span class="o">=</span><span class="n">sce</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">()[</span><span class="s2">&quot;sum&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span><span class="o">.</span><span class="n">get_assay_names</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">()[</span><span class="s2">&quot;size_factor&quot;</span><span class="p">]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.normalize_rna_counts_se">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">normalize_rna_counts_se</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size_factors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">center</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'lowest'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pseudo_count</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assay_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'counts'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'logcounts'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">factor_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'size_factor'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_se_normalize_rna_counts.html#normalize_rna_counts_se"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.normalize_rna_counts_se" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute (log-)normalized expression values after performing scaling normalization of an RNA count matrix.
This calls <a class="reference internal" href="#scranpy.center_size_factors" title="scranpy.center_size_factors"><code class="xref py py-func docutils literal notranslate"><span class="pre">center_size_factors()</span></code></a> to center the library sizes,
and then <a class="reference internal" href="#scranpy.normalize_counts" title="scranpy.normalize_counts"><code class="xref py py-func docutils literal notranslate"><span class="pre">normalize_counts()</span></code></a> to compute normalized log-expression values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a></span>) – A <a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a> object or one of its subclasses.
Rows correspond to genes and columns correspond to cells.</p></li>
<li><p><strong>size_factors</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Size factor for each cell in <code class="docutils literal notranslate"><span class="pre">x</span></code>, to be passed to <a class="reference internal" href="#scranpy.normalize_counts" title="scranpy.normalize_counts"><code class="xref py py-func docutils literal notranslate"><span class="pre">normalize_counts()</span></code></a>.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, this defaults to the column sums of the count matrix in <code class="docutils literal notranslate"><span class="pre">x</span></code>.</p></li>
<li><p><strong>center</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to center the <code class="docutils literal notranslate"><span class="pre">size_factors</span></code> by passing them to <a class="reference internal" href="#scranpy.center_size_factors" title="scranpy.center_size_factors"><code class="xref py py-func docutils literal notranslate"><span class="pre">center_size_factors()</span></code></a>.</p></li>
<li><p><strong>block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Block assignment for each cell in <code class="docutils literal notranslate"><span class="pre">x</span></code>, to be passed to <a class="reference internal" href="#scranpy.center_size_factors" title="scranpy.center_size_factors"><code class="xref py py-func docutils literal notranslate"><span class="pre">center_size_factors()</span></code></a>.</p></li>
<li><p><strong>mode</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Literal" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'lowest'</span></code>, <code class="docutils literal notranslate"><span class="pre">'per-block'</span></code>]</span>) – How to scale the size factors across blocks, see <a class="reference internal" href="#scranpy.center_size_factors" title="scranpy.center_size_factors"><code class="xref py py-func docutils literal notranslate"><span class="pre">center_size_factors()</span></code></a>.
Only relevant if <code class="docutils literal notranslate"><span class="pre">block</span></code> is provided.</p></li>
<li><p><strong>log</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to log-transform the normalized expression values, see <a class="reference internal" href="#scranpy.normalize_counts" title="scranpy.normalize_counts"><code class="xref py py-func docutils literal notranslate"><span class="pre">normalize_counts()</span></code></a>.</p></li>
<li><p><strong>pseudo_count</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Pseudo-count for the log-transformation, see <a class="reference internal" href="#scranpy.normalize_counts" title="scranpy.normalize_counts"><code class="xref py py-func docutils literal notranslate"><span class="pre">normalize_counts()</span></code></a>.
Only relevant if <code class="docutils literal notranslate"><span class="pre">log</span> <span class="pre">=</span> <span class="pre">True</span></code>.</p></li>
<li><p><strong>assay_type</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>) – Name or index specifying the assay of <code class="docutils literal notranslate"><span class="pre">x</span></code> that contains the count matrix to be normalized.</p></li>
<li><p><strong>output_name</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></span>) – Name of the assay in which to store the normalized matrix in the output object.</p></li>
<li><p><strong>factor_name</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – Name of the column of the column data in which to store the size factors in the output object.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, the size factors are not stored.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A copy of <code class="docutils literal notranslate"><span class="pre">x</span></code>, with an additional assay containing the (log-)normalized matrix.
Size factors are also stored in the column data.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">get_test_rna_data_se</span><span class="p">(</span><span class="s2">&quot;qc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">normalize_rna_counts_se</span><span class="p">(</span><span class="n">sce</span><span class="p">,</span> <span class="n">size_factors</span><span class="o">=</span><span class="n">sce</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">()[</span><span class="s2">&quot;sum&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span><span class="o">.</span><span class="n">get_assay_names</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">()[</span><span class="s2">&quot;size_factor&quot;</span><span class="p">]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.preview_markers">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">preview_markers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">columns</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['mean',</span> <span class="pre">'detected',</span> <span class="pre">('lfc',</span> <span class="pre">'delta_mean_mean')]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order_by</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_order_by</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.preview_markers" title="Permalink to this definition">¶</a></dt>
<dd><p>Preview the top markers for a group in a pretty format.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a></span>) – <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a> containing the marker statistics for a single group.</p></li>
<li><p><strong>columns</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – <p>Names of columns of <code class="docutils literal notranslate"><span class="pre">df</span></code> to retain in the preview.</p>
<p>Alternatively, each entry may be a tuple of two strings.
The first string is the name of the column in the output <code class="docutils literal notranslate"><span class="pre">BiocFrame</span></code>, and the second string is the name of the column of <code class="docutils literal notranslate"><span class="pre">df</span></code> to retain.</p>
</p></li>
<li><p><strong>rows</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>) – Number of rows to show.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, all rows are returned.</p></li>
<li><p><strong>order_by</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/constants.html#None" title="(in Python v3.14)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]</span>) – Name of the column to use for ordering the rows of the output <code class="xref py py-class docutils literal notranslate">
<span class="pre">Alternatively</span> <span class="pre">``True`</span></code>, in which case a column is automatically chosen from the effect size summaries.
If <code class="docutils literal notranslate"><span class="pre">None</span></code> or <code class="docutils literal notranslate"><span class="pre">False</span></code>, no ordering is performed.</p></li>
<li><p><strong>include_order_by</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to include the column named by <code class="docutils literal notranslate"><span class="pre">order_by</span></code> in the output <code class="docutils literal notranslate"><span class="pre">BiocFrame</span></code>.
This may also be a string, which is treated as <code class="docutils literal notranslate"><span class="pre">True</span></code>;
the value is used as the name of the column in the output <code class="docutils literal notranslate"><span class="pre">BiocFrame</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a> containing important columns for the top markers.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">get_test_rna_data_se</span><span class="p">(</span><span class="s2">&quot;cluster&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">markers</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">score_markers_se</span><span class="p">(</span><span class="n">sce</span><span class="p">,</span> <span class="n">sce</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">()[</span><span class="s2">&quot;clusters&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">scranpy</span><span class="o">.</span><span class="n">preview_markers</span><span class="p">(</span><span class="n">markers</span><span class="p">[</span><span class="s2">&quot;1&quot;</span><span class="p">]))</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.quick_adt_qc_se">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">quick_adt_qc_se</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subsets</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">more_suggest_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assay_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'counts'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">meta_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'qc'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flatten</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.quick_adt_qc_se" title="Permalink to this definition">¶</a></dt>
<dd><p>Quickly compute quality control (QC) metrics, thresholds and filters from ADT data in a <a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>.
This calls <a class="reference internal" href="#scranpy.compute_adt_qc_metrics" title="scranpy.compute_adt_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_adt_qc_metrics()</span></code></a>, <a class="reference internal" href="#scranpy.suggest_adt_qc_thresholds" title="scranpy.suggest_adt_qc_thresholds"><code class="xref py py-func docutils literal notranslate"><span class="pre">suggest_adt_qc_thresholds()</span></code></a>, and <a class="reference internal" href="#scranpy.filter_adt_qc_metrics" title="scranpy.filter_adt_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">filter_adt_qc_metrics()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a></span>) – A <a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a> or one of its subclasses.
Rows correspond to antibody-derived tags (ADTs) and columns correspond to cells.</p></li>
<li><p><strong>subsets</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Mapping" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Mapping</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – List of subsets of control features, passed to <a class="reference internal" href="#scranpy.compute_adt_qc_metrics" title="scranpy.compute_adt_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_adt_qc_metrics()</span></code></a>.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads, passed to <a class="reference internal" href="#scranpy.compute_adt_qc_metrics" title="scranpy.compute_adt_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_adt_qc_metrics()</span></code></a>.</p></li>
<li><p><strong>more_suggest_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Additional arguments to pass to <a class="reference internal" href="#scranpy.suggest_adt_qc_thresholds" title="scranpy.suggest_adt_qc_thresholds"><code class="xref py py-func docutils literal notranslate"><span class="pre">suggest_adt_qc_thresholds()</span></code></a>.</p></li>
<li><p><strong>block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Blocking factor specifying the block of origin (e.g., batch, sample) for each cell in <code class="docutils literal notranslate"><span class="pre">metrics</span></code>.
If supplied, a separate threshold is computed from the cells in each block.
Alternatively <code class="docutils literal notranslate"><span class="pre">None</span></code>, if all cells are from the same block.</p></li>
<li><p><strong>assay_type</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – Index or name of the assay of <code class="docutils literal notranslate"><span class="pre">x</span></code> containing the ADT count matrix.</p></li>
<li><p><strong>output_prefix</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – Prefix to add to the column names of the column data containing the output QC statistics.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, no prefix is added.</p></li>
<li><p><strong>meta_name</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – Name of the metadata entry in which to store additional outputs like the filtering thresholds.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, additional outputs are not reported.</p></li>
<li><p><strong>flatten</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to flatten the subset proportions into separate columns of the column data.
If <code class="docutils literal notranslate"><span class="pre">False</span></code>, the subset proportions are stored in a nested <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A copy of <code class="docutils literal notranslate"><span class="pre">x</span></code> with additional columns added to its column data.
Each column contains per-cell values for one of the ADT-related QC metrics, see <a class="reference internal" href="#scranpy.compute_adt_qc_metrics" title="scranpy.compute_adt_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_adt_qc_metrics()</span></code></a> for details.
The suggested thresholds are stored as a list in the metadata.
The column data also contains a <code class="docutils literal notranslate"><span class="pre">keep</span></code> column, specifying which cells are to be retained.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">get_test_adt_data_se</span><span class="p">()</span><span class="o">.</span><span class="n">get_alternative_experiment</span><span class="p">(</span><span class="s2">&quot;ADT&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">is_igg</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">y</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s2">&quot;IgG&quot;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">sce</span><span class="o">.</span><span class="n">get_row_names</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">quick_adt_qc_se</span><span class="p">(</span><span class="n">sce</span><span class="p">,</span> <span class="n">subsets</span><span class="o">=</span><span class="p">{</span> <span class="s2">&quot;igg&quot;</span><span class="p">:</span> <span class="n">is_igg</span> <span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">sce</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">()[:,[</span><span class="s2">&quot;sum&quot;</span><span class="p">,</span> <span class="s2">&quot;detected&quot;</span><span class="p">,</span> <span class="s2">&quot;subset_sum_igg&quot;</span><span class="p">]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">sce</span><span class="o">.</span><span class="n">get_metadata</span><span class="p">()[</span><span class="s2">&quot;qc&quot;</span><span class="p">][</span><span class="s2">&quot;thresholds&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">biocutils</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">biocutils</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">sce</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">()[</span><span class="s2">&quot;keep&quot;</span><span class="p">]))</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.quick_crispr_qc_se">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">quick_crispr_qc_se</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">more_suggest_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assay_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'counts'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">meta_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'qc'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.quick_crispr_qc_se" title="Permalink to this definition">¶</a></dt>
<dd><p>Quickly compute quality control (QC) metrics, thresholds and filters from CRISPR data in a <a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>.
This calls <a class="reference internal" href="#scranpy.compute_crispr_qc_metrics" title="scranpy.compute_crispr_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_crispr_qc_metrics()</span></code></a>, <a class="reference internal" href="#scranpy.suggest_crispr_qc_thresholds" title="scranpy.suggest_crispr_qc_thresholds"><code class="xref py py-func docutils literal notranslate"><span class="pre">suggest_crispr_qc_thresholds()</span></code></a>, and <a class="reference internal" href="#scranpy.filter_crispr_qc_metrics" title="scranpy.filter_crispr_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">filter_crispr_qc_metrics()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a></span>) – A <a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a> or one of its subclasses.
Rows correspond to antibody-derived tags (ADTs) and columns correspond to cells.</p></li>
<li><p><strong>subsets</strong> – Passed to <a class="reference internal" href="#scranpy.compute_crispr_qc_metrics" title="scranpy.compute_crispr_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_crispr_qc_metrics()</span></code></a>.</p></li>
<li><p><strong>more_suggest_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Additional arguments to pass to <a class="reference internal" href="#scranpy.suggest_crispr_qc_thresholds" title="scranpy.suggest_crispr_qc_thresholds"><code class="xref py py-func docutils literal notranslate"><span class="pre">suggest_crispr_qc_thresholds()</span></code></a>.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Passed to <a class="reference internal" href="#scranpy.compute_crispr_qc_metrics" title="scranpy.compute_crispr_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_crispr_qc_metrics()</span></code></a>.</p></li>
<li><p><strong>block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Blocking factor specifying the block of origin (e.g., batch, sample) for each cell in <code class="docutils literal notranslate"><span class="pre">metrics</span></code>.
If supplied, a separate threshold is computed from the cells in each block.
Alternatively <code class="docutils literal notranslate"><span class="pre">None</span></code>, if all cells are from the same block.</p></li>
<li><p><strong>assay_type</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – Index or name of the assay of <code class="docutils literal notranslate"><span class="pre">x</span></code> containing the ADT count matrix.</p></li>
<li><p><strong>output_prefix</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – Prefix to add to the column names of the column data containing the output QC statistics.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, no prefix is added.</p></li>
<li><p><strong>meta_name</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></span>) – Name of the metadata entry in which to store additional outputs like the filtering thresholds.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, additional outputs are not reported.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><code class="docutils literal notranslate"><span class="pre">x</span></code>, with additional columns added to its column data.
Each column contains per-cell values for one of the ADT-related QC metrics, see <a class="reference internal" href="#scranpy.compute_crispr_qc_metrics" title="scranpy.compute_crispr_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_crispr_qc_metrics()</span></code></a> for details.
The suggested thresholds are stored as a list in the metadata.
The column data also contains a <code class="docutils literal notranslate"><span class="pre">keep</span></code> column, specifying which cells are to be retained.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">get_test_crispr_data_se</span><span class="p">()</span><span class="o">.</span><span class="n">get_alternative_experiment</span><span class="p">(</span><span class="s2">&quot;CRISPR Guide Capture&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">quick_crispr_qc_se</span><span class="p">(</span><span class="n">sce</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">sce</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">()[:,[</span><span class="s2">&quot;sum&quot;</span><span class="p">,</span> <span class="s2">&quot;detected&quot;</span><span class="p">,</span> <span class="s2">&quot;max_value&quot;</span><span class="p">,</span> <span class="s2">&quot;max_index&quot;</span><span class="p">]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">sce</span><span class="o">.</span><span class="n">get_metadata</span><span class="p">()[</span><span class="s2">&quot;qc&quot;</span><span class="p">][</span><span class="s2">&quot;thresholds&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">()[</span><span class="s2">&quot;keep&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.quick_rna_qc_se">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">quick_rna_qc_se</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subsets</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">altexp_proportions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">more_suggest_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assay_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'counts'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">meta_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'qc'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flatten</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.quick_rna_qc_se" title="Permalink to this definition">¶</a></dt>
<dd><p>Quickly compute quality control (QC) metrics, thresholds and filters from RNA data in a <a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>.
This calls <a class="reference internal" href="#scranpy.compute_rna_qc_metrics" title="scranpy.compute_rna_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_rna_qc_metrics()</span></code></a>, <a class="reference internal" href="#scranpy.suggest_rna_qc_thresholds" title="scranpy.suggest_rna_qc_thresholds"><code class="xref py py-func docutils literal notranslate"><span class="pre">suggest_rna_qc_thresholds()</span></code></a>, and <a class="reference internal" href="#scranpy.filter_rna_qc_metrics" title="scranpy.filter_rna_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">filter_rna_qc_metrics()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a></span>) – A <a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a> or one of its subclasses.
Rows correspond to genes and columns correspond to cells.</p></li>
<li><p><strong>subsets</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Mapping" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Mapping</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – List of subsets of control genes, passed to <a class="reference internal" href="#scranpy.compute_rna_qc_metrics_with_altexps" title="scranpy.compute_rna_qc_metrics_with_altexps"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_rna_qc_metrics_with_altexps()</span></code></a>.</p></li>
<li><p><strong>altexp_proportions</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>, <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/constants.html#None" title="(in Python v3.14)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]</span>) – Indices or names of alternative experiments for which to compute QC metrics, see <a class="reference internal" href="#scranpy.compute_rna_qc_metrics_with_altexps" title="scranpy.compute_rna_qc_metrics_with_altexps"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_rna_qc_metrics_with_altexps()</span></code></a> for details.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads, passed to <a class="reference internal" href="#scranpy.compute_rna_qc_metrics_with_altexps" title="scranpy.compute_rna_qc_metrics_with_altexps"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_rna_qc_metrics_with_altexps()</span></code></a>.</p></li>
<li><p><strong>more_suggest_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Additional arguments to pass to <a class="reference internal" href="#scranpy.suggest_rna_qc_thresholds" title="scranpy.suggest_rna_qc_thresholds"><code class="xref py py-func docutils literal notranslate"><span class="pre">suggest_rna_qc_thresholds()</span></code></a>.</p></li>
<li><p><strong>block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Blocking factor specifying the block of origin (e.g., batch, sample) for each cell in <code class="docutils literal notranslate"><span class="pre">metrics</span></code>.
If supplied, a separate threshold is computed from the cells in each block.
Alternatively <code class="docutils literal notranslate"><span class="pre">None</span></code>, if all cells are from the same block.</p></li>
<li><p><strong>assay_type</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – Index or name of the assay of <code class="docutils literal notranslate"><span class="pre">x</span></code> containing the RNA count matrix.</p></li>
<li><p><strong>output_prefix</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – Prefix to add to the column names of the column data containing the output QC statistics.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, no prefix is added.</p></li>
<li><p><strong>meta_name</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – Name of the metadata entry in which to store additional outputs like the filtering thresholds.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, additional outputs are not reported.</p></li>
<li><p><strong>flatten</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to flatten the subset proportions into separate columns of the column data.
If <code class="docutils literal notranslate"><span class="pre">False</span></code>, the subset proportions are stored in a nested <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><code class="docutils literal notranslate"><span class="pre">x</span></code>, with additional columns added to its column data.
Each column contains per-cell values for one of the RNA-related QC metrics, see <a class="reference internal" href="#scranpy.compute_rna_qc_metrics" title="scranpy.compute_rna_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_rna_qc_metrics()</span></code></a> for details.
The suggested thresholds are stored as a list in the metadata.
The column data also contains a <code class="docutils literal notranslate"><span class="pre">keep</span></code> column, specifying which cells are to be retained.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">get_test_rna_data_se</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">is_mito</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">y</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;mt-&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">sce</span><span class="o">.</span><span class="n">get_row_names</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">quick_rna_qc_se</span><span class="p">(</span><span class="n">sce</span><span class="p">,</span> <span class="n">subsets</span><span class="o">=</span><span class="p">{</span> <span class="s2">&quot;mito&quot;</span><span class="p">:</span> <span class="n">is_mito</span> <span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">sce</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">()[:,[</span><span class="s2">&quot;sum&quot;</span><span class="p">,</span> <span class="s2">&quot;detected&quot;</span><span class="p">,</span> <span class="s2">&quot;subset_proportion_mito&quot;</span><span class="p">]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">sce</span><span class="o">.</span><span class="n">get_metadata</span><span class="p">()[</span><span class="s2">&quot;qc&quot;</span><span class="p">][</span><span class="s2">&quot;thresholds&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">biocutils</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">biocutils</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">sce</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">()[</span><span class="s2">&quot;keep&quot;</span><span class="p">]))</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Computing spike-in proportions, if available.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">get_test_rna_data_se</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">quick_rna_qc_se</span><span class="p">(</span><span class="n">sce</span><span class="p">,</span> <span class="n">subsets</span><span class="o">=</span><span class="p">{</span> <span class="s2">&quot;mito&quot;</span><span class="p">:</span> <span class="n">is_mito</span> <span class="p">},</span> <span class="n">altexp_proportions</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;ERCC&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">sce</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">()[:,[</span><span class="s2">&quot;sum&quot;</span><span class="p">,</span> <span class="s2">&quot;detected&quot;</span><span class="p">,</span> <span class="s2">&quot;subset_proportion_mito&quot;</span><span class="p">,</span> <span class="s2">&quot;subset_proportion_ERCC&quot;</span><span class="p">]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">sce</span><span class="o">.</span><span class="n">get_alternative_experiment</span><span class="p">(</span><span class="s2">&quot;ERCC&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">()[:,[</span><span class="s2">&quot;sum&quot;</span><span class="p">,</span> <span class="s2">&quot;detected&quot;</span><span class="p">]])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.run_all_neighbor_steps">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">run_all_neighbor_steps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">run_umap_options={}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">run_tsne_options={}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">build_snn_graph_options={}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster_graph_options={}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nn_parameters=&lt;knncolle.annoy.AnnoyParameters</span> <span class="pre">object&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">collapse_search=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads=3</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.run_all_neighbor_steps" title="Permalink to this definition">¶</a></dt>
<dd><p>Run all steps that depend on the nearest neighbor search,
i.e., <a class="reference internal" href="#scranpy.run_tsne" title="scranpy.run_tsne"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_tsne()</span></code></a>, <a class="reference internal" href="#scranpy.run_umap" title="scranpy.run_umap"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_umap()</span></code></a>, <a class="reference internal" href="#scranpy.build_snn_graph" title="scranpy.build_snn_graph"><code class="xref py py-func docutils literal notranslate"><span class="pre">build_snn_graph()</span></code></a>, and <a class="reference internal" href="#scranpy.cluster_graph" title="scranpy.cluster_graph"><code class="xref py py-func docutils literal notranslate"><span class="pre">cluster_graph()</span></code></a>.
This builds the index once and re-uses it for the neighbor search in each step; the various steps are also run in parallel to save more time.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.classes.Index" title="(in knncolle v0.0.post1.dev1+g951fea368)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Index</span></code></a>, <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a>]</span>) – <p>Matrix of principal components where rows are cells and columns are PCs, typically produced by <a class="reference internal" href="#scranpy.run_pca" title="scranpy.run_pca"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_pca()</span></code></a>.</p>
<p>Alternatively, a <code class="xref py py-class docutils literal notranslate"><span class="pre">Index</span></code> instance containing a prebuilt search index for the cells.</p>
</p></li>
<li><p><strong>run_umap_options</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>]</span>) – Optional arguments for <a class="reference internal" href="#scranpy.run_umap" title="scranpy.run_umap"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_umap()</span></code></a>.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, UMAP is not performed.</p></li>
<li><p><strong>run_tsne_options</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>]</span>) – Optional arguments for <a class="reference internal" href="#scranpy.run_tsne" title="scranpy.run_tsne"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_tsne()</span></code></a>.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, t-SNE is not performed.</p></li>
<li><p><strong>build_snn_graph_options</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>]</span>) – Optional arguments for <a class="reference internal" href="#scranpy.build_snn_graph" title="scranpy.build_snn_graph"><code class="xref py py-func docutils literal notranslate"><span class="pre">build_snn_graph()</span></code></a>.
Ignored if <code class="docutils literal notranslate"><span class="pre">cluster_graph_options</span> <span class="pre">=</span> <span class="pre">None</span></code>.</p></li>
<li><p><strong>cluster_graph_options</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Optional arguments for <a class="reference internal" href="#scranpy.cluster_graph" title="scranpy.cluster_graph"><code class="xref py py-func docutils literal notranslate"><span class="pre">cluster_graph()</span></code></a>.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, graph-based clustering is not performed.</p></li>
<li><p><strong>nn_parameters</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.classes.Parameters" title="(in knncolle v0.0.post1.dev1+g951fea368)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Parameters</span></code></a></span>) – Parameters for the nearest-neighbor search.</p></li>
<li><p><strong>collapse_search</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to collapse the nearest-neighbor search for each step into a single search.
Steps that need fewer neighbors will use a subset of the neighbors from the collapsed search.
This is faster but may not give the same results as separate searches for some approximate search algorithms.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use for the parallel execution of UMAP, t-SNE and SNN graph construction.
This overrides the specified number of threads in the various <code class="docutils literal notranslate"><span class="pre">*_options</span></code> arguments.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><p>A <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a> containing one entry for each step.</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">run_tsne</span></code>: results of <a class="reference internal" href="#scranpy.run_tsne" title="scranpy.run_tsne"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_tsne()</span></code></a>.
Omitted if t-SNE was not performed.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">run_umap</span></code>: results of <a class="reference internal" href="#scranpy.run_tsne" title="scranpy.run_tsne"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_tsne()</span></code></a>.
Omitted if UMAP was not performed.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">build_snn_graph</span></code>: results of <a class="reference internal" href="#scranpy.build_snn_graph" title="scranpy.build_snn_graph"><code class="xref py py-func docutils literal notranslate"><span class="pre">build_snn_graph()</span></code></a>.
Omitted if graph-based clustering was not performed.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">cluster_graph</span></code>: results of <a class="reference internal" href="#scranpy.cluster_graph" title="scranpy.cluster_graph"><code class="xref py py-func docutils literal notranslate"><span class="pre">cluster_graph()</span></code></a>.
Omitted if graph-based clustering was not performed.</p></li>
</ul>
<p>If <code class="docutils literal notranslate"><span class="pre">collapse_search</span> <span class="pre">=</span> <span class="pre">False</span></code>, results should be identical to the result of running each step in serial.</p>
</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pcs</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">200</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">output</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">run_all_neighbor_steps</span><span class="p">(</span><span class="n">pcs</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">output</span><span class="p">[</span><span class="s2">&quot;run_tsne&quot;</span><span class="p">][:</span><span class="mi">5</span><span class="p">,:])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">output</span><span class="p">[</span><span class="s2">&quot;run_umap&quot;</span><span class="p">][:</span><span class="mi">5</span><span class="p">,:])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">biocutils</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">biocutils</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">output</span><span class="p">[</span><span class="s2">&quot;cluster_graph&quot;</span><span class="p">][</span><span class="s2">&quot;membership&quot;</span><span class="p">]))</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.run_all_neighbor_steps_se">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">run_all_neighbor_steps_se</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">umap_output_name='UMAP'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">more_umap_args={}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tsne_output_name='TSNE'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">more_tsne_args={}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">build_graph_name=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">more_build_graph_args={}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster_output_name='clusters'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster_meta_name=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">more_cluster_graph_args={}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nn_parameters=&lt;knncolle.annoy.AnnoyParameters</span> <span class="pre">object&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads=3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">more_neighbor_args={}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reddim_type='PCA'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_se_run_all_neighbor_steps.html#run_all_neighbor_steps_se"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.run_all_neighbor_steps_se" title="Permalink to this definition">¶</a></dt>
<dd><p>Concurrently run all steps involving a nearest-neighbor search, using the same nearest-neighbor search index built from an existing embedding.
This includes the t-SNE, UMAP and graph-based clustering.
Internally, this uses <a class="reference internal" href="#scranpy.run_all_neighbor_steps" title="scranpy.run_all_neighbor_steps"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_all_neighbor_steps()</span></code></a> on a reduced dimension entry in <a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a></span>) – A <a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a> object or one of its subclasses.
Rows correspond to genomic features and columns correspond to cells.</p></li>
<li><p><strong>umap_output_name</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – Name of the reduced dimension entry in which to store the UMAP coordinates in the output object.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, the UMAP is not computed.</p></li>
<li><p><strong>more_umap_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>]</span>) – Additional arguments for UMAP, to pass to <a class="reference internal" href="#scranpy.run_all_neighbor_steps" title="scranpy.run_all_neighbor_steps"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_all_neighbor_steps()</span></code></a> as <code class="docutils literal notranslate"><span class="pre">run_umap_options</span></code>.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, the UMAP is not computed.</p></li>
<li><p><strong>tsne_output_name</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – Name of the reduced dimension entry in which to store the t-SNE coordinates in the output object.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, the t-SNE is not computed.</p></li>
<li><p><strong>more_tsne_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>]</span>) – Additional arguments for t-SNE, to pass to <a class="reference internal" href="#scranpy.run_all_neighbor_steps" title="scranpy.run_all_neighbor_steps"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_all_neighbor_steps()</span></code></a> as <code class="docutils literal notranslate"><span class="pre">run_tsne_options</span></code>.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, the t-SNE is not computed.</p></li>
<li><p><strong>build_graph_name</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – Name of the metadata entry in which to store the nearest neighbor graph.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, the graph is not stored.</p></li>
<li><p><strong>more_build_graph_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>]</span>) – Additional arguments for graph construction, to pass to <a class="reference internal" href="#scranpy.run_all_neighbor_steps" title="scranpy.run_all_neighbor_steps"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_all_neighbor_steps()</span></code></a> as <code class="docutils literal notranslate"><span class="pre">more_build_graph_args</span></code>.</p></li>
<li><p><strong>cluster_output_name</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – Name of the column of the column data in which to store the cluster assignments.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, graph-based clustering is not performed.</p></li>
<li><p><strong>cluster_meta_name</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – Name of the metadata entry in which to store additional clustering outputs.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, additional outputs are not stored.</p></li>
<li><p><strong>more_cluster_graph_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>]</span>) – Additional arguments for graph-based clustering, to pass to <a class="reference internal" href="#scranpy.run_all_neighbor_steps" title="scranpy.run_all_neighbor_steps"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_all_neighbor_steps()</span></code></a> as <code class="docutils literal notranslate"><span class="pre">more_cluster_graph_args</span></code>.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, the graph-based clustering is not performed.</p></li>
<li><p><strong>nn_parameters</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.classes.Parameters" title="(in knncolle v0.0.post1.dev1+g951fea368)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Parameters</span></code></a></span>) – Parameters for the nearest-neighbor search.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use, passed to <a class="reference internal" href="#scranpy.run_all_neighbor_steps" title="scranpy.run_all_neighbor_steps"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_all_neighbor_steps()</span></code></a>.</p></li>
<li><p><strong>more_neighbor_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Additional arguments to pass to <a class="reference internal" href="#scranpy.run_all_neighbor_steps" title="scranpy.run_all_neighbor_steps"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_all_neighbor_steps()</span></code></a>.</p></li>
<li><p><strong>reddim_type</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>) – Name or index of the reduced dimensions of <code class="docutils literal notranslate"><span class="pre">x</span></code> on which to perform the nearest neighbor search.
Alternatively a tuple, where the first element contains the name of an alternative experiment of <code class="docutils literal notranslate"><span class="pre">x</span></code>,
and the second element contains the name/index of an embedding in that alternative experiment.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A copy of <code class="docutils literal notranslate"><span class="pre">x</span></code>, with additional coordinates in its reduced dimensions and clustering output in its column data.
Additional information may also be stored in its metadata.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">get_test_rna_data_se</span><span class="p">(</span><span class="s2">&quot;pca&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">run_all_neighbor_steps_se</span><span class="p">(</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">sce</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">more_tsne_args</span><span class="o">=</span><span class="p">{</span> <span class="s2">&quot;max_iterations&quot;</span><span class="p">:</span> <span class="mi">50</span> <span class="p">},</span> <span class="c1"># turned down for brevity</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">more_umap_args</span><span class="o">=</span><span class="p">{</span> <span class="s2">&quot;num_epochs&quot;</span><span class="p">:</span> <span class="mi">50</span> <span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span><span class="o">.</span><span class="n">get_reduced_dimension_names</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">biocutils</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">biocutils</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">sce</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">()[</span><span class="s2">&quot;clusters&quot;</span><span class="p">]))</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.run_pca">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">run_pca</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">number</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">25</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block_weight_policy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'variable'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variable_block_weight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(0,</span> <span class="pre">1000)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">components_from_residuals</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extra_work</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">iterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5489</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">realized</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_run_pca.html#run_pca"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.run_pca" title="Permalink to this definition">¶</a></dt>
<dd><p>Run a PCA on the gene-by-cell log-expression matrix to obtain a low-dimensional representation for downstream analyses.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Any" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a></span>) – A matrix-like object where rows correspond to genes or genomic features and columns correspond to cells.
Typically, the matrix is expected to contain log-expression values, and the rows should be filtered to relevant (e.g., highly variable) genes.</p></li>
<li><p><strong>number</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of PCs to retain.</p></li>
<li><p><strong>scale</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to scale all genes to have the same variance.</p></li>
<li><p><strong>block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Array of length equal to the number of columns of <code class="docutils literal notranslate"><span class="pre">x</span></code>, containing the block of origin (e.g., batch, sample) for each cell.
Alternatively <code class="docutils literal notranslate"><span class="pre">None</span></code>, if all cells are from the same block.</p></li>
<li><p><strong>block_weight_policy</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Literal" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'variable'</span></code>, <code class="docutils literal notranslate"><span class="pre">'equal'</span></code>, <code class="docutils literal notranslate"><span class="pre">'none'</span></code>]</span>) – Policy to use for weighting different blocks when computing the average for each statistic.
Only used if <code class="docutils literal notranslate"><span class="pre">block</span></code> is provided.</p></li>
<li><p><strong>variable_block_weight</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Tuple" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code></a></span>) – Parameters for variable block weighting.
This should be a tuple of length 2 where the first and second values are used as the lower and upper bounds, respectively, for the variable weight calculation.
Only used if <code class="docutils literal notranslate"><span class="pre">block</span></code> is provided and <code class="docutils literal notranslate"><span class="pre">block_weight_policy</span> <span class="pre">=</span> <span class="pre">&quot;variable&quot;</span></code>.</p></li>
<li><p><strong>components_from_residuals</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to compute the PC scores from the residuals in the presence of a blocking factor.
If <code class="docutils literal notranslate"><span class="pre">False</span></code>, the residuals are only used to compute the rotation matrix, and the original expression values of the cells are projected onto this new space.
Only used if <code class="docutils literal notranslate"><span class="pre">block</span></code> is provided.</p></li>
<li><p><strong>extra_work</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of extra dimensions for the IRLBA workspace.</p></li>
<li><p><strong>iterations</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Maximum number of restart iterations for IRLBA.</p></li>
<li><p><strong>seed</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Seed for the initial random vector in IRLBA.</p></li>
<li><p><strong>realized</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to realize <code class="docutils literal notranslate"><span class="pre">x</span></code> into an optimal memory layout for IRLBA.
This speeds up computation at the cost of increased memory usage.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><p>A <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a> containing the following entries.</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">components</span></code>: a double-precision NumPy matrix of principal component (PC) scores.
Rows are dimensions (i.e., PCs) and columns are cells.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">rotation</span></code>: a double-precision NumPy matrix containing the rotation vectors.
Rows are genes and columns are dimensions (i.e., PCs).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">variance_explained</span></code>: a double-precision NumPy array containing the variance explained by each successive PC.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">total_variance</span></code>: total variance in the input data.
Guaranteed to be greater than the sum of <code class="docutils literal notranslate"><span class="pre">variance_explained</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">center</span></code>: a double-precision NumPy array containing the mean for each gene across all cells.
If <code class="docutils literal notranslate"><span class="pre">block</span></code> was supplied, this is instead a matrix containing the mean for each gene (column) in each block of cells (row).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">block_ids</span></code>: list containing the levels of the blocking factor corresponding to each row of <code class="docutils literal notranslate"><span class="pre">center</span></code>.
Only reported if <code class="docutils literal notranslate"><span class="pre">block</span></code> was supplied.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">scale</span></code>: a double-precision NumPy arrary containing the scaling for each gene.
Only reported if <code class="docutils literal notranslate"><span class="pre">scale</span> <span class="pre">=</span> <span class="pre">True</span></code> in the function call.</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">References</p>
<p><a class="reference external" href="https://libscran.github.io/scran_pca">https://libscran.github.io/scran_pca</a>, which describes the approach in more detail.
In particular, the documentation for the <code class="docutils literal notranslate"><span class="pre">blocked_pca</span></code> function explains the blocking strategy.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">normed</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="mi">500</span><span class="p">,</span> <span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">run_pca</span><span class="p">(</span><span class="n">normed</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="s2">&quot;variance_explained&quot;</span><span class="p">]</span> <span class="o">/</span> <span class="n">res</span><span class="p">[</span><span class="s2">&quot;total_variance&quot;</span><span class="p">])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.run_pca_se">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">run_pca_se</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">features</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">number</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">25</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">more_pca_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assay_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'logcounts'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'PCA'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">meta_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'PCA'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_se_run_pca.html#run_pca_se"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.run_pca_se" title="Permalink to this definition">¶</a></dt>
<dd><p>Compact and denoise the dataset by performing PCA on the (log-)normalized expression matrix.
This calls <a class="reference internal" href="#scranpy.run_pca" title="scranpy.run_pca"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_pca()</span></code></a> on an assay of a <a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a></span>) – A <a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a> object or one of its subclasses.
Rows correspond to genomic features and columns correspond to cells.</p></li>
<li><p><strong>features</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Features of interest to use in the PCA.
This can be a sequence of row indices, row names, or booleans indicating the rows of <code class="docutils literal notranslate"><span class="pre">x</span></code> to use.
For RNA data, this is typically the <code class="docutils literal notranslate"><span class="pre">hvg</span></code> vector added to the row data of <code class="docutils literal notranslate"><span class="pre">x</span></code> by <a class="reference internal" href="#scranpy.choose_rna_hvgs_se" title="scranpy.choose_rna_hvgs_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">choose_rna_hvgs_se()</span></code></a>.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, all available features are used.</p></li>
<li><p><strong>number</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of PCs to retain, passed to <a class="reference internal" href="#scranpy.run_pca" title="scranpy.run_pca"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_pca()</span></code></a>.</p></li>
<li><p><strong>block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Block assignment of each cell, passed to <a class="reference internal" href="#scranpy.run_pca" title="scranpy.run_pca"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_pca()</span></code></a>.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads, passed to <a class="reference internal" href="#scranpy.run_pca" title="scranpy.run_pca"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_pca()</span></code></a>.</p></li>
<li><p><strong>more_pca_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Additional arguments passed to <a class="reference internal" href="#scranpy.run_pca" title="scranpy.run_pca"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_pca()</span></code></a>.</p></li>
<li><p><strong>assay_type</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – Name or index of the assay of <code class="docutils literal notranslate"><span class="pre">x</span></code> to be used for PCA.
This is typically the log-normalized expression matrix created by <a class="reference internal" href="#scranpy.normalize_rna_counts_se" title="scranpy.normalize_rna_counts_se"><code class="xref py py-func docutils literal notranslate"><span class="pre">normalize_rna_counts_se()</span></code></a> or equivalent.</p></li>
<li><p><strong>output_name</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></span>) – Name of the reduced dimension entry in which to store the PC scores in the output object.</p></li>
<li><p><strong>meta_name</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – Name of the metadata entry in which to store other PCA statistics in the output object.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A copy of <code class="docutils literal notranslate"><span class="pre">x</span></code> with the newly computed principal component scores in the reduced dimensions.
Additional outputs (e.g., rotation matrix, variance explained) are stored in the metadata.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">get_test_rna_data_se</span><span class="p">(</span><span class="s2">&quot;hvg&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">run_pca_se</span><span class="p">(</span><span class="n">sce</span><span class="p">,</span> <span class="n">features</span><span class="o">=</span><span class="n">sce</span><span class="o">.</span><span class="n">get_row_data</span><span class="p">()[</span><span class="s2">&quot;hvg&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span><span class="o">.</span><span class="n">get_reduced_dimension</span><span class="p">(</span><span class="s2">&quot;PCA&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">shape</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pcameta</span> <span class="o">=</span> <span class="n">sce</span><span class="o">.</span><span class="n">get_metadata</span><span class="p">()[</span><span class="s2">&quot;PCA&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pcameta</span><span class="p">[</span><span class="s2">&quot;variance_explained&quot;</span><span class="p">]</span> <span class="o">/</span> <span class="n">pcameta</span><span class="p">[</span><span class="s2">&quot;total_variance&quot;</span><span class="p">]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.run_tsne">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">run_tsne</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">perplexity=30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_neighbors=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">theta=1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">early_exaggeration_iterations=250</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exaggeration_factor=12</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">momentum_switch_iterations=250</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start_momentum=0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">final_momentum=0.8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">eta=200</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_depth=7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">leaf_approximation=False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_iterations=500</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed=42</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads=1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nn_parameters=&lt;knncolle.annoy.AnnoyParameters</span> <span class="pre">object&gt;</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.run_tsne" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute t-SNE coordinates to visualize similarities between cells.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a>, <a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.find_knn.FindKnnResults" title="(in knncolle v0.0.post1.dev1+g951fea368)"><code class="xref py py-class docutils literal notranslate"><span class="pre">FindKnnResults</span></code></a>, <a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.classes.Index" title="(in knncolle v0.0.post1.dev1+g951fea368)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Index</span></code></a>]</span>) – <p>Numeric matrix where rows are dimensions and columns are cells, typically containing a low-dimensional representation from, e.g., <a class="reference internal" href="#scranpy.run_pca" title="scranpy.run_pca"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_pca()</span></code></a>.</p>
<p>Alternatively, a <a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.find_knn.FindKnnResults" title="(in knncolle v0.0.post1.dev1+g951fea368)"><code class="xref py py-class docutils literal notranslate"><span class="pre">FindKnnResults</span></code></a> object containing existing neighbor search results.</p>
<p>Alternatively, a <code class="xref py py-class docutils literal notranslate"><span class="pre">Index</span></code> object.</p>
</p></li>
<li><p><strong>perplexity</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Perplexity to use in the t-SNE algorithm.
Larger values cause the embedding to focus on global structure.</p></li>
<li><p><strong>num_neighbors</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>) – Number of neighbors in the nearest-neighbor graph.
Typically derived from <code class="docutils literal notranslate"><span class="pre">perplexity</span></code> using <a class="reference internal" href="#scranpy.tsne_perplexity_to_neighbors" title="scranpy.tsne_perplexity_to_neighbors"><code class="xref py py-func docutils literal notranslate"><span class="pre">tsne_perplexity_to_neighbors()</span></code></a>.
If <code class="docutils literal notranslate"><span class="pre">x</span></code> is a <a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.find_knn.FindKnnResults" title="(in knncolle v0.0.post1.dev1+g951fea368)"><code class="xref py py-class docutils literal notranslate"><span class="pre">FindKnnResults</span></code></a> object and the number of neighbors is not equal to <code class="docutils literal notranslate"><span class="pre">num_neighbors</span></code>, a warning is raised;
this can be suppressed by setting <code class="docutils literal notranslate"><span class="pre">num_neighbors</span> <span class="pre">=</span> <span class="pre">None</span></code>.</p></li>
<li><p><strong>theta</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Approximation level for the Barnes-Hut calculation of repulsive forces.
Lower values increase accuracy at the cost of computational time.</p></li>
<li><p><strong>early_exaggeration_iterations</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of iterations of the early exaggeration phase, where the conditional probabilities are multiplied by <code class="docutils literal notranslate"><span class="pre">exaggeration_factor</span></code>.
In this phase, the empty space between clusters is increased so that clusters can easily relocate to find a good global organization.
Larger values improve convergence within this phase at the cost of reducing the remaining iterations in <code class="docutils literal notranslate"><span class="pre">max_iterations</span></code>.</p></li>
<li><p><strong>exaggeration_factor</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Exaggeration factor to scale the probabilities during the early exaggeration phase (see <code class="docutils literal notranslate"><span class="pre">early_exaggeration_iterations</span></code>).
Larger values increase the attraction between nearest neighbors to favor local structure during this phase.</p></li>
<li><p><strong>momentum_switch_iterations</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of iterations to perform before switching from the starting momentum (<code class="docutils literal notranslate"><span class="pre">start_momentum</span></code>) to the final momentum (<code class="docutils literal notranslate"><span class="pre">final_momentum</span></code>).
Greater momentum can improve convergence by increasing the step size and smoothing over local oscillations, at the risk of potentially skipping over relevant minima.</p></li>
<li><p><strong>start_momentum</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Starting momentum in [0, 1) to be used in the iterations before the momentum switch at <code class="docutils literal notranslate"><span class="pre">momentum_switch_iterations</span></code>.
This is usually lower than <code class="docutils literal notranslate"><span class="pre">final_momentum</span></code> to avoid skipping over suitable local minima.</p></li>
<li><p><strong>final_momentum</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Final momentum in [0, 1) to be used in the iterations after the momentum switch at <code class="docutils literal notranslate"><span class="pre">momentum_switch_iterations</span></code>.
This is usually higher than <code class="docutils literal notranslate"><span class="pre">start_momentum</span></code> to accelerate convergence to the local minima once the observations are moderately well-organized.</p></li>
<li><p><strong>eta</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – The learning rate, used to scale the updates to the coordinates at each iteration.
Larger values can speed up convergence at the cost of potentially skipping over local minima.</p></li>
<li><p><strong>max_depth</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Maximum depth of the Barnes-Hut quadtree.
If neighboring observations cannot be separated before the maximum depth is reached, they will be assigned to the same leaf node.
This effectively approximates each observation’s coordinates with the center of mass of its leaf node.
Smaller values (7-10) improve speed at the cost of accuracy.</p></li>
<li><p><strong>leaf_approximation</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to use the “leaf approximation” approach, which sacrifices some accuracy for greater speed.
This replaces a observation with the center of mass of its leaf node when computing the repulsive forces to all other observations.
Only effective when <code class="docutils literal notranslate"><span class="pre">max_depth</span></code> is small enough for multiple cells to be assigned to the same leaf node of the quadtree.</p></li>
<li><p><strong>max_iterations</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Maximum number of iterations to perform.</p></li>
<li><p><strong>seed</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Random seed to use for generating the initial coordinates.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.</p></li>
<li><p><strong>nn_parameters</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.classes.Parameters" title="(in knncolle v0.0.post1.dev1+g951fea368)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Parameters</span></code></a></span>) – The algorithm to use for the nearest-neighbor search.
Only used if <code class="docutils literal notranslate"><span class="pre">x</span></code> is not a pre-built nearest-neighbor search index or a list of existing nearest-neighbor search results.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Double-precision NumPy matrix containing the coordinates of each cell in a 2-dimensional embedding.
Each row corresponds to a cell and each column corresponds to a dimension.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p><a class="reference external" href="https://libscran.github.io/qdtsne">https://libscran.github.io/qdtsne</a>, for some more details on the approximations.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pcs</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="mi">500</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tout</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">run_tsne</span><span class="p">(</span><span class="n">pcs</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">tout</span><span class="p">[:</span><span class="mi">5</span><span class="p">,:])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.run_tsne_se">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">run_tsne_se</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">perplexity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">more_tsne_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reddim_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'PCA'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'TSNE'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_se_run_tsne.html#run_tsne_se"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.run_tsne_se" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a t-SNE visualization from an existing embedding.
This calls <a class="reference internal" href="#scranpy.run_tsne" title="scranpy.run_tsne"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_tsne()</span></code></a> on a reduced dimension entry in a <a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a></span>) – A <a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a> object or one of its subclasses.
Rows correspond to genomic features and columns correspond to cells.</p></li>
<li><p><strong>perplexity</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Perplexity for the t-SNE algorithm, passed to <a class="reference internal" href="#scranpy.run_tsne" title="scranpy.run_tsne"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_tsne()</span></code></a>.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads for the neighbor search and optimization, passed to <a class="reference internal" href="#scranpy.run_tsne" title="scranpy.run_tsne"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_tsne()</span></code></a>.</p></li>
<li><p><strong>more_tsne_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Additional arguments to pass to <a class="reference internal" href="#scranpy.run_tsne" title="scranpy.run_tsne"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_tsne()</span></code></a>.</p></li>
<li><p><strong>reddim_type</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – Name or index of the existing reduced dimension embedding in <code class="docutils literal notranslate"><span class="pre">x</span></code> from which to generate t-SNE coordinates.
Alternatively a tuple, where the first element contains the name of an alternative experiment of <code class="docutils literal notranslate"><span class="pre">x</span></code>,
and the second element contains the name/index of an embedding in that alternative experiment.</p></li>
<li><p><strong>output_name</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></span>) – Name of the reduced dimension entry in which to store the t-SNE coordinates in the output object.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A copy of <code class="docutils literal notranslate"><span class="pre">x</span></code> with a new reduced dimension entry containing the t-SNE coordinates.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">get_test_rna_data_se</span><span class="p">(</span><span class="s2">&quot;pca&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Using fewer iterations for a faster-running example.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">run_tsne_se</span><span class="p">(</span><span class="n">sce</span><span class="p">,</span> <span class="n">more_tsne_args</span><span class="o">=</span><span class="p">{</span> <span class="s2">&quot;max_iterations&quot;</span><span class="p">:</span> <span class="mi">50</span> <span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span><span class="o">.</span><span class="n">get_reduced_dimension</span><span class="p">(</span><span class="s2">&quot;TSNE&quot;</span><span class="p">)[:</span><span class="mi">5</span><span class="p">,:]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.run_umap">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">run_umap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_dim=2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parallel_optimization=False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">local_connectivity=1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bandwidth=1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mix_ratio=1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spread=1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_dist=0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">a=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">b=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">repulsion_strength=1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">initialize_method='spectral'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">initial_coordinates=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">initialize_random_on_spectral_fail=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">initialize_spectral_scale=10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">initialize_spectral_jitter=False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">initialize_spectral_jitter_sd=0.0001</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">initialize_random_scale=10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">initialize_seed=9876543210</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_epochs=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">learning_rate=1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">negative_sample_rate=5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_neighbors=15</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">optimize_seed=1234567890</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads=1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nn_parameters=&lt;knncolle.annoy.AnnoyParameters</span> <span class="pre">object&gt;</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.run_umap" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute UMAP coordinates to visualize similarities between cells.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a>, <a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.find_knn.FindKnnResults" title="(in knncolle v0.0.post1.dev1+g951fea368)"><code class="xref py py-class docutils literal notranslate"><span class="pre">FindKnnResults</span></code></a>, <a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.classes.Index" title="(in knncolle v0.0.post1.dev1+g951fea368)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Index</span></code></a>]</span>) – <p>Numeric matrix where rows are dimensions and columns are cells, typically containing a low-dimensional representation from, e.g., <a class="reference internal" href="#scranpy.run_pca" title="scranpy.run_pca"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_pca()</span></code></a>.</p>
<p>Alternatively, a <a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.find_knn.FindKnnResults" title="(in knncolle v0.0.post1.dev1+g951fea368)"><code class="xref py py-class docutils literal notranslate"><span class="pre">FindKnnResults</span></code></a> object containing existing neighbor search results.</p>
<p>Alternatively, a <code class="xref py py-class docutils literal notranslate"><span class="pre">Index</span></code> object.</p>
</p></li>
<li><p><strong>num_dim</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of dimensions in the UMAP embedding.</p></li>
<li><p><strong>local_connectivity</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Number of nearest neighbors that are assumed to be always connected, with maximum membership confidence.
Larger values increase the connectivity of the embedding and reduce the focus on local structure.
This may be a fractional number of neighbors, in which case interpolation is performed when computing the membership confidence.</p></li>
<li><p><strong>bandwidth</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Effective bandwidth of the kernel when converting the distance to a neighbor into a fuzzy set membership confidence.
Larger values reduce the decay in confidence with respect to distance, increasing connectivity and favoring global structure.</p></li>
<li><p><strong>mix_ratio</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Number between 0 and 1 specifying the mixing ratio when combining fuzzy sets.
A mixing ratio of 1 will take the union of confidences, a ratio of 0 will take the intersection, and intermediate values will interpolate between them.
Larger values favor connectivity and more global structure.</p></li>
<li><p><strong>spread</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Scale of the coordinates of the final low-dimensional embedding.
Ignored if <code class="docutils literal notranslate"><span class="pre">a</span></code> and <code class="docutils literal notranslate"><span class="pre">b</span></code> are provided.</p></li>
<li><p><strong>min_dist</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Minimum distance between observations in the final low-dimensional embedding.
Smaller values will increase local clustering while larger values favor a more even distribution of observations throughout the low-dimensional space.
This is interpreted relative to <code class="docutils literal notranslate"><span class="pre">spread</span></code>.
Ignored if <code class="docutils literal notranslate"><span class="pre">a</span></code> and <code class="docutils literal notranslate"><span class="pre">b</span></code> are provided.</p></li>
<li><p><strong>a</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a>]</span>) – The <code class="docutils literal notranslate"><span class="pre">a</span></code> parameter for the fuzzy set membership strength calculations.
Larger values yield a sharper decay in membership strength with increasing distance between observations.
If this or <code class="docutils literal notranslate"><span class="pre">b</span></code> are <code class="docutils literal notranslate"><span class="pre">None</span></code>, a suitable value for this parameter is automatically determined from <code class="docutils literal notranslate"><span class="pre">spread</span></code> and <code class="docutils literal notranslate"><span class="pre">min_dist</span></code>.</p></li>
<li><p><strong>b</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a>]</span>) – The <code class="docutils literal notranslate"><span class="pre">b</span></code> parameter for the fuzzy set membership strength calculations.
Larger values yield an earlier decay in membership strength with increasing distance between observations.
If this or <code class="docutils literal notranslate"><span class="pre">a</span></code> are <code class="docutils literal notranslate"><span class="pre">None</span></code>, a suitable value for this parameter is automatically determined from <code class="docutils literal notranslate"><span class="pre">spread</span></code> and <code class="docutils literal notranslate"><span class="pre">min_dist</span></code>.</p></li>
<li><p><strong>repulsion_strength</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Modifier for the repulsive force.
Larger values increase repulsion and favor local structure.</p></li>
<li><p><strong>initialize_method</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Literal" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'spectral'</span></code>, <code class="docutils literal notranslate"><span class="pre">'random'</span></code>, <code class="docutils literal notranslate"><span class="pre">'none'</span></code>]</span>) – <p>How to initialize the embedding.</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">spectral</span></code>: spectral decomposition of the normalized graph Laplacian.
Specifically, the initial coordinates are defined from the eigenvectors corresponding to the smallest non-zero eigenvalues.
This fails in the presence of multiple graph components or if the approximate SVD fails to converge.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">random</span></code>:  fills the embedding with random draws from a normal distribution.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">none</span></code>: uses initial values from <code class="docutils literal notranslate"><span class="pre">initial_coordinates</span></code>.</p></li>
</ul>
</p></li>
<li><p><strong>initialize_random_on_spectral_fail</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to fall back to random sampling (i.e., <code class="docutils literal notranslate"><span class="pre">random</span></code>) if spectral initialization fails due to the presence of multiple components in the graph.
If False, the values in <code class="docutils literal notranslate"><span class="pre">initial_coordinates</span></code> will be used instead, i.e., same as <code class="docutils literal notranslate"><span class="pre">none</span></code>.
Only relevant if <code class="docutils literal notranslate"><span class="pre">initialize_method</span> <span class="pre">=</span> <span class="pre">&quot;spectral&quot;</span></code> and spectral initialization fails.</p></li>
<li><p><strong>initialize_spectral_scale</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Maximum absolute magnitude of the coordinates after spectral initialization.
The default is chosen to avoid outlier observations with large absolute distances that may interfere with optimization.
Only relevant if <code class="docutils literal notranslate"><span class="pre">initialize_method</span> <span class="pre">=</span> <span class="pre">&quot;spectral&quot;</span></code> and spectral initialization does not fail.</p></li>
<li><p><strong>initialize_spectral_jitter</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to jitter coordinates after spectral initialization to separate duplicate observations (e.g., to avoid overplotting).
This is done using normally-distributed noise of mean zero and standard deviation of <code class="docutils literal notranslate"><span class="pre">initialize_spectral_jitter_sd</span></code>.
Only relevant if <code class="docutils literal notranslate"><span class="pre">initialize_method</span> <span class="pre">=</span> <span class="pre">&quot;spectral&quot;</span></code> and spectral initialization does not fail.</p></li>
<li><p><strong>initialize_spectral_jitter_sd</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Standard deviation of the jitter to apply after spectral initialization.
Only relevant if <code class="docutils literal notranslate"><span class="pre">initialize_method</span> <span class="pre">=</span> <span class="pre">&quot;spectral&quot;</span></code> and spectral initialization does not fail and <code class="docutils literal notranslate"><span class="pre">initialize_spectral_jitter</span> <span class="pre">=</span> <span class="pre">True</span></code>.</p></li>
<li><p><strong>initialize.random.scale</strong> – Scale of the randomly generated initial coordinates.
Coordinates are sampled from a uniform distribution from [-x, x) where x is <code class="docutils literal notranslate"><span class="pre">initialize_random_scale</span></code>.
Only relevant if <code class="docutils literal notranslate"><span class="pre">initialize_method</span> <span class="pre">=</span> <span class="pre">&quot;random&quot;</span></code>,
or <code class="docutils literal notranslate"><span class="pre">initialize_method</span> <span class="pre">=</span> <span class="pre">&quot;spectral&quot;</span></code> and spectral initialization fails and <code class="docutils literal notranslate"><span class="pre">initialize_random_on_spectral_fail</span> <span class="pre">=</span> <span class="pre">True</span></code>.</p></li>
<li><p><strong>initialize_seed</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Seed for the random number generation during initialization.
Only relevant if <code class="docutils literal notranslate"><span class="pre">initialize_method</span> <span class="pre">=</span> <span class="pre">&quot;random&quot;</span></code>,
or <code class="docutils literal notranslate"><span class="pre">initialize_method</span> <span class="pre">=</span> <span class="pre">&quot;spectral&quot;</span></code> and <code class="docutils literal notranslate"><span class="pre">initialize_spectral_jitter</span> <span class="pre">=</span> <span class="pre">True</span></code>;
or <code class="docutils literal notranslate"><span class="pre">initialize_method</span> <span class="pre">=</span> <span class="pre">&quot;spectral&quot;</span></code> and spectral initialization fails and <code class="docutils literal notranslate"><span class="pre">initialize_random_on_spectral_fail</span> <span class="pre">=</span> <span class="pre">True</span></code>.</p></li>
<li><p><strong>initial_coordinates</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<code class="xref py py-class docutils literal notranslate"><span class="pre">array</span></code>]</span>) – Double-precision matrix of initial coordinates with number of rows and columns equal to the number of observations and <code class="docutils literal notranslate"><span class="pre">num_dim</span></code>, respectively.
Only relevant if <code class="docutils literal notranslate"><span class="pre">initialize_method</span> <span class="pre">=</span> <span class="pre">&quot;none&quot;</span></code>;
or <code class="docutils literal notranslate"><span class="pre">initialize_method</span> <span class="pre">=</span> <span class="pre">&quot;spectral&quot;</span></code> and spectral initialization fails and <code class="docutils literal notranslate"><span class="pre">initialize_random_on_spectral_fail</span> <span class="pre">=</span> <span class="pre">False</span></code>.</p></li>
<li><p><strong>num_epochs</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>) – Number of epochs to perform.
If set to None, an appropriate number of epochs is chosen based on the number of points in <code class="docutils literal notranslate"><span class="pre">x</span></code>.</p></li>
<li><p><strong>num_neighbors</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of neighbors to use in the UMAP algorithm.
Larger values cause the embedding to focus on global structure.
Ignored if <code class="docutils literal notranslate"><span class="pre">x</span></code> is a <a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.find_knn.FindKnnResults" title="(in knncolle v0.0.post1.dev1+g951fea368)"><code class="xref py py-class docutils literal notranslate"><span class="pre">FindKnnResults</span></code></a> object.</p></li>
<li><p><strong>optimize_seed</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Integer scalar specifying the seed to use.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.</p></li>
<li><p><strong>nn_parameters</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.classes.Parameters" title="(in knncolle v0.0.post1.dev1+g951fea368)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Parameters</span></code></a></span>) – The algorithm to use for the nearest-neighbor search.
Only used if <code class="docutils literal notranslate"><span class="pre">x</span></code> is not a pre-built nearest-neighbor search index or a list of existing nearest-neighbor search results.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Double-precision NumPy matrix containing the coordinates of each cell in a 2-dimensional embedding.
Each row corresponds to a cell and each column corresponds to a dimension.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p><a class="reference external" href="https://libscran.github.io/umappp">https://libscran.github.io/umappp</a>, for the underlying implementation.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pcs</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="mi">500</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">uout</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">run_umap</span><span class="p">(</span><span class="n">pcs</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">uout</span><span class="p">[:</span><span class="mi">5</span><span class="p">,:])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.run_umap_se">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">run_umap_se</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_dim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_dist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_neighbors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">15</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">more_umap_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reddim_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'PCA'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'UMAP'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_se_run_umap.html#run_umap_se"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.run_umap_se" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a UMAP visualization from an existing embedding.
This calls <a class="reference internal" href="#scranpy.run_umap" title="scranpy.run_umap"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_umap()</span></code></a> on a reduced dimension entry in a <a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a></span>) – A <a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a> object or one of its subclasses.
Rows correspond to genomic features and columns correspond to cells.</p></li>
<li><p><strong>num_dim</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of dimensions in the UMAP, passed to <a class="reference internal" href="#scranpy.run_umap" title="scranpy.run_umap"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_umap()</span></code></a>.</p></li>
<li><p><strong>min_dist</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Minimum distance between cells in the UMAP, passed to <a class="reference internal" href="#scranpy.run_umap" title="scranpy.run_umap"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_umap()</span></code></a>.</p></li>
<li><p><strong>num_neighbors</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of neighbors, passed to <a class="reference internal" href="#scranpy.run_umap" title="scranpy.run_umap"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_umap()</span></code></a>.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads for the neighbor search and optimization, passed to <a class="reference internal" href="#scranpy.run_umap" title="scranpy.run_umap"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_umap()</span></code></a>.</p></li>
<li><p><strong>more_umap_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Additional arguments to pass to <a class="reference internal" href="#scranpy.run_umap" title="scranpy.run_umap"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_umap()</span></code></a>.</p></li>
<li><p><strong>reddim_type</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – Name or index of the existing reduced dimension embedding in <code class="docutils literal notranslate"><span class="pre">x</span></code> from which to generate UMAP coordinates.
Alternatively a tuple, where the first element contains the name of an alternative experiment of <code class="docutils literal notranslate"><span class="pre">x</span></code>,
and the second element contains the name/index of an embedding in that alternative experiment.</p></li>
<li><p><strong>output_name</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></span>) – Name of the reduced dimension entry in which to store the UMAP coordinates in the output object.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A copy of <code class="docutils literal notranslate"><span class="pre">x</span></code> with a new reduced dimension entry containing the UMAP coordinates.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">get_test_rna_data_se</span><span class="p">(</span><span class="s2">&quot;pca&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Using fewer iterations for a faster-running example.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">run_umap_se</span><span class="p">(</span><span class="n">sce</span><span class="p">,</span> <span class="n">more_umap_args</span><span class="o">=</span><span class="p">{</span> <span class="s2">&quot;num_epochs&quot;</span><span class="p">:</span> <span class="mi">50</span> <span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span><span class="o">.</span><span class="n">get_reduced_dimension</span><span class="p">(</span><span class="s2">&quot;UMAP&quot;</span><span class="p">)[:</span><span class="mi">5</span><span class="p">,:]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.sanitize_size_factors">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">sanitize_size_factors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">size_factors</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">replace_zero</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">replace_negative</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">replace_infinite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">replace_nan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">in_place</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.sanitize_size_factors" title="Permalink to this definition">¶</a></dt>
<dd><p>Replace invalid size factors, i.e., zero, negative, infinite or NaNs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>size_factors</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span>) – Floating-point array containing size factors for all cells.</p></li>
<li><p><strong>replace_zero</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to replace size factors of zero with the lowest positive factor.
If <code class="docutils literal notranslate"><span class="pre">False</span></code>, zeros are retained.</p></li>
<li><p><strong>replace_negative</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to replace negative size factors with the lowest positive factor.
If <code class="docutils literal notranslate"><span class="pre">False</span></code>, negative values are retained.</p></li>
<li><p><strong>replace_infinite</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to replace infinite size factors with the largest positive factor.
If <code class="docutils literal notranslate"><span class="pre">False</span></code>, infinite values are retained.</p></li>
<li><p><strong>replace_nan</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to replace NaN size factors with unity.
If <code class="docutils literal notranslate"><span class="pre">False</span></code>, NaN values are retained.</p></li>
<li><p><strong>in_place</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to modify <code class="docutils literal notranslate"><span class="pre">size_factors</span></code> in place.
If <code class="docutils literal notranslate"><span class="pre">False</span></code>, a new array is returned.
This argument only used if <code class="docutils literal notranslate"><span class="pre">size_factors</span></code> is double-precision, otherwise a new array is always returned.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Double-precision NumPy array containing sanitized size factors.
If <code class="docutils literal notranslate"><span class="pre">in_place</span> <span class="pre">=</span> <span class="pre">True</span></code>, this is a reference to <code class="docutils literal notranslate"><span class="pre">size_factors</span></code>.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The <code class="docutils literal notranslate"><span class="pre">sanitize_size_factors</span></code> function in the <a class="reference external" href="https://libscran.github.io/scran_norm">scran_norm</a> C++ library, which provides the underlying implementation.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sf</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="mi">1000</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sf</span><span class="p">[:</span><span class="mi">4</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">inf</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">nan</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">san</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">sanitize_size_factors</span><span class="p">(</span><span class="n">sf</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">san</span><span class="p">[:</span><span class="mi">4</span><span class="p">]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.scale_by_neighbors">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">scale_by_neighbors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_neighbors=20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block_weight_policy='variable'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variable_block_weight=(0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">1000)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads=1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weights=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nn_parameters=&lt;knncolle.annoy.AnnoyParameters</span> <span class="pre">object&gt;</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_scale_by_neighbors.html#scale_by_neighbors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.scale_by_neighbors" title="Permalink to this definition">¶</a></dt>
<dd><p>Scale multiple embeddings (usually derived from different modalities across the same set of cells) so that their within-population variances are comparable.
Then, combine them into a single embedding matrix for combined downstream analysis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a></span>) – Sequence of of numeric matrices of principal components or other embeddings, one for each modality.
For each entry, rows are dimensions and columns are cells.
All entries should have the same number of columns but may have different numbers of rows.</p></li>
<li><p><strong>num_neighbors</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of neighbors to use to define the scaling factor.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.</p></li>
<li><p><strong>nn_parameters</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.classes.Parameters" title="(in knncolle v0.0.post1.dev1+g951fea368)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Parameters</span></code></a></span>) – Algorithm for the nearest-neighbor search.</p></li>
<li><p><strong>weights</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Array of length equal to <code class="docutils literal notranslate"><span class="pre">x</span></code>, specifying the weights to apply to each modality.
Each value represents a multiplier of the within-population variance of its modality, i.e., larger values increase the contribution of that modality in the combined output matrix.
The default of <code class="docutils literal notranslate"><span class="pre">None</span></code> is equivalent to an all-1 vector, i.e., all modalities are scaled to have the same within-population variance.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><p>A <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a> containing the following entries.</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">scaling</span></code>: a double-precision NumPy array containing the scaling factor to be applied to each embedding in <code class="docutils literal notranslate"><span class="pre">x</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">combined</span></code>: a double-precision NumPy matrix of scaled and concatenated embeddings.
This is formed by scaling each embedding in <code class="docutils literal notranslate"><span class="pre">x</span></code> by its corresponding entry of <code class="docutils literal notranslate"><span class="pre">scaling</span></code>, and then concatenating them together by row.
Each row corresponds to a dimension and each column corresponds to a cell.</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">References</p>
<p><a class="reference external" href="https://libscran.github.io/mumosa">https://libscran.github.io/mumosa</a>, for the basis and caveats of this approach.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rna_pcs</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="mi">25</span><span class="p">,</span> <span class="mi">200</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adt_pcs</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span> <span class="mi">200</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">other_pcs</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">200</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">scale_by_neighbors</span><span class="p">([</span><span class="n">rna_pcs</span><span class="p">,</span> <span class="n">adt_pcs</span><span class="p">,</span> <span class="n">other_pcs</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="s2">&quot;scaling&quot;</span><span class="p">])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.scale_by_neighbors_se">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">scale_by_neighbors_se</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">altexp_reddims</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">main_reddims='PCA'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_neighbors=20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nn_parameters=&lt;knncolle.annoy.AnnoyParameters</span> <span class="pre">object&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads=1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">more_scale_args={}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_name='combined'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">meta_name='combined'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_se_scale_by_neighbors.html#scale_by_neighbors_se"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.scale_by_neighbors_se" title="Permalink to this definition">¶</a></dt>
<dd><p>Scale embeddings for different modalities to equalize their intra-population variance, and then combine them into a single embedding for downstream analysis.
This calls <a class="reference internal" href="#scranpy.scale_by_neighbors" title="scranpy.scale_by_neighbors"><code class="xref py py-func docutils literal notranslate"><span class="pre">scale_by_neighbors()</span></code></a> on the main/alternative experiments of a <a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a></span>) – A <a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a> object or one of its subclasses.
Rows correspond to genomic features and columns correspond to cells.</p></li>
<li><p><strong>altexp_reddims</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>, <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a>]</span>) – <p>Dictionary specifying the reduced dimensions of the alternative experiments of <code class="docutils literal notranslate"><span class="pre">x</span></code> to be combined.
Specifically, each key is the name of an alternative experiment and the value is a list of strings/integers specifying the names/indices of the reduced dimensions of that experiment.</p>
<p>Alternatively, each value may be a single string or integer that will be converted into a list of length 1.
Each value may also be <code class="docutils literal notranslate"><span class="pre">None</span></code> in which case the corresponding alternative experiment is ignored.</p>
<p>Alternatively, a <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a> where each element is named after one or more alternative experiments,
and each value is as described above for a dictionary.</p>
</p></li>
<li><p><strong>main_reddims</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – <p>List of strings/integers specifying the names/indices of the reduced dimensions from the main experiment of <code class="docutils literal notranslate"><span class="pre">x</span></code>.</p>
<p>Alternatively, a single string or integer that will be converted into a list of length 1.</p>
<p>Alternatively <code class="docutils literal notranslate"><span class="pre">None</span></code>, which will be treated as a</p>
</p></li>
<li><p><strong>num_neighbors</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of neighbors to compute the scaling, see <a class="reference internal" href="#scranpy.scale_by_neighbors" title="scranpy.scale_by_neighbors"><code class="xref py py-func docutils literal notranslate"><span class="pre">scale_by_neighbors()</span></code></a>.</p></li>
<li><p><strong>block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Block assignment for each cell, passed to <a class="reference internal" href="#scranpy.scale_by_neighbors" title="scranpy.scale_by_neighbors"><code class="xref py py-func docutils literal notranslate"><span class="pre">scale_by_neighbors()</span></code></a>.</p></li>
<li><p><strong>nn_parameters</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.classes.Parameters" title="(in knncolle v0.0.post1.dev1+g951fea368)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Parameters</span></code></a></span>) – Algorithm for the nearest neighbor search, passed to <a class="reference internal" href="#scranpy.scale_by_neighbors" title="scranpy.scale_by_neighbors"><code class="xref py py-func docutils literal notranslate"><span class="pre">scale_by_neighbors()</span></code></a>.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads, passed to <a class="reference internal" href="#scranpy.scale_by_neighbors" title="scranpy.scale_by_neighbors"><code class="xref py py-func docutils literal notranslate"><span class="pre">scale_by_neighbors()</span></code></a>.</p></li>
<li><p><strong>more_scale_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Additional arguments to pass to <a class="reference internal" href="#scranpy.scale_by_neighbors" title="scranpy.scale_by_neighbors"><code class="xref py py-func docutils literal notranslate"><span class="pre">scale_by_neighbors()</span></code></a>.</p></li>
<li><p><strong>output_name</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></span>) – Name of the reduced dimension entry in which to store the combined embeddings in the output object.</p></li>
<li><p><strong>meta_name</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – Name of the metadata entry in which to store additional metrics in the output object.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, additional metrics are not stored.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SingleCellExperiment/api/singlecellexperiment.html#singlecellexperiment.SingleCellExperiment.SingleCellExperiment" title="(in SingleCellExperiment v0.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A copy of <code class="docutils literal notranslate"><span class="pre">x</span></code> with the combined embeddings stored in its row data.
The scaling factors for all embeddings are stored in the metadata.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">get_test_adt_data_se</span><span class="p">(</span><span class="s2">&quot;pca&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">scale_by_neighbors_se</span><span class="p">(</span><span class="n">sce</span><span class="p">,</span> <span class="n">altexp_reddims</span><span class="o">=</span><span class="p">{</span> <span class="s2">&quot;ADT&quot;</span><span class="p">:</span> <span class="s2">&quot;PCA&quot;</span> <span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span><span class="o">.</span><span class="n">get_reduced_dimension_names</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span><span class="o">.</span><span class="n">get_metadata</span><span class="p">()[</span><span class="s2">&quot;combined&quot;</span><span class="p">]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.score_gene_set">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">score_gene_set</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">set</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rank</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block_weight_policy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'variable'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variable_block_weight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(0,</span> <span class="pre">1000)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extra_work</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">iterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5489</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">realized</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_score_gene_set.html#score_gene_set"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.score_gene_set" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute per-cell scores for a gene set, defined as the column sums of a rank-1 approximation to the submatrix for the feature set.
This uses the same approach implemented in the GSDecon package by Jason Hackney.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Any" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a></span>) – A matrix-like object where rows correspond to genes or genomic features and columns correspond to cells.
The matrix is expected to contain log-expression values.</p></li>
<li><p><strong>set</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a></span>) – Array of integer indices specifying the rows of <code class="docutils literal notranslate"><span class="pre">x</span></code> belonging to the gene set.
Alternatively, a sequence of boolean values of length equal to the number of rows, where truthy elements indicate that the corresponding row belongs to the gene set.</p></li>
<li><p><strong>rank</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Rank of the approximation.</p></li>
<li><p><strong>scale</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to scale all genes to have the same variance.</p></li>
<li><p><strong>block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Array of length equal to the number of columns of <code class="docutils literal notranslate"><span class="pre">x</span></code>, containing the block of origin (e.g., batch, sample) for each cell.
Alternatively <code class="docutils literal notranslate"><span class="pre">None</span></code>, if all cells are from the same block.</p></li>
<li><p><strong>block_weight_policy</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Literal" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'variable'</span></code>, <code class="docutils literal notranslate"><span class="pre">'equal'</span></code>, <code class="docutils literal notranslate"><span class="pre">'none'</span></code>]</span>) – Policy to use for weighting different blocks when computing the average for each statistic.
Only used if <code class="docutils literal notranslate"><span class="pre">block</span></code> is provided.</p></li>
<li><p><strong>variable_block_weight</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Tuple" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code></a></span>) – Parameters for variable block weighting.
This should be a tuple of length 2 where the first and second values are used as the lower and upper bounds, respectively, for the variable weight calculation.
Only used if <code class="docutils literal notranslate"><span class="pre">block</span></code> is provided and <code class="docutils literal notranslate"><span class="pre">block_weight_policy</span> <span class="pre">=</span> <span class="pre">&quot;variable&quot;</span></code>.</p></li>
<li><p><strong>extra_work</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of extra dimensions for the IRLBA workspace.</p></li>
<li><p><strong>iterations</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Maximum number of restart iterations for IRLBA.</p></li>
<li><p><strong>seed</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Seed for the initial random vector in IRLBA.</p></li>
<li><p><strong>realized</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to realize <code class="docutils literal notranslate"><span class="pre">x</span></code> into an optimal memory layout for IRLBA.
This speeds up computation at the cost of increased memory usage.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><p>A <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a> containing the following entries.</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">scores</span></code>: a double-precision NumPy array containing the gene set score for each cell.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">weights</span></code>: a double-precision NumPy array containing containing the weight of each gene in <code class="docutils literal notranslate"><span class="pre">set</span></code>.
A larger weight indicates that the corresponding gene in <code class="docutils literal notranslate"><span class="pre">set</span></code> has more influence on <code class="docutils literal notranslate"><span class="pre">scores</span></code>.</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">References</p>
<p><a class="reference external" href="https://libscran.github.io/gsdecon">https://libscran.github.io/gsdecon</a>, which describes the approach in more detail.
In particular, the documentation for the <code class="docutils literal notranslate"><span class="pre">compute_blocked</span></code> function explains the blocking strategy.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">normed</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="mi">200</span><span class="p">,</span> <span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">score_gene_set</span><span class="p">(</span><span class="n">normed</span><span class="p">,</span> <span class="nb">set</span><span class="o">=</span><span class="nb">range</span><span class="p">(</span><span class="mi">20</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.score_gene_set_se">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">score_gene_set_se</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">set</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">more_score_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assay_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'logcounts'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_se_score_gene_set.html#score_gene_set_se"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.score_gene_set_se" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute a gene set activity score for each cell based on the expression values of the genes in the set.
This calls <a class="reference internal" href="#scranpy.score_gene_set" title="scranpy.score_gene_set"><code class="xref py py-func docutils literal notranslate"><span class="pre">score_gene_set()</span></code></a> on an assay of a <code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a></span>) – A <a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a> object or one of its subclasses.
Rows correspond to genes and columns correspond to cells.</p></li>
<li><p><strong>set</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a></span>) – Names or indices of genes in the set, see <a class="reference internal" href="#scranpy.score_gene_set" title="scranpy.score_gene_set"><code class="xref py py-func docutils literal notranslate"><span class="pre">score_gene_set()</span></code></a> for details.</p></li>
<li><p><strong>block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Block assignment for each cell, passed to <a class="reference internal" href="#scranpy.score_gene_set" title="scranpy.score_gene_set"><code class="xref py py-func docutils literal notranslate"><span class="pre">score_gene_set()</span></code></a>.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads, passed to <a class="reference internal" href="#scranpy.score_gene_set" title="scranpy.score_gene_set"><code class="xref py py-func docutils literal notranslate"><span class="pre">score_gene_set()</span></code></a>.</p></li>
<li><p><strong>more_score_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Additional arguments to pass to <a class="reference internal" href="#scranpy.score_gene_set" title="scranpy.score_gene_set"><code class="xref py py-func docutils literal notranslate"><span class="pre">score_gene_set()</span></code></a>.</p></li>
<li><p><strong>assay_type</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>) – Name or index of the assay of <code class="docutils literal notranslate"><span class="pre">x</span></code> from which to compute the gene set scores.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a> containing per-cell scores and per-gene weights,
see <a class="reference internal" href="#scranpy.score_gene_set" title="scranpy.score_gene_set"><code class="xref py py-func docutils literal notranslate"><span class="pre">score_gene_set()</span></code></a> for details.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">get_test_rna_data_se</span><span class="p">(</span><span class="s2">&quot;norm&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">custom_set</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">7</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">custom_scores</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">score_gene_set_se</span><span class="p">(</span><span class="n">sce</span><span class="p">,</span> <span class="n">custom_set</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">custom_scores</span><span class="p">[</span><span class="s2">&quot;scores&quot;</span><span class="p">]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.score_markers">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">score_markers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groups</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block_average_policy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'mean'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block_weight_policy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'variable'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variable_block_weight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(0,</span> <span class="pre">1000)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block_quantile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compute_group_mean</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compute_group_detected</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compute_cohens_d</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compute_auc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compute_delta_mean</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compute_delta_detected</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compute_summary_min</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compute_summary_mean</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compute_summary_median</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compute_summary_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compute_summary_quantiles</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compute_summary_min_rank</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_rank_limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">500</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">all_pairwise</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_score_markers.html#score_markers"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.score_markers" title="Permalink to this definition">¶</a></dt>
<dd><p>Score marker genes for each group using a variety of effect sizes from pairwise comparisons between groups.
This includes Cohen’s d, the area under the curve (AUC), the difference in the means (delta-mean) and the difference in the proportion of detected cells (delta-detected).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Any" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a></span>) – A matrix-like object where rows correspond to genes or genomic features and columns correspond to cells.
It is typically expected to contain log-expression values, e.g., from <a class="reference internal" href="#scranpy.normalize_counts" title="scranpy.normalize_counts"><code class="xref py py-func docutils literal notranslate"><span class="pre">normalize_counts()</span></code></a>.</p></li>
<li><p><strong>groups</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a></span>) – Group assignment for each cell in <code class="docutils literal notranslate"><span class="pre">x</span></code>.
This should have length equal to the number of columns in <code class="docutils literal notranslate"><span class="pre">x</span></code>.</p></li>
<li><p><strong>block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Array of length equal to the number of columns of <code class="docutils literal notranslate"><span class="pre">x</span></code>, containing the block of origin (e.g., batch, sample) for each cell.
Alternatively <code class="docutils literal notranslate"><span class="pre">None</span></code>, if all cells are from the same block.</p></li>
<li><p><strong>block_average_policy</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Literal" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'mean'</span></code>, <code class="docutils literal notranslate"><span class="pre">'quantile'</span></code>]</span>) – Policy to use for average statistics across blocks.
This can either be a (weighted) <code class="docutils literal notranslate"><span class="pre">mean</span></code> or a <code class="docutils literal notranslate"><span class="pre">quantile</span></code>.
Only used if <code class="docutils literal notranslate"><span class="pre">block</span></code> is supplied.</p></li>
<li><p><strong>block_weight_policy</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Literal" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'variable'</span></code>, <code class="docutils literal notranslate"><span class="pre">'equal'</span></code>, <code class="docutils literal notranslate"><span class="pre">'none'</span></code>]</span>) – Policy to use for weighting different blocks when computing the average for each statistic.
Only used if <code class="docutils literal notranslate"><span class="pre">block</span></code> is provided.</p></li>
<li><p><strong>variable_block_weight</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Tuple" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code></a></span>) – Parameters for variable block weighting.
This should be a tuple of length 2 where the first and second values are used as the lower and upper bounds, respectively, for the variable weight calculation.
Only used if <code class="docutils literal notranslate"><span class="pre">block</span></code> is provided and <code class="docutils literal notranslate"><span class="pre">block_weight_policy</span> <span class="pre">=</span> <span class="pre">&quot;variable&quot;</span></code>.</p></li>
<li><p><strong>block_quantile</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Probability of the quantile of statistics across blocks.
Defaults to 0.5, i.e., the median of per-block statistics.
Only used if <code class="docutils literal notranslate"><span class="pre">block</span></code> is provided and <code class="docutils literal notranslate"><span class="pre">block_average_policy</span> <span class="pre">=&quot;quantile&quot;</span></code>.</p></li>
<li><p><strong>threshold</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Non-negative value specifying the minimum threshold on the differences in means (i.e., the log-fold change, if <code class="docutils literal notranslate"><span class="pre">x</span></code> contains log-expression values).
This is incorporated into the calculation for Cohen’s d and the AUC.</p></li>
<li><p><strong>compute_group_mean</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to compute the group-wise mean expression for each gene.</p></li>
<li><p><strong>compute_group_detected</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to compute the group-wise proportion of detected cells for each gene.</p></li>
<li><p><strong>compute_cohens_d</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to compute Cohen’s d, i.e., the ratio of the difference in means to the standard deviation.</p></li>
<li><p><strong>compute_auc</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to compute the AUC.
Setting this to <code class="docutils literal notranslate"><span class="pre">False</span></code> can improve speed and memory efficiency.</p></li>
<li><p><strong>compute_delta_mean</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to compute the delta-means, i.e., the log-fold change when <code class="docutils literal notranslate"><span class="pre">x</span></code> contains log-expression values.</p></li>
<li><p><strong>compute_delta_detected</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to compute the delta-detected, i.e., differences in the proportion of cells with detected expression.</p></li>
<li><p><strong>compute_summary_min</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to compute the minimum as a summary statistic for each effect size.
Only used if <code class="docutils literal notranslate"><span class="pre">all_pairwise</span> <span class="pre">=</span> <span class="pre">False</span></code>.</p></li>
<li><p><strong>compute_summary_mean</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to compute the mean as a summary statistic for each effect size.
Only used if <code class="docutils literal notranslate"><span class="pre">all.pairwise</span> <span class="pre">=</span> <span class="pre">False</span></code>.</p></li>
<li><p><strong>compute_summary_median</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to compute the median as a summary statistic for each effect size.
Only used if <code class="docutils literal notranslate"><span class="pre">all_pairwise</span> <span class="pre">=</span> <span class="pre">False</span></code>.</p></li>
<li><p><strong>compute_summary_max</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to compute the maximum as a summary statistic for each effect size.
Only used if <code class="docutils literal notranslate"><span class="pre">all_pairwise</span> <span class="pre">=</span> <span class="pre">False</span></code>.</p></li>
<li><p><strong>compute_summary_quantiles</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Probabilities of quantiles to compute as summary statistics for each effect size.
This should be in [0, 1] and sorted in order of increasing size.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, no quantiles are computed.
Only used if <code class="docutils literal notranslate"><span class="pre">all_pairwise</span> <span class="pre">=</span> <span class="pre">False</span></code>.</p></li>
<li><p><strong>compute_summary_min_rank</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to compute the mininum rank as a summary statistic for each effect size.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, no quantiles are computed.
Only used if <code class="docutils literal notranslate"><span class="pre">all_pairwise</span> <span class="pre">=</span> <span class="pre">False</span></code>.</p></li>
<li><p><strong>min_rank_limit</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Maximum value of the min-rank to report.
Lower values improve memory efficiency at the cost of discarding information about lower-ranked genes.
Only used if <code class="docutils literal notranslate"><span class="pre">all_pairwise</span> <span class="pre">=</span> <span class="pre">False</span></code> and <code class="docutils literal notranslate"><span class="pre">compute_summary_min_rank</span> <span class="pre">=</span> <span class="pre">True</span></code>.</p></li>
<li><p><strong>all_pairwise</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to report the full effects for every pairwise comparison between groups.
Alternatively, an integer scalar indicating the number of top markers to report from each pairwise comparison between groups.
If <code class="docutils literal notranslate"><span class="pre">False</span></code>, only summaries are reported.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><p>A <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a> containing various marker statistics for each group.
This has the following entries:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">nrow</span></code>: integer specifying the number of genes in the dataset.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">group_ids</span></code>: list containing the identities of the groups.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">means</span></code>: double-precision NumPy matrix containing the mean expression for each gene in each group.
Each row is a gene and each column is a group, ordered as in <code class="docutils literal notranslate"><span class="pre">group_ids</span></code>.
Omitted if <code class="docutils literal notranslate"><span class="pre">compute_group_means</span> <span class="pre">=</span> <span class="pre">False</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">means</span></code>: double-precision NumPy matrix containing the proportion of cells with detected expression for each gene in each group.
Each row is a gene and each column is a group, ordered as in <code class="docutils literal notranslate"><span class="pre">group_ids</span></code>.
Omitted if <code class="docutils literal notranslate"><span class="pre">compute_group_means</span> <span class="pre">=</span> <span class="pre">False</span></code>.</p></li>
</ul>
<p>If <code class="docutils literal notranslate"><span class="pre">all_pairwise</span> <span class="pre">=</span> <span class="pre">False</span></code>, the <code class="docutils literal notranslate"><span class="pre">NamedList</span></code> contains the following additional entries:</p>
<ul class="simple">
<li><dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">cohens_d</span></code>: a <code class="docutils literal notranslate"><span class="pre">NamedList</span></code> with the same structure as returned by <code class="xref py py-func docutils literal notranslate"><span class="pre">summarized_effects()</span></code>.</dt><dd><p>Briefly, each entry corresponds to a group in <code class="docutils literal notranslate"><span class="pre">group_ids</span></code> and is a <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a> with one row per gene.
Each column contains a summary statistic of the Cohen’s d from pairwise comparisons to all other groups, e.g., min, mean, median, max, min-rank, and any requested quantiles.
Columns are omitted if the relevant <code class="docutils literal notranslate"><span class="pre">compute_summary_*</span></code> option is set to <code class="docutils literal notranslate"><span class="pre">False</span></code>.</p>
</dd>
</dl>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">auc</span></code>: Same as <code class="docutils literal notranslate"><span class="pre">cohens_d</span></code> but for the AUCs.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">delta_mean</span></code>: Same as <code class="docutils literal notranslate"><span class="pre">cohens_d</span></code> but for the delta-mean.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">delta_detected</span></code>: Same as <code class="docutils literal notranslate"><span class="pre">cohens_d</span></code> but for the delta-detected.</p></li>
</ul>
<p>If <code class="docutils literal notranslate"><span class="pre">all_pairwise</span> <span class="pre">=</span> <span class="pre">True</span></code>, the <code class="docutils literal notranslate"><span class="pre">NamedList</span></code> contains the following addditional entries:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">cohens_d</span></code>: a 3-dimensional double-precision NumPy array containing the Cohen’s d from each pairwise comparison between groups.
The extents of the first two dimensions are equal to the number of groups, while the extent of the final dimension is equal to the number of genes.
Specifically, the entry <code class="docutils literal notranslate"><span class="pre">[i,</span> <span class="pre">j,</span> <span class="pre">k]</span></code> represents Cohen’s d from the comparison of group <code class="docutils literal notranslate"><span class="pre">j</span></code> over group <code class="docutils literal notranslate"><span class="pre">i</span></code> for gene <code class="docutils literal notranslate"><span class="pre">k</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">auc</span></code>: Same as <code class="docutils literal notranslate"><span class="pre">cohens_d</span></code> but for the AUCs.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">delta_mean</span></code>: Same as <code class="docutils literal notranslate"><span class="pre">cohens_d</span></code> but for the delta-mean.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">delta_detected</span></code>: Same as <code class="docutils literal notranslate"><span class="pre">cohens_d</span></code> but for the delta-detected.</p></li>
</ul>
<p>If <code class="docutils literal notranslate"><span class="pre">all_pairwise</span></code> is an integer, the <code class="docutils literal notranslate"><span class="pre">NamedList</span></code> contains the following additional entries:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">cohens_d</span></code>: a <code class="docutils literal notranslate"><span class="pre">NamedList</span></code> list of named lists of <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a> objects.
The <code class="docutils literal notranslate"><span class="pre">BiocFrame</span></code> at <code class="docutils literal notranslate"><span class="pre">cohens_d[m][n]</span></code> contains the top markers for the comparison of group <code class="docutils literal notranslate"><span class="pre">m</span></code> over group <code class="docutils literal notranslate"><span class="pre">n</span></code>.
Each <code class="docutils literal notranslate"><span class="pre">BiocFrame</span></code> has the <code class="docutils literal notranslate"><span class="pre">index</span></code> and <code class="docutils literal notranslate"><span class="pre">effect</span></code> columns, containing the row indices and effect sizes of the top genes, respectively.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">auc</span></code>: Same as <code class="docutils literal notranslate"><span class="pre">cohens_d</span></code> but for the AUCs.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">delta_mean</span></code>: Same as <code class="docutils literal notranslate"><span class="pre">cohens_d</span></code> but for the delta-mean.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">delta_detected</span></code>: Same as <code class="docutils literal notranslate"><span class="pre">cohens_d</span></code> but for the delta-detected.</p></li>
</ul>
<p>Entries will be omitted if the relevant <code class="docutils literal notranslate"><span class="pre">compute_*</span></code> option is set to <code class="docutils literal notranslate"><span class="pre">False</span></code>.
For example, if <code class="docutils literal notranslate"><span class="pre">compute_cohens_d</span> <span class="pre">=</span> <span class="pre">False</span></code>, the output will not contain any <code class="docutils literal notranslate"><span class="pre">cohens_d</span></code> entry.</p>
</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The <code class="docutils literal notranslate"><span class="pre">score_markers_summary</span></code> and <code class="docutils literal notranslate"><span class="pre">score_markers_pairwise</span></code> functions in the <a class="reference external" href="https://libscran.github.io/scran_markers">scran_markers</a> C++ library,
which describes the rationale behind the choice of effect sizes and summary statistics.
Also see their blocked equivalents <code class="docutils literal notranslate"><span class="pre">score_markers_summary_blocked</span></code> and <code class="docutils literal notranslate"><span class="pre">score_markers_pairwise_blocked</span></code> when <code class="docutils literal notranslate"><span class="pre">block</span></code> is provided.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">normed</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="mi">200</span><span class="p">,</span> <span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">group</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="s2">&quot;D&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">25</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">score_markers</span><span class="p">(</span><span class="n">normed</span><span class="p">,</span> <span class="n">group</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="s2">&quot;cohens_d&quot;</span><span class="p">][</span><span class="s2">&quot;A&quot;</span><span class="p">])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.score_markers_se">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">score_markers_se</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groups</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">more_marker_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assay_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'logcounts'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extra_columns</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order_by</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.score_markers_se" title="Permalink to this definition">¶</a></dt>
<dd><p>Identify candidate marker genes based on effect sizes from pairwise comparisons between groups of cells.
This calls <a class="reference internal" href="#scranpy.score_markers" title="scranpy.score_markers"><code class="xref py py-func docutils literal notranslate"><span class="pre">score_markers()</span></code></a> on an assay of a <a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>,
and then uses <a class="reference internal" href="#scranpy.format_score_markers_result" title="scranpy.format_score_markers_result"><code class="xref py py-func docutils literal notranslate"><span class="pre">format_score_markers_result()</span></code></a> to reformat the results.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a></span>) – A <a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a> object or one of its subclasses.
Rows correspond to genes and columns correspond to cells.</p></li>
<li><p><strong>groups</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a></span>) – Group assignment for each cell, passed to <a class="reference internal" href="#scranpy.score_markers" title="scranpy.score_markers"><code class="xref py py-func docutils literal notranslate"><span class="pre">score_markers()</span></code></a>.</p></li>
<li><p><strong>block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Block assignment for each cell, passed to <a class="reference internal" href="#scranpy.score_markers" title="scranpy.score_markers"><code class="xref py py-func docutils literal notranslate"><span class="pre">score_markers()</span></code></a>.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads for marker scoring, passed to <a class="reference internal" href="#scranpy.score_markers" title="scranpy.score_markers"><code class="xref py py-func docutils literal notranslate"><span class="pre">score_markers()</span></code></a>.</p></li>
<li><p><strong>more_marker_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Additional arguments to pass to <a class="reference internal" href="#scranpy.score_markers" title="scranpy.score_markers"><code class="xref py py-func docutils literal notranslate"><span class="pre">score_markers()</span></code></a>.</p></li>
<li><p><strong>assay_type</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>) – Name or index of the assay of <code class="docutils literal notranslate"><span class="pre">x</span></code> to use for marker detection, usually containing log-normalized expression values.</p></li>
<li><p><strong>extra_columns</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/constants.html#None" title="(in Python v3.14)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]</span>) – A <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a> with the same number of rows as <code class="docutils literal notranslate"><span class="pre">x</span></code>, containing extra columns to add each DataFrame.
Alternatively, a list of strings specifying the columns of the row data of <code class="docutils literal notranslate"><span class="pre">x</span></code> to be added.
A single string is treated as a list of length 1.</p></li>
<li><p><strong>order_by</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/constants.html#None" title="(in Python v3.14)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]</span>) – Name of the column to use for ordering the rows of each output <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a>.
Alternatively <code class="docutils literal notranslate"><span class="pre">True</span></code>, in which case a column is automatically chosen from the effect size summaries.
If <code class="docutils literal notranslate"><span class="pre">None</span></code> or <code class="docutils literal notranslate"><span class="pre">False</span></code>, no ordering is performed.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><p>A <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a> of <code class="xref py py-class docutils literal notranslate">
<span class="pre">Each</span> <span class="pre">``BiocFrame`</span></code> corresponds to a unique group in <code class="docutils literal notranslate"><span class="pre">groups</span></code>.
Each row contains statistics for a gene in <code class="docutils literal notranslate"><span class="pre">x</span></code>, with the following columns:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">mean</span></code>, the mean expression in the current group.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">detected</span></code>, the proportion of cells with detected expression in the current group.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&lt;effect&gt;_&lt;summary&gt;</span></code>, a summary statistic for an effect size.
For example, <code class="docutils literal notranslate"><span class="pre">cohens_d_mean</span></code> contains the mean Cohen’s d across comparisons involving the current group.</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sce</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">get_test_rna_data_se</span><span class="p">(</span><span class="s2">&quot;cluster&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">markers</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">score_markers_se</span><span class="p">(</span><span class="n">sce</span><span class="p">,</span> <span class="n">sce</span><span class="o">.</span><span class="n">get_column_data</span><span class="p">()[</span><span class="s2">&quot;clusters&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">scranpy</span><span class="o">.</span><span class="n">preview_markers</span><span class="p">(</span><span class="n">markers</span><span class="p">[</span><span class="s2">&quot;0&quot;</span><span class="p">]))</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.subsample_by_neighbors">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">subsample_by_neighbors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_neighbors=20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_remaining=10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nn_parameters=&lt;knncolle.annoy.AnnoyParameters</span> <span class="pre">object&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads=1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.subsample_by_neighbors" title="Permalink to this definition">¶</a></dt>
<dd><p>Subsample a dataset by selecting cells to represent all of their nearest neighbors.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a>, <a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.find_knn.FindKnnResults" title="(in knncolle v0.0.post1.dev1+g951fea368)"><code class="xref py py-class docutils literal notranslate"><span class="pre">FindKnnResults</span></code></a>, <a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.classes.Index" title="(in knncolle v0.0.post1.dev1+g951fea368)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Index</span></code></a>]</span>) – <p>Numeric matrix where rows are dimensions and columns are cells, typically containing a low-dimensional representation from, e.g., <a class="reference internal" href="#scranpy.run_pca" title="scranpy.run_pca"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_pca()</span></code></a>.</p>
<p>Alternatively, a <a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.classes.Index" title="(in knncolle v0.0.post1.dev1+g951fea368)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Index</span></code></a> object containing a pre-built search index for a dataset.</p>
<p>Alternatively, a <a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.find_knn.FindKnnResults" title="(in knncolle v0.0.post1.dev1+g951fea368)"><code class="xref py py-class docutils literal notranslate"><span class="pre">FindKnnResults</span></code></a> object containing pre-computed search results for a dataset.</p>
</p></li>
<li><p><strong>num_neighbors</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of neighbors to use.
Larger values result in greater downsampling.
Ignored if <code class="docutils literal notranslate"><span class="pre">x</span></code> is a <a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.find_knn.FindKnnResults" title="(in knncolle v0.0.post1.dev1+g951fea368)"><code class="xref py py-class docutils literal notranslate"><span class="pre">FindKnnResults</span></code></a> object.</p></li>
<li><p><strong>nn_parameters</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://knncolle.github.io/knncolle-py/api/knncolle.html#knncolle.classes.Parameters" title="(in knncolle v0.0.post1.dev1+g951fea368)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Parameters</span></code></a></span>) – Neighbor search algorithm to use.
Only used if <code class="docutils literal notranslate"><span class="pre">x</span></code> does not contain existing neighbor search results.</p></li>
<li><p><strong>min_remaining</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Minimum number of remaining (i.e., unselected) neighbors that a cell must have in order to be considered for selection.
This should be less than or equal to <code class="docutils literal notranslate"><span class="pre">num_neighbors</span></code>.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use for the nearest-neighbor search.
Only used if <code class="docutils literal notranslate"><span class="pre">x</span></code> does not contain existing neighbor search results.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Integer NumPy array containing the indices of the cells selected to be in the subsample.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p><a class="reference external" href="https://libscran.github.io/nenesub">https://libscran.github.io/nenesub</a>, for the rationale behind this approach.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pcs</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="mi">500</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">keep</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">subsample_by_neighbors</span><span class="p">(</span><span class="n">pcs</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">keep</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.suggest_adt_qc_thresholds">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">suggest_adt_qc_thresholds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">metrics</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_detected_drop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_mads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3.0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_adt_quality_control.html#suggest_adt_qc_thresholds"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.suggest_adt_qc_thresholds" title="Permalink to this definition">¶</a></dt>
<dd><p>Suggest filter thresholds for the ADT-derived QC metrics from <a class="reference internal" href="#scranpy.compute_adt_qc_metrics" title="scranpy.compute_adt_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_adt_qc_metrics()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>metrics</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a></span>) – ADT-derived QC metrics generated by <a class="reference internal" href="#scranpy.compute_adt_qc_metrics" title="scranpy.compute_adt_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_adt_qc_metrics()</span></code></a>.
This should contain the <code class="docutils literal notranslate"><span class="pre">sum</span></code>, <code class="docutils literal notranslate"><span class="pre">detected</span></code> and <code class="docutils literal notranslate"><span class="pre">subset_sum</span></code> columns.</p></li>
<li><p><strong>block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Blocking factor specifying the block of origin (e.g., batch, sample) for each cell in <code class="docutils literal notranslate"><span class="pre">metrics</span></code>.
If supplied, a separate threshold is computed from the cells in each block.
Alternatively <code class="docutils literal notranslate"><span class="pre">None</span></code>, if all cells are from the same block.</p></li>
<li><p><strong>min_detected_drop</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Minimum proportional drop in the number of detected ADTs to consider a cell to be of low quality.
This avoids overly aggressive filtering when the MAD is zero.</p></li>
<li><p><strong>num_mads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Number of MADs from the median to define the threshold for outliers in each QC metric.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><p>If <code class="docutils literal notranslate"><span class="pre">block</span> <span class="pre">=</span> <span class="pre">None</span></code>, a <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a> is returned, containing the following entries.</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">detected</span></code>: a number specifying the lower threshold on the number of detected ADTs.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">subsets</span></code>: a <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.FloatList.FloatList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">FloatList</span></code></a> of length equal to the number of control subsets (and named accordingly).
Each entry represents the upper bound on the sum of counts in the corresponding control subset.</p></li>
</ul>
<p>If <code class="docutils literal notranslate"><span class="pre">block</span></code> is provided, the <code class="docutils literal notranslate"><span class="pre">NamedList</span></code> instead contains:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">detected</span></code>, a <code class="docutils literal notranslate"><span class="pre">FloatList</span></code> of length equal to the number of blocks (and named accordingly).
Each entry represents the lower threshold on the number of detected ADTs in the corresponding block.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">subset_sum</span></code>, a <code class="docutils literal notranslate"><span class="pre">NamedList</span></code> of length equal to the number of control subsets.
Each entry is another <code class="docutils literal notranslate"><span class="pre">FloatList</span></code> that contains the upper threshold on the sum of counts for that subset in each block.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">block_ids</span></code>, a list containing the unique levels of the blocking factor.
This is in the same order as the blocks in <code class="docutils literal notranslate"><span class="pre">detected</span></code> and <code class="docutils literal notranslate"><span class="pre">subset_sum</span></code>.</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The <code class="docutils literal notranslate"><span class="pre">compute_adt_qc_filters</span></code> and <code class="docutils literal notranslate"><span class="pre">compute_adt_qc_filters_blocked</span></code> functions in the <a class="reference external" href="https://github.com/libscran/scran_qc">scran_qc</a> C++ library,
which describe the rationale behind the suggested filters.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mat</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="n">lam</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">1000</span><span class="p">),</span> <span class="p">(</span><span class="mi">50</span><span class="p">,</span> <span class="mi">20</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">compute_adt_qc_metrics</span><span class="p">(</span><span class="n">mat</span><span class="p">,</span> <span class="p">{</span> <span class="s2">&quot;IgG&quot;</span><span class="p">:</span> <span class="p">[</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">40</span> <span class="p">]</span> <span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">suggest_adt_qc_thresholds</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">filt</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.suggest_crispr_qc_thresholds">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">suggest_crispr_qc_thresholds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">metrics</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_mads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3.0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_crispr_quality_control.html#suggest_crispr_qc_thresholds"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.suggest_crispr_qc_thresholds" title="Permalink to this definition">¶</a></dt>
<dd><p>Suggest filter thresholds for the CRISPR-derived QC metrics, typically generated from <a class="reference internal" href="#scranpy.compute_crispr_qc_metrics" title="scranpy.compute_crispr_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_crispr_qc_metrics()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>metrics</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a></span>) – CRISPR-derived QC metrics from <a class="reference internal" href="#scranpy.compute_crispr_qc_metrics" title="scranpy.compute_crispr_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_crispr_qc_metrics()</span></code></a>.
This should contain the <code class="docutils literal notranslate"><span class="pre">sum</span></code>, <code class="docutils literal notranslate"><span class="pre">detected</span></code>, <code class="docutils literal notranslate"><span class="pre">max_value</span></code> and <code class="docutils literal notranslate"><span class="pre">max_index</span></code> columns.</p></li>
<li><p><strong>block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Blocking factor specifying the block of origin (e.g., batch, sample) for each cell in <code class="docutils literal notranslate"><span class="pre">metrics</span></code>.
If supplied, a separate threshold is computed from the cells in each block.
Alternatively <code class="docutils literal notranslate"><span class="pre">None</span></code>, if all cells are from the same block.</p></li>
<li><p><strong>num_mads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Number of MADs from the median to define the threshold for outliers in each QC metric.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><p>If <code class="docutils literal notranslate"><span class="pre">block</span> <span class="pre">=</span> <span class="pre">None</span></code>, a <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a> is returned, containing the following entries.</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">max_value</span></code>, a number specifying the lower threshold on the maximum count in each cell.</p></li>
</ul>
<p>If <code class="docutils literal notranslate"><span class="pre">block</span></code> is provided, the <code class="docutils literal notranslate"><span class="pre">NamedList</span></code> instead contains:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">max_value</span></code>, a FloatList of length equal to the number of blocks (and named accordingly).
Each entry represents the lower threshold on the maximum count in the corresponding block.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">block_ids</span></code>, a list containing the unique levels of the blocking factor.
This is in the same order as the blocks in <code class="docutils literal notranslate"><span class="pre">detected</span></code> and <code class="docutils literal notranslate"><span class="pre">subset_sum</span></code>.</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The <code class="docutils literal notranslate"><span class="pre">compute_crispr_qc_filters</span></code> and <code class="docutils literal notranslate"><span class="pre">compute_crispr_qc_filters_blocked</span></code> functions in the <a class="reference external" href="https://github.com/libscran/scran_qc">scran_qc</a> C++ library,
which describes the rationale behind the suggested filters.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mat</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="n">lam</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">1000</span><span class="p">),</span> <span class="p">(</span><span class="mi">50</span><span class="p">,</span> <span class="mi">20</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">compute_crispr_qc_metrics</span><span class="p">(</span><span class="n">mat</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">suggest_crispr_qc_thresholds</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">filt</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.suggest_rna_qc_thresholds">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">suggest_rna_qc_thresholds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">metrics</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_mads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3.0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_rna_quality_control.html#suggest_rna_qc_thresholds"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.suggest_rna_qc_thresholds" title="Permalink to this definition">¶</a></dt>
<dd><p>Suggest filter thresholds for the RNA-derived QC metrics, typically generated from <a class="reference internal" href="#scranpy.compute_rna_qc_metrics" title="scranpy.compute_rna_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_rna_qc_metrics()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>metrics</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a></span>) – RNA-derived QC metrics from <a class="reference internal" href="#scranpy.compute_rna_qc_metrics" title="scranpy.compute_rna_qc_metrics"><code class="xref py py-func docutils literal notranslate"><span class="pre">compute_rna_qc_metrics()</span></code></a>.
This should contain the <code class="docutils literal notranslate"><span class="pre">sum</span></code>, <code class="docutils literal notranslate"><span class="pre">detected</span></code> and <code class="docutils literal notranslate"><span class="pre">subset_proportion</span></code> columns.</p></li>
<li><p><strong>block</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Blocking factor specifying the block of origin (e.g., batch, sample) for each cell in <code class="docutils literal notranslate"><span class="pre">metrics</span></code>.
If supplied, a separate threshold is computed from the cells in each block.
Alternatively <code class="docutils literal notranslate"><span class="pre">None</span></code>, if all cells are from the same block.</p></li>
<li><p><strong>num_mads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Number of MADs from the median to define the threshold for outliers in each QC metric.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><p>If <code class="docutils literal notranslate"><span class="pre">block</span> <span class="pre">=</span> <span class="pre">None</span></code>, a <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a> is returned, containing the following entries.</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">sum</span></code>, a number specifying the lower threshold on the sum of counts in each cell.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">detected</span></code>, a number specifying the lower threshold on the number of detected genes.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">subset_proportion</span></code>, a <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.FloatList.FloatList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">FloatList</span></code></a> of length equal to the number of control subsets (and named accordingly).
Each entry represents the upper bound on the proportion of counts in the corresponding control subset.</p></li>
</ul>
<p>If <code class="docutils literal notranslate"><span class="pre">block</span></code> is provided, the <code class="docutils literal notranslate"><span class="pre">NamedList</span></code> instead contains:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">sum</span></code>, a FloatList of length equal to the number of blocks (and named accordingly).
Each entry represents the lower threshold on the sum of counts in the corresponding block.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">detected</span></code>, a FloatList of length equal to the number of blocks (and named accordingly).
Each entry represents the lower threshold on the number of detected genes in the corresponding block.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">subset_proportion</span></code>, a <code class="docutils literal notranslate"><span class="pre">NamedList</span></code> of length equal to the number of control subsets.
Each entry is another FloatList that contains the upper threshold on the proportion of counts for that subset in each block.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">block_ids</span></code>, a list containing the unique levels of the blocking factor.
This is in the same order as the blocks in <code class="docutils literal notranslate"><span class="pre">detected</span></code> and <code class="docutils literal notranslate"><span class="pre">subset_sum</span></code>.</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The <code class="docutils literal notranslate"><span class="pre">compute_rna_qc_filters</span></code> and <code class="docutils literal notranslate"><span class="pre">compute_rna_qc_filters_blocked</span></code> functions in the <a class="reference external" href="https://github.com/libscran/scran_qc">scran_qc</a> C++ library,
which describes the rationale behind the suggested filters.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mat</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="n">lam</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">1000</span><span class="p">),</span> <span class="p">(</span><span class="mi">50</span><span class="p">,</span> <span class="mi">20</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">compute_rna_qc_metrics</span><span class="p">(</span><span class="n">mat</span><span class="p">,</span> <span class="p">{</span> <span class="s2">&quot;mito&quot;</span><span class="p">:</span> <span class="p">[</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">40</span> <span class="p">]</span> <span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">suggest_rna_qc_thresholds</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">filt</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.summarize_effects">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">summarize_effects</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">effects</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compute_min</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compute_mean</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compute_median</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compute_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compute_quantiles</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compute_min_rank</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_summarize_effects.html#summarize_effects"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.summarize_effects" title="Permalink to this definition">¶</a></dt>
<dd><p>For each group, summarize the effect sizes for all pairwise comparisons
to other groups. This yields a set of summary statistics that can be used
to rank marker genes for each group.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>effects</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span>) – A 3-dimensional numeric containing the effect sizes from each pairwise comparison between groups.
The extents of the first two dimensions should be equal to the number of groups, while the extent of the final dimension is equal to the number of genes.
The entry <code class="docutils literal notranslate"><span class="pre">[i,</span> <span class="pre">j,</span> <span class="pre">k]</span></code> should represent the effect size from the comparison of group <code class="docutils literal notranslate"><span class="pre">j</span></code> against group <code class="docutils literal notranslate"><span class="pre">i</span></code> for gene <code class="docutils literal notranslate"><span class="pre">k</span></code>.
See also the output of <a class="reference internal" href="#scranpy.score_markers" title="scranpy.score_markers"><code class="xref py py-func docutils literal notranslate"><span class="pre">score_markers()</span></code></a> with <code class="docutils literal notranslate"><span class="pre">all_pairwise</span> <span class="pre">=</span> <span class="pre">True</span></code>.</p></li>
<li><p><strong>compute_min</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to compute the minimum as a summary statistic for each effect size.</p></li>
<li><p><strong>compute_mean</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to compute the mean as a summary statistic for each effect size.</p></li>
<li><p><strong>compute_median</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to compute the median as a summary statistic for each effect size.</p></li>
<li><p><strong>compute_max</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to compute the maximum as a summary statistic for each effect size.</p></li>
<li><p><strong>compute_quantiles</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Optional" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Probabilities of quantiles to compute as summary statistics for each effect size.
This should be in [0, 1] and sorted in order of increasing size.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, no quantiles are computed.</p></li>
<li><p><strong>compute_min_rank</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to compute the mininum rank as a summary statistic for each effect size.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, no quantiles are computed.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><p>A <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a> of length equal to the number of groups (i.e., the extents of the first two dimensions of <code class="docutils literal notranslate"><span class="pre">effects</span></code>).
Each entry is a <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a> where each row corresponds to a gene in <code class="docutils literal notranslate"><span class="pre">effects</span></code>.
Each column contain a summary statistic of the effect sizes of the comparisons involving its corresponding group.</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">min</span></code>: double-precision NumPy array containing the minimum effect size across all pairwise comparisons to other groups.
Only present if <code class="docutils literal notranslate"><span class="pre">compute_min</span> <span class="pre">=</span> <span class="pre">true</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">median</span></code>: double-precision NumPy array containing the median effect size across all pairwise comparisons to other groups.
Only present if <code class="docutils literal notranslate"><span class="pre">compute_median</span> <span class="pre">=</span> <span class="pre">True</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">mean</span></code>: double-precision NumPy array containing the mean effect size from across all pairwise comparisons to other groups.
Only present if <code class="docutils literal notranslate"><span class="pre">compute_median</span> <span class="pre">=</span> <span class="pre">True</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">quantile</span></code>: nested <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a> containing the specified quantiles of the effect sizes across all pairwise comparisons to other groups.
Only present if <code class="docutils literal notranslate"><span class="pre">compute_quantiles</span></code> is provided.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">max</span></code>: double-precision NumPy array containing the maximum effect size across all pairwise comparisons to other groups.
Only present if <code class="docutils literal notranslate"><span class="pre">compute_max</span> <span class="pre">=</span> <span class="pre">true</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">min_rank</span></code>: integer array containing the minimum rank of each gene across all pairwise comparisons to other groups.
Only present if <code class="docutils literal notranslate"><span class="pre">compute_min_rank</span> <span class="pre">=</span> <span class="pre">true</span></code>.</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>The <code class="docutils literal notranslate"><span class="pre">summarize_effects</span></code> function in the <a class="reference external" href="https://libscran.github.io/scran_markers">scran_markers</a> C++ library, for more details on the statistics.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">normed</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="mi">200</span><span class="p">,</span> <span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">group</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="s2">&quot;D&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">25</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">score_markers</span><span class="p">(</span><span class="n">normed</span><span class="p">,</span> <span class="n">group</span><span class="p">,</span> <span class="n">all_pairwise</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">summaries</span> <span class="o">=</span> <span class="n">scranpy</span><span class="o">.</span><span class="n">summarize_effects</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="s2">&quot;cohens_d&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">summaries</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.test_enrichment">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">test_enrichment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sets</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">universe</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scranpy/_test_enrichment.html#test_enrichment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scranpy.test_enrichment" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform a hypergeometric test for enrichment of interesting genes (e.g., markers) in one or more pre-defined gene sets.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a></span>) – Sequence of identifiers for the interesting genes.</p></li>
<li><p><strong>sets</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – <p>Sequence of gene sets, where each entry corresponds to a gene set and contains a sequence of identifiers for genes in that set.</p>
<p>Alternatively, a dictionary where each key is the name of a gene set and each value is a sequence of identifiers for that gene set.</p>
</p></li>
<li><p><strong>universe</strong> (<span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>, <a class="reference external" href="https://docs.python.org/3.14/library/typing.html#typing.Sequence" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – <p>Sequence of identifiers for the universe of genes in the dataset.
It is assumed that <code class="docutils literal notranslate"><span class="pre">x</span></code> is a subset of <code class="docutils literal notranslate"><span class="pre">universe</span></code>.
Identifiers in <code class="docutils literal notranslate"><span class="pre">sets</span></code> that are not in <code class="docutils literal notranslate"><span class="pre">universe</span></code> will be ignored.</p>
<p>Alternatively, an integer specifying the number of genes in the universe.</p>
</p></li>
<li><p><strong>log</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to report the log-transformed p-values.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Double-precision NumPy array of (log-transformed) p-values to test for significant enrichment of <code class="docutils literal notranslate"><span class="pre">x</span></code> in each entry of <code class="docutils literal notranslate"><span class="pre">sets</span></code>.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p><a class="reference external" href="https://libscran.github.io/phyper">https://libscran.github.io/phyper</a>, for the underlying implementation.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">LETTERS</span> <span class="o">=</span> <span class="s2">&quot;ABCDEFGHIJKLMNOPQRSTUVWXYZ&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">scranpy</span><span class="o">.</span><span class="n">test_enrichment</span><span class="p">(</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">x</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="s2">&quot;D&quot;</span><span class="p">,</span> <span class="s2">&quot;E&quot;</span><span class="p">],</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">sets</span><span class="o">=</span><span class="p">{</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="s2">&quot;set1&quot;</span><span class="p">:</span> <span class="n">LETTERS</span><span class="p">[:</span><span class="mi">10</span><span class="p">],</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="s2">&quot;set2&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;D&quot;</span><span class="p">,</span> <span class="s2">&quot;F&quot;</span><span class="p">,</span> <span class="s2">&quot;H&quot;</span><span class="p">,</span> <span class="s2">&quot;J&quot;</span><span class="p">],</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="s2">&quot;set3&quot;</span><span class="p">:</span> <span class="n">LETTERS</span><span class="p">[</span><span class="mi">10</span><span class="p">:</span><span class="mi">20</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="p">},</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">universe</span><span class="o">=</span><span class="n">LETTERS</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scranpy.tsne_perplexity_to_neighbors">
<span class="sig-prename descclassname"><span class="pre">scranpy.</span></span><span class="sig-name descname"><span class="pre">tsne_perplexity_to_neighbors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">perplexity</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scranpy.tsne_perplexity_to_neighbors" title="Permalink to this definition">¶</a></dt>
<dd><p>Determine the number of nearest neighbors required to support a given
perplexity in the t-SNE algorithm.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>perplexity</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Perplexity to use in <a class="reference internal" href="#scranpy.run_tsne" title="scranpy.run_tsne"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_tsne()</span></code></a>.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.14/library/functions.html#int" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The corresponding number of nearest neighbors.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scranpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">scranpy</span><span class="o">.</span><span class="n">tsne_perplexity_to_neighbors</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">scranpy</span><span class="o">.</span><span class="n">tsne_perplexity_to_neighbors</span><span class="p">(</span><span class="mi">30</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">scranpy</span><span class="o">.</span><span class="n">tsne_perplexity_to_neighbors</span><span class="p">(</span><span class="mi">50</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          
          <a class="prev-page" href="modules.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">scranpy</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2023, Jayaram Kancherla
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">scranpy package</a><ul>
<li><a class="reference internal" href="#module-scranpy">Module contents</a><ul>
<li><a class="reference internal" href="#scranpy.aggregate_across_cells"><code class="docutils literal notranslate"><span class="pre">aggregate_across_cells()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.aggregate_across_cells_se"><code class="docutils literal notranslate"><span class="pre">aggregate_across_cells_se()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.aggregate_across_genes"><code class="docutils literal notranslate"><span class="pre">aggregate_across_genes()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.aggregate_across_genes_se"><code class="docutils literal notranslate"><span class="pre">aggregate_across_genes_se()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.aggregate_column_data"><code class="docutils literal notranslate"><span class="pre">aggregate_column_data()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze"><code class="docutils literal notranslate"><span class="pre">analyze()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.analyze_se"><code class="docutils literal notranslate"><span class="pre">analyze_se()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.build_snn_graph"><code class="docutils literal notranslate"><span class="pre">build_snn_graph()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.center_size_factors"><code class="docutils literal notranslate"><span class="pre">center_size_factors()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.choose_highly_variable_genes"><code class="docutils literal notranslate"><span class="pre">choose_highly_variable_genes()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.choose_pseudo_count"><code class="docutils literal notranslate"><span class="pre">choose_pseudo_count()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.choose_rna_hvgs_se"><code class="docutils literal notranslate"><span class="pre">choose_rna_hvgs_se()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_graph"><code class="docutils literal notranslate"><span class="pre">cluster_graph()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_graph_se"><code class="docutils literal notranslate"><span class="pre">cluster_graph_se()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_kmeans"><code class="docutils literal notranslate"><span class="pre">cluster_kmeans()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.cluster_kmeans_se"><code class="docutils literal notranslate"><span class="pre">cluster_kmeans_se()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.combine_factors"><code class="docutils literal notranslate"><span class="pre">combine_factors()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.compute_adt_qc_metrics"><code class="docutils literal notranslate"><span class="pre">compute_adt_qc_metrics()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.compute_clrm1_factors"><code class="docutils literal notranslate"><span class="pre">compute_clrm1_factors()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.compute_crispr_qc_metrics"><code class="docutils literal notranslate"><span class="pre">compute_crispr_qc_metrics()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.compute_rna_qc_metrics"><code class="docutils literal notranslate"><span class="pre">compute_rna_qc_metrics()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.compute_rna_qc_metrics_with_altexps"><code class="docutils literal notranslate"><span class="pre">compute_rna_qc_metrics_with_altexps()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.correct_mnn"><code class="docutils literal notranslate"><span class="pre">correct_mnn()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.correct_mnn_se"><code class="docutils literal notranslate"><span class="pre">correct_mnn_se()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.filter_adt_qc_metrics"><code class="docutils literal notranslate"><span class="pre">filter_adt_qc_metrics()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.filter_crispr_qc_metrics"><code class="docutils literal notranslate"><span class="pre">filter_crispr_qc_metrics()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.filter_rna_qc_metrics"><code class="docutils literal notranslate"><span class="pre">filter_rna_qc_metrics()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.fit_variance_trend"><code class="docutils literal notranslate"><span class="pre">fit_variance_trend()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.format_compute_adt_qc_metrics_result"><code class="docutils literal notranslate"><span class="pre">format_compute_adt_qc_metrics_result()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.format_compute_crispr_qc_metrics_result"><code class="docutils literal notranslate"><span class="pre">format_compute_crispr_qc_metrics_result()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.format_compute_rna_qc_metrics_result"><code class="docutils literal notranslate"><span class="pre">format_compute_rna_qc_metrics_result()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.format_score_markers_result"><code class="docutils literal notranslate"><span class="pre">format_score_markers_result()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.get_test_adt_data_se"><code class="docutils literal notranslate"><span class="pre">get_test_adt_data_se()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.get_test_crispr_data_se"><code class="docutils literal notranslate"><span class="pre">get_test_crispr_data_se()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.get_test_rna_data_se"><code class="docutils literal notranslate"><span class="pre">get_test_rna_data_se()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.model_gene_variances"><code class="docutils literal notranslate"><span class="pre">model_gene_variances()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.normalize_adt_counts_se"><code class="docutils literal notranslate"><span class="pre">normalize_adt_counts_se()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.normalize_counts"><code class="docutils literal notranslate"><span class="pre">normalize_counts()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.normalize_crispr_counts_se"><code class="docutils literal notranslate"><span class="pre">normalize_crispr_counts_se()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.normalize_rna_counts_se"><code class="docutils literal notranslate"><span class="pre">normalize_rna_counts_se()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.preview_markers"><code class="docutils literal notranslate"><span class="pre">preview_markers()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.quick_adt_qc_se"><code class="docutils literal notranslate"><span class="pre">quick_adt_qc_se()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.quick_crispr_qc_se"><code class="docutils literal notranslate"><span class="pre">quick_crispr_qc_se()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.quick_rna_qc_se"><code class="docutils literal notranslate"><span class="pre">quick_rna_qc_se()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.run_all_neighbor_steps"><code class="docutils literal notranslate"><span class="pre">run_all_neighbor_steps()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.run_all_neighbor_steps_se"><code class="docutils literal notranslate"><span class="pre">run_all_neighbor_steps_se()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.run_pca"><code class="docutils literal notranslate"><span class="pre">run_pca()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.run_pca_se"><code class="docutils literal notranslate"><span class="pre">run_pca_se()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.run_tsne"><code class="docutils literal notranslate"><span class="pre">run_tsne()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.run_tsne_se"><code class="docutils literal notranslate"><span class="pre">run_tsne_se()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.run_umap"><code class="docutils literal notranslate"><span class="pre">run_umap()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.run_umap_se"><code class="docutils literal notranslate"><span class="pre">run_umap_se()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.sanitize_size_factors"><code class="docutils literal notranslate"><span class="pre">sanitize_size_factors()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.scale_by_neighbors"><code class="docutils literal notranslate"><span class="pre">scale_by_neighbors()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.scale_by_neighbors_se"><code class="docutils literal notranslate"><span class="pre">scale_by_neighbors_se()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.score_gene_set"><code class="docutils literal notranslate"><span class="pre">score_gene_set()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.score_gene_set_se"><code class="docutils literal notranslate"><span class="pre">score_gene_set_se()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.score_markers"><code class="docutils literal notranslate"><span class="pre">score_markers()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.score_markers_se"><code class="docutils literal notranslate"><span class="pre">score_markers_se()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.subsample_by_neighbors"><code class="docutils literal notranslate"><span class="pre">subsample_by_neighbors()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.suggest_adt_qc_thresholds"><code class="docutils literal notranslate"><span class="pre">suggest_adt_qc_thresholds()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.suggest_crispr_qc_thresholds"><code class="docutils literal notranslate"><span class="pre">suggest_crispr_qc_thresholds()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.suggest_rna_qc_thresholds"><code class="docutils literal notranslate"><span class="pre">suggest_rna_qc_thresholds()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.summarize_effects"><code class="docutils literal notranslate"><span class="pre">summarize_effects()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.test_enrichment"><code class="docutils literal notranslate"><span class="pre">test_enrichment()</span></code></a></li>
<li><a class="reference internal" href="#scranpy.tsne_perplexity_to_neighbors"><code class="docutils literal notranslate"><span class="pre">tsne_perplexity_to_neighbors()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js?v=c02300dd"></script>
    <script src="../_static/doctools.js?v=888ff710"></script>
    <script src="../_static/sphinx_highlight.js?v=4825356b"></script>
    <script src="../_static/scripts/furo.js?v=46bd48cc"></script>
    </body>
</html>